diff --git a/Marlin/src/gcode/temperature/M303.cpp b/Marlin/src/gcode/temperature/M303.cpp
index 8473ecd20ca..f7012778548 100644
--- a/Marlin/src/gcode/temperature/M303.cpp
+++ b/Marlin/src/gcode/temperature/M303.cpp
@@ -37,7 +37,7 @@
  */
 void GcodeSuite::M303() {
 
-  const int8_t e = parser.byteval('E');
+  const int8_t e = parser.intval('E');
 
   if (!WITHIN(e, 0
     #if ENABLED(PIDTEMPBED)
