diff --git a/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java b/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
index b131f11fbc03..2e2497fe7cb8 100644
--- a/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
+++ b/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
@@ -420,6 +420,7 @@ public int compare(StoreFileMetaData o1, StoreFileMetaData o2) {
                                     for (StoreFileMetaData md : metadata) {
                                         logger.debug("{} checking integrity for file {} after remove corruption exception", shard.shardId(), md);
                                         if (store.checkIntegrityNoException(md) == false) { // we are corrupted on the primary -- fail!
+                                            shard.engine().failEngine("recovery", corruptIndexException);
                                             logger.warn("{} Corrupted file detected {} checksum mismatch", shard.shardId(), md);
                                             throw corruptIndexException;
                                         }
