diff --git a/Marlin/src/lcd/HD44780/ultralcd_HD44780.h b/Marlin/src/lcd/HD44780/ultralcd_HD44780.h
index 5ca36cb2263..c742f59e31b 100644
--- a/Marlin/src/lcd/HD44780/ultralcd_HD44780.h
+++ b/Marlin/src/lcd/HD44780/ultralcd_HD44780.h
@@ -90,7 +90,6 @@
 //https://github.com/mikeshub/SailfishLCD
 //uses the code directly from Sailfish
 
-  #include <LCD.h>
   #include <SailfishLCD.h>
   #define LCD_CLASS LiquidCrystalSerial
 
diff --git a/platformio.ini b/platformio.ini
index 5a298e42b4e..d683e2cd792 100644
--- a/platformio.ini
+++ b/platformio.ini
@@ -167,6 +167,7 @@ lib_deps          = Servo
   U8glib-HAL=https://github.com/MarlinFirmware/U8glib-HAL/archive/dev.zip
   TMCStepper@<1.0.0
   Adafruit NeoPixel=https://github.com/p3p/Adafruit_NeoPixel/archive/master.zip
+  https://github.com/mikeshub/SailfishLCD.git
 
 [env:LPC1769]
 platform          = https://github.com/p3p/pio-nxplpc-arduino-lpc176x/archive/master.zip
