diff --git a/Marlin/planner.cpp b/Marlin/planner.cpp
index 9d955574038..6daa22d2b0a 100644
--- a/Marlin/planner.cpp
+++ b/Marlin/planner.cpp
@@ -1252,7 +1252,7 @@ void Planner::_buffer_line(const float &a, const float &b, const float &c, const
     }
     else {
       block->use_advance_lead = true;
-      block->abs_adv_steps_multiplier8 = lround(extruder_advance_k * (de_float / mm_D_float) * block->nominal_speed / (float)block->nominal_rate * axis_steps_per_mm[Z_AXIS] * 256.0);
+      block->abs_adv_steps_multiplier8 = lround(extruder_advance_k * (de_float / mm_D_float) * block->nominal_speed / (float)block->nominal_rate * axis_steps_per_mm[E_AXIS] * 256.0);
     }
 
   #elif ENABLED(ADVANCE)
