diff --git a/third_party/py/python_configure.bzl b/third_party/py/python_configure.bzl
index 6f9a178a057..c87a2289f25 100644
--- a/third_party/py/python_configure.bzl
+++ b/third_party/py/python_configure.bzl
@@ -260,8 +260,8 @@ def _create_single_version_package(repository_ctx,
     # See https://docs.python.org/3/extending/windows.html
     if _is_windows(repository_ctx):
         python_include = _normalize_path(python_include)
-        python_import_lib_name = _get_python_import_lib_name, bin_path_key(
-            repository_ctx, python_bin)
+        python_import_lib_name = _get_python_import_lib_name(
+            repository_ctx, python_bin, bin_path_key)
         python_import_lib_src = python_include.rsplit(
             '/', 1)[0] + "/libs/" + python_import_lib_name
         python_import_lib_genrule = _symlink_genrule_for_dir(
