diff --git a/core/src/main/java/org/elasticsearch/search/fetch/subphase/ParentFieldSubFetchPhase.java b/core/src/main/java/org/elasticsearch/search/fetch/subphase/ParentFieldSubFetchPhase.java
index cb8310975fc1..5cd19096beb1 100644
--- a/core/src/main/java/org/elasticsearch/search/fetch/subphase/ParentFieldSubFetchPhase.java
+++ b/core/src/main/java/org/elasticsearch/search/fetch/subphase/ParentFieldSubFetchPhase.java
@@ -63,6 +63,10 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
     public static String getParentId(ParentFieldMapper fieldMapper, LeafReader reader, int docId) {
         try {
             SortedDocValues docValues = reader.getSortedDocValues(fieldMapper.name());
+            if (docValues == null) {
+                // hit has no _parent field.
+                return null;
+            }
             BytesRef parentId = docValues.get(docId);
             return parentId.length > 0 ? parentId.utf8ToString() : null;
         } catch (IOException e) {
