diff --git a/docs/reference/security/securing-communications/configuring-tls-docker.asciidoc b/docs/reference/security/securing-communications/configuring-tls-docker.asciidoc
index 1d23430e37ee..e4d5a2a09304 100644
--- a/docs/reference/security/securing-communications/configuring-tls-docker.asciidoc
+++ b/docs/reference/security/securing-communications/configuring-tls-docker.asciidoc
@@ -69,6 +69,7 @@ services:
     image: {docker-image}
     command: >
       bash -c '
+        yum install unzip -y
         if [[ ! -d config/certificates/certs ]]; then
           mkdir config/certificates/certs;
         fi;
