diff --git a/ext/soap/soap.c b/ext/soap/soap.c
index 1751f4d2f944..fd008282629e 100644
--- a/ext/soap/soap.c
+++ b/ext/soap/soap.c
@@ -303,7 +303,7 @@ ZEND_END_ARG_INFO()
 
 ZEND_BEGIN_ARG_INFO_EX(arginfo_soapserver_setclass, 0, 0, 1)
 	ZEND_ARG_INFO(0, class_name)
-	ZEND_ARG_INFO(0, args)
+	ZEND_ARG_VARIADIC_INFO(0, args)
 ZEND_END_ARG_INFO()
 
 ZEND_BEGIN_ARG_INFO_EX(arginfo_soapserver_setobject, 0, 0, 1)
