diff --git a/cmake/OpenCVFindLibsGUI.cmake b/cmake/OpenCVFindLibsGUI.cmake
index 3ff993fd466..82f7e0ff123 100644
--- a/cmake/OpenCVFindLibsGUI.cmake
+++ b/cmake/OpenCVFindLibsGUI.cmake
@@ -64,7 +64,7 @@ if(WITH_GTK AND NOT HAVE_QT)
   if(WITH_OPENGL AND NOT HAVE_GTK3)
     ocv_check_modules(GTKGLEXT gtkglext-1.0)
     if(HAVE_GTKGLEXT)
-      ocv_append_build_options(GTKGLEXT GTHREAD)
+      ocv_append_build_options(HIGHGUI GTKGLEXT)
     endif()
   endif()
 endif()
