diff --git a/build/buildcheck.sh b/build/buildcheck.sh
index 6601fc9943ce..8857a34b18f0 100755
--- a/build/buildcheck.sh
+++ b/build/buildcheck.sh
@@ -28,16 +28,16 @@ fi
 # autoconf 2.68 or newer
 ac_version=`$PHP_AUTOCONF --version 2>/dev/null|head -n 1|sed -e 's/^[^0-9]*//' -e 's/[a-z]* *$//'`
 if test -z "$ac_version"; then
-echo "buildconf: autoconf not found."
-echo "           You need autoconf version 2.68 or newer installed"
-echo "           to build PHP from Git."
+echo "buildconf: autoconf not found." >&2
+echo "           You need autoconf version 2.68 or newer installed" >&2
+echo "           to build PHP from Git." >&2
 exit 1
 fi
 IFS=.; set $ac_version; IFS=' '
 if test "$1" = "2" -a "$2" -lt "68" || test "$1" -lt "2"; then
-echo "buildconf: autoconf version $ac_version found."
-echo "           You need autoconf version 2.68 or newer installed"
-echo "           to build PHP from Git."
+echo "buildconf: autoconf version $ac_version found." >&2
+echo "           You need autoconf version 2.68 or newer installed" >&2
+echo "           to build PHP from Git." >&2
 exit 1
 else
 echo "buildconf: autoconf version $ac_version (ok)"
diff --git a/build/genif.sh b/build/genif.sh
index 1df7c9316466..a6eb627b0c5f 100755
--- a/build/genif.sh
+++ b/build/genif.sh
@@ -12,7 +12,7 @@ awk=$1
 shift
 
 if test -z "$infile" || test -z "$srcdir"; then
-	echo "please supply infile and srcdir"
+	echo "please supply infile and srcdir" >&2
 	exit 1
 fi
 
diff --git a/buildconf b/buildconf
index 23d10061f2d0..4b72ef8efba8 100755
--- a/buildconf
+++ b/buildconf
@@ -27,8 +27,8 @@ while test $# -gt 0; do
 done
 
 if test "$dev" = "0" -a "$devok" = "0"; then
-  echo "You should not run buildconf in a release package."
-  echo "use buildconf --force to override this check."
+  echo "You should not run buildconf in a release package." >&2
+  echo "use buildconf --force to override this check." >&2
   exit 1
 fi
 
diff --git a/makedist b/makedist
index ea17bafa7748..452768683aa4 100755
--- a/makedist
+++ b/makedist
@@ -65,8 +65,8 @@ DIR=php-$VER
 DIRPATH=$MY_OLDPWD/$DIR
 
 if test -d "$DIRPATH"; then
-    echo "The directory $DIR"
-    echo "already exists, rename or remove it and run makedist again."
+    echo "The directory $DIR" >&2
+    echo "already exists, rename or remove it and run makedist again." >&2
     exit 1
 fi
 
