diff --git a/server/src/main/java/org/elasticsearch/search/rescore/QueryRescorerBuilder.java b/server/src/main/java/org/elasticsearch/search/rescore/QueryRescorerBuilder.java
index 5690e584eac2..02e748dfbc95 100644
--- a/server/src/main/java/org/elasticsearch/search/rescore/QueryRescorerBuilder.java
+++ b/server/src/main/java/org/elasticsearch/search/rescore/QueryRescorerBuilder.java
@@ -250,6 +250,9 @@ public QueryRescorerBuilder rewrite(QueryRewriteContext ctx) throws IOException
         queryRescoreBuilder.setQueryWeight(queryWeight);
         queryRescoreBuilder.setRescoreQueryWeight(rescoreQueryWeight);
         queryRescoreBuilder.setScoreMode(scoreMode);
+        if (windowSize() != null) {
+            queryRescoreBuilder.windowSize(windowSize());
+        }
         return queryRescoreBuilder;
     }
 }
diff --git a/server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java b/server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java
index 24621a12d39e..0f647353e95a 100644
--- a/server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java
+++ b/server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java
@@ -200,11 +200,13 @@ public MappedFieldType fieldMapper(String name) {
         rescoreBuilder.setQueryWeight(randomFloat());
         rescoreBuilder.setRescoreQueryWeight(randomFloat());
         rescoreBuilder.setScoreMode(QueryRescoreMode.Max);
+        rescoreBuilder.windowSize(randomIntBetween(0, 100));
 
         QueryRescorerBuilder rescoreRewritten = rescoreBuilder.rewrite(mockShardContext);
         assertEquals(rescoreRewritten.getQueryWeight(), rescoreBuilder.getQueryWeight(), 0.01f);
         assertEquals(rescoreRewritten.getRescoreQueryWeight(), rescoreBuilder.getRescoreQueryWeight(), 0.01f);
         assertEquals(rescoreRewritten.getScoreMode(), rescoreBuilder.getScoreMode());
+        assertEquals(rescoreRewritten.windowSize(), rescoreBuilder.windowSize());
     }
 
     /**
