diff --git a/doc/source/conf.py b/doc/source/conf.py
index cce0716d54c..91d413bd70a 100644
--- a/doc/source/conf.py
+++ b/doc/source/conf.py
@@ -126,6 +126,8 @@
 pngmath_use_preview = True
 pngmath_dvipng_args = ['-gamma', '1.5', '-D', '96', '-bg', 'Transparent']
 
+plot_html_show_formats = False
+plot_html_show_source_link = False
 
 # -----------------------------------------------------------------------------
 # LaTeX output
