diff --git a/numpy/core/arrayprint.py b/numpy/core/arrayprint.py
index e4be810b921..e4a048a42d8 100644
--- a/numpy/core/arrayprint.py
+++ b/numpy/core/arrayprint.py
@@ -427,12 +427,14 @@ def wrapper(self, *args, **kwargs):
 # gracefully handle recursive calls, when object arrays contain themselves
 @_recursive_guard()
 def _array2string(a, options, separator=' ', prefix=""):
+    # The formatter __init__s cannot deal with subclasses yet
+    data = asarray(a)
+
     if a.size > options['threshold']:
         summary_insert = "..."
-        data = _leading_trailing(a)
+        data = _leading_trailing(data)
     else:
         summary_insert = ""
-        data = asarray(a)
 
     # find the right formatting function for the array
     format_function = _get_format_function(data, **options)
