diff --git a/ext/standard/config.m4 b/ext/standard/config.m4
index 3b668a2724a8..52faa91cb5a1 100644
--- a/ext/standard/config.m4
+++ b/ext/standard/config.m4
@@ -409,11 +409,6 @@ dnl Check for arc4random on BSD systems
 dnl
 AC_CHECK_DECLS([arc4random_buf])
 
-dnl
-dnl Check for getrandom on newer Linux kernels
-dnl
-AC_CHECK_DECLS([getrandom])
-
 dnl
 dnl Check for argon2
 dnl
diff --git a/ext/standard/random.c b/ext/standard/random.c
index a5a5940dcfa8..5896b3171bf6 100644
--- a/ext/standard/random.c
+++ b/ext/standard/random.c
@@ -93,7 +93,7 @@ PHPAPI int php_random_bytes(void *bytes, size_t size, zend_bool should_throw)
 	}
 #elif HAVE_DECL_ARC4RANDOM_BUF && ((defined(__OpenBSD__) && OpenBSD >= 201405) || (defined(__NetBSD__) && __NetBSD_Version__ >= 700000001))
 	arc4random_buf(bytes, size);
-#elif HAVE_DECL_GETRANDOM
+#elif defined(__linux__) && defined(SYS_getrandom)
 	/* Linux getrandom(2) syscall */
 	size_t read_bytes = 0;
 	size_t amount_to_read = 0;
