diff --git a/modules/objdetect/CMakeLists.txt b/modules/objdetect/CMakeLists.txt
index a51740c280d..414e5780999 100644
--- a/modules/objdetect/CMakeLists.txt
+++ b/modules/objdetect/CMakeLists.txt
@@ -4,5 +4,5 @@ ocv_define_module(objdetect opencv_core opencv_imgproc opencv_calib3d WRAP java
 if(HAVE_QUIRC)
     get_property(QUIRC_INCLUDE GLOBAL PROPERTY QUIRC_INCLUDE_DIR)
     ocv_include_directories(${QUIRC_INCLUDE})
-    ocv_target_link_libraries(${PROJECT_NAME} quirc)
+    ocv_target_link_libraries(${the_module} quirc)
 endif()
