diff --git a/ext/zlib/tests/bug75299.phpt b/ext/zlib/tests/bug75299.phpt
new file mode 100644
index 000000000000..1886b3ecafcd
--- /dev/null
+++ b/ext/zlib/tests/bug75299.phpt
@@ -0,0 +1,20 @@
+--TEST--
+Bug #75299 Wrong reflection on inflate_init and inflate_add
+--SKIPIF--
+<?php if (!extension_loaded("zlib") || !extension_loaded('reflection')) die("skip"); ?>
+--FILE--
+<?php
+$r = new ReflectionFunction('inflate_init');
+var_dump($r->getNumberOfRequiredParameters());
+var_dump($r->getNumberOfParameters());
+$r = new ReflectionFunction('inflate_add');
+var_dump($r->getNumberOfRequiredParameters());
+var_dump($r->getNumberOfParameters());
+?>
+===DONE===
+--EXPECT--
+int(1)
+int(2)
+int(2)
+int(3)
+===DONE===
diff --git a/ext/zlib/zlib.c b/ext/zlib/zlib.c
index b3c1a88fc508..6f2519d00ff3 100644
--- a/ext/zlib/zlib.c
+++ b/ext/zlib/zlib.c
@@ -1315,11 +1315,13 @@ ZEND_END_ARG_INFO()
 
 ZEND_BEGIN_ARG_INFO_EX(arginfo_inflate_init, 0, 0, 1)
 	ZEND_ARG_INFO(0, encoding)
+	ZEND_ARG_INFO(0, options)
 ZEND_END_ARG_INFO()
 
 ZEND_BEGIN_ARG_INFO_EX(arginfo_inflate_add, 0, 0, 2)
-	ZEND_ARG_INFO(0, resource)
-	ZEND_ARG_INFO(0, flush_behavior)
+	ZEND_ARG_INFO(0, context)
+	ZEND_ARG_INFO(0, encoded_data)
+	ZEND_ARG_INFO(0, flush_mode)
 ZEND_END_ARG_INFO()
 
 /* }}} */
