diff --git a/ext/mysqlnd/mysqlnd_priv.h b/ext/mysqlnd/mysqlnd_priv.h
index 9b898bed42f9..779fbc2b5ce2 100644
--- a/ext/mysqlnd/mysqlnd_priv.h
+++ b/ext/mysqlnd/mysqlnd_priv.h
@@ -105,7 +105,9 @@
 	} else { \
 		(error_info).error_no = (a); \
 		strlcpy((error_info).sqlstate, (b), sizeof((error_info).sqlstate)); \
-		strlcpy((error_info).error, (c), sizeof((error_info).error)); \
+		if ((error_info).error != (c)) { \
+			strlcpy((error_info).error, (c), sizeof((error_info).error)); \
+		} \
 		if ((error_info).error_list) {\
 			MYSQLND_ERROR_LIST_ELEMENT error_for_the_list = {0}; \
 																	\
