diff --git a/numpy/core/src/npymath/npy_math_complex.c.src b/numpy/core/src/npymath/npy_math_complex.c.src
index a82a6737c7f..f2b3913944c 100644
--- a/numpy/core/src/npymath/npy_math_complex.c.src
+++ b/numpy/core/src/npymath/npy_math_complex.c.src
@@ -184,8 +184,6 @@ npy_carg@c@(@ctype@ z)
 #define SCALED_CEXP_LOWERL 11357.216553474703895L
 #define SCALED_CEXP_UPPERL 22756.021937783004509L
 
-#ifndef HAVE_CEXP@C@
-
 static
 @ctype@
 _npy_scaled_cexp@c@(@type@ x, @type@ y, npy_int expt)
@@ -212,6 +210,7 @@ _npy_scaled_cexp@c@(@type@ x, @type@ y, npy_int expt)
                          npy_ldexp@c@(mant * mantsin, expt + exsin));
 }
 
+#ifndef HAVE_CEXP@C@
 @ctype@
 npy_cexp@c@(@ctype@ z)
 {
