diff --git a/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java b/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java
index abf6696efb40..2a8c525095da 100644
--- a/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java
+++ b/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java
@@ -1272,7 +1272,7 @@ private void writeValue(Object value) throws IOException {
             generator.writeEndArray();
         } else if (value instanceof short[]) {
             generator.writeStartArray();
-            for (float v : (short[]) value) {
+            for (short v : (short[]) value) {
                 generator.writeNumber(v);
             }
             generator.writeEndArray();
