diff --git a/wscript b/wscript
index 6e8893c8c2e0..d8c6ef329c3f 100644
--- a/wscript
+++ b/wscript
@@ -26,7 +26,8 @@ from waflib import Build, ConfigSet, Context, Utils
 def init(ctx):
     env = ConfigSet.ConfigSet()
     try:
-        env.load('build/c4che/_cache.py')
+        p = os.path.join(Context.out_dir, Build.CACHE_DIR, Build.CACHE_SUFFIX)
+        env.load(p)
     except:
         return
 
