diff --git a/modules/python/src2/gen2.py b/modules/python/src2/gen2.py
index 1b3c329fbf8..69d03a66fd4 100755
--- a/modules/python/src2/gen2.py
+++ b/modules/python/src2/gen2.py
@@ -73,7 +73,7 @@
             return true;
         }
         ${mappable_code}
-        failmsg("Expected ${cname} for argument '%%s'", name);
+        failmsg("Expected ${cname} for argument '%s'", name);
         return false;
     }
 };
