diff --git a/plugins/repository-gcs/build.gradle b/plugins/repository-gcs/build.gradle
index d35a248f5a5e..b86e5fb81f73 100644
--- a/plugins/repository-gcs/build.gradle
+++ b/plugins/repository-gcs/build.gradle
@@ -23,8 +23,8 @@ esplugin {
 }
 
 dependencies {
-  compile 'com.google.cloud:google-cloud-storage:1.40.0'
-  compile 'com.google.cloud:google-cloud-core:1.40.0'
+  compile 'com.google.cloud:google-cloud-storage:1.55.0'
+  compile 'com.google.cloud:google-cloud-core:1.55.0'
   compile 'com.google.guava:guava:20.0'
   compile "joda-time:joda-time:${versions.joda}"
   compile 'com.google.http-client:google-http-client:1.24.1'
@@ -40,7 +40,7 @@ dependencies {
   compile 'com.google.code.gson:gson:2.7'
   compile 'com.google.api.grpc:proto-google-common-protos:1.12.0'
   compile 'com.google.api.grpc:proto-google-iam-v1:0.12.0'
-  compile 'com.google.cloud:google-cloud-core-http:1.40.0'
+  compile 'com.google.cloud:google-cloud-core-http:1.55.0'
   compile 'com.google.auth:google-auth-library-credentials:0.10.0'
   compile 'com.google.auth:google-auth-library-oauth2-http:0.10.0'
   compile 'com.google.oauth-client:google-oauth-client:1.24.1'
diff --git a/plugins/repository-gcs/licenses/google-cloud-core-1.40.0.jar.sha1 b/plugins/repository-gcs/licenses/google-cloud-core-1.40.0.jar.sha1
deleted file mode 100644
index 7562ead12e9f..000000000000
--- a/plugins/repository-gcs/licenses/google-cloud-core-1.40.0.jar.sha1
+++ /dev/null
@@ -1 +0,0 @@
-4985701f989030e262cf8f4e38cc954115f5b082
\ No newline at end of file
diff --git a/plugins/repository-gcs/licenses/google-cloud-core-1.55.0.jar.sha1 b/plugins/repository-gcs/licenses/google-cloud-core-1.55.0.jar.sha1
new file mode 100644
index 000000000000..7c00bf52c419
--- /dev/null
+++ b/plugins/repository-gcs/licenses/google-cloud-core-1.55.0.jar.sha1
@@ -0,0 +1 @@
+9e50a2a559128b7938cfd6598753d4c7383472dc
\ No newline at end of file
diff --git a/plugins/repository-gcs/licenses/google-cloud-core-http-1.40.0.jar.sha1 b/plugins/repository-gcs/licenses/google-cloud-core-http-1.40.0.jar.sha1
deleted file mode 100644
index 2761bfdc745c..000000000000
--- a/plugins/repository-gcs/licenses/google-cloud-core-http-1.40.0.jar.sha1
+++ /dev/null
@@ -1 +0,0 @@
-67f5806beda32894f1e6c9527925b64199fd2e4f
\ No newline at end of file
diff --git a/plugins/repository-gcs/licenses/google-cloud-core-http-1.55.0.jar.sha1 b/plugins/repository-gcs/licenses/google-cloud-core-http-1.55.0.jar.sha1
new file mode 100644
index 000000000000..f4179201bcca
--- /dev/null
+++ b/plugins/repository-gcs/licenses/google-cloud-core-http-1.55.0.jar.sha1
@@ -0,0 +1 @@
+f26862445efffd8cb3a7f4b1f2a91b7c5143ee1f
\ No newline at end of file
diff --git a/plugins/repository-gcs/licenses/google-cloud-storage-1.40.0.jar.sha1 b/plugins/repository-gcs/licenses/google-cloud-storage-1.40.0.jar.sha1
deleted file mode 100644
index 33e83b73712f..000000000000
--- a/plugins/repository-gcs/licenses/google-cloud-storage-1.40.0.jar.sha1
+++ /dev/null
@@ -1 +0,0 @@
-fabefef46f07d1e334123f0de17702708b4dfbd1
\ No newline at end of file
diff --git a/plugins/repository-gcs/licenses/google-cloud-storage-1.55.0.jar.sha1 b/plugins/repository-gcs/licenses/google-cloud-storage-1.55.0.jar.sha1
new file mode 100644
index 000000000000..abe0065ab5c8
--- /dev/null
+++ b/plugins/repository-gcs/licenses/google-cloud-storage-1.55.0.jar.sha1
@@ -0,0 +1 @@
+ca19f55eeb96609243bf3a15fdafd497432f6673
\ No newline at end of file
diff --git a/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java b/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java
index cf7395ea1f1f..97c7e2ab76bd 100644
--- a/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java
+++ b/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java
@@ -80,6 +80,11 @@ public Bucket get(String bucket, BucketGetOption... options) {
         }
     }
 
+    @Override
+    public Bucket lockRetentionPolicy(final BucketInfo bucket, final BucketTargetOption... options) {
+        return null;
+    }
+
     @Override
     public Blob get(BlobId blob) {
         if (bucketName.equals(blob.getBucket())) {
