diff --git a/Tools/autotest/default_params/copter-AVC2013.parm b/Tools/autotest/default_params/copter-AVC2013.parm
index 440fccfa5df2..f2e12d078568 100644
--- a/Tools/autotest/default_params/copter-AVC2013.parm
+++ b/Tools/autotest/default_params/copter-AVC2013.parm
@@ -13,6 +13,9 @@ COMPASS_OFS_Z   -18
 COMPASS_OFS2_X   5
 COMPASS_OFS2_Y   13
 COMPASS_OFS2_Z   -18
+COMPASS_OFS3_X   5
+COMPASS_OFS3_Y   13
+COMPASS_OFS3_Z   -18
 RC1_MAX         2000.000000
 RC1_MIN         1000.000000
 RC1_TRIM        1500.000000
diff --git a/Tools/autotest/default_params/copter.parm b/Tools/autotest/default_params/copter.parm
index 6c0cd4684281..0d78343cb6d5 100644
--- a/Tools/autotest/default_params/copter.parm
+++ b/Tools/autotest/default_params/copter.parm
@@ -10,6 +10,9 @@ COMPASS_OFS_Z   -18
 COMPASS_OFS2_X   5
 COMPASS_OFS2_Y   13
 COMPASS_OFS2_Z   -18
+COMPASS_OFS3_X   5
+COMPASS_OFS3_Y   13
+COMPASS_OFS3_Z   -18
 FENCE_RADIUS    150
 FRAME_CLASS     1
 RC1_MAX         2000.000000
diff --git a/Tools/autotest/default_params/firefly.parm b/Tools/autotest/default_params/firefly.parm
index 1c59f7a9642b..27fe1485e1fa 100644
--- a/Tools/autotest/default_params/firefly.parm
+++ b/Tools/autotest/default_params/firefly.parm
@@ -15,9 +15,9 @@ AUTOTUNE_LEVEL   8
 COMPASS_OFS2_X   -0.420265
 COMPASS_OFS2_Y   -0.726942
 COMPASS_OFS2_Z   6.665476
-COMPASS_OFS3_X   0
-COMPASS_OFS3_Y   0
-COMPASS_OFS3_Z   0
+COMPASS_OFS3_X   -0.420265
+COMPASS_OFS3_Y   -0.726942
+COMPASS_OFS3_Z   6.665476
 COMPASS_OFS_X    -0.453570
 COMPASS_OFS_Y    -0.585846
 COMPASS_OFS_Z    6.815743
diff --git a/Tools/autotest/default_params/quadplane-cl84.parm b/Tools/autotest/default_params/quadplane-cl84.parm
index 3cbce18ce79c..c7ce7d839970 100644
--- a/Tools/autotest/default_params/quadplane-cl84.parm
+++ b/Tools/autotest/default_params/quadplane-cl84.parm
@@ -6,9 +6,9 @@ AUTOTUNE_LEVEL   8
 COMPASS_OFS2_X   -0.420265
 COMPASS_OFS2_Y   -0.726942
 COMPASS_OFS2_Z   6.665476
-COMPASS_OFS3_X   0
-COMPASS_OFS3_Y   0
-COMPASS_OFS3_Z   0
+COMPASS_OFS3_X   -0.420265
+COMPASS_OFS3_Y   -0.726942
+COMPASS_OFS3_Z   6.665476
 COMPASS_OFS_X    -0.453570
 COMPASS_OFS_Y    -0.585846
 COMPASS_OFS_Z    6.815743
diff --git a/Tools/autotest/default_params/quadplane-tilttri.parm b/Tools/autotest/default_params/quadplane-tilttri.parm
index 8cde89726f5f..517db738f5e7 100644
--- a/Tools/autotest/default_params/quadplane-tilttri.parm
+++ b/Tools/autotest/default_params/quadplane-tilttri.parm
@@ -7,9 +7,9 @@ AUTOTUNE_LEVEL   8
 COMPASS_OFS2_X   -0.420265
 COMPASS_OFS2_Y   -0.726942
 COMPASS_OFS2_Z   6.665476
-COMPASS_OFS3_X   0
-COMPASS_OFS3_Y   0
-COMPASS_OFS3_Z   0
+COMPASS_OFS3_X   -0.420265
+COMPASS_OFS3_Y   -0.726942
+COMPASS_OFS3_Z   6.665476
 COMPASS_OFS_X    -0.453570
 COMPASS_OFS_Y    -0.585846
 COMPASS_OFS_Z    6.815743
diff --git a/Tools/autotest/default_params/quadplane-tilttrivec.parm b/Tools/autotest/default_params/quadplane-tilttrivec.parm
index 33fe48c3cd14..949eb5bc6179 100644
--- a/Tools/autotest/default_params/quadplane-tilttrivec.parm
+++ b/Tools/autotest/default_params/quadplane-tilttrivec.parm
@@ -7,9 +7,9 @@ AUTOTUNE_LEVEL   8
 COMPASS_OFS2_X   -0.420265
 COMPASS_OFS2_Y   -0.726942
 COMPASS_OFS2_Z   6.665476
-COMPASS_OFS3_X   0
-COMPASS_OFS3_Y   0
-COMPASS_OFS3_Z   0
+COMPASS_OFS3_X   -0.420265
+COMPASS_OFS3_Y   -0.726942
+COMPASS_OFS3_Z   6.665476
 COMPASS_OFS_X    -0.453570
 COMPASS_OFS_Y    -0.585846
 COMPASS_OFS_Z    6.815743
diff --git a/Tools/autotest/default_params/quadplane-tri.parm b/Tools/autotest/default_params/quadplane-tri.parm
index b4a87a91c4b2..555383113fb0 100644
--- a/Tools/autotest/default_params/quadplane-tri.parm
+++ b/Tools/autotest/default_params/quadplane-tri.parm
@@ -7,9 +7,9 @@ AUTOTUNE_LEVEL   8
 COMPASS_OFS2_X   -0.420265
 COMPASS_OFS2_Y   -0.726942
 COMPASS_OFS2_Z   6.665476
-COMPASS_OFS3_X   0
-COMPASS_OFS3_Y   0
-COMPASS_OFS3_Z   0
+COMPASS_OFS3_X   -0.420265
+COMPASS_OFS3_Y   -0.726942
+COMPASS_OFS3_Z   6.665476
 COMPASS_OFS_X    -0.453570
 COMPASS_OFS_Y    -0.585846
 COMPASS_OFS_Z    6.815743
diff --git a/Tools/autotest/default_params/quadplane.parm b/Tools/autotest/default_params/quadplane.parm
index 1b683defef2f..ade7761534d5 100644
--- a/Tools/autotest/default_params/quadplane.parm
+++ b/Tools/autotest/default_params/quadplane.parm
@@ -7,9 +7,9 @@ AUTOTUNE_LEVEL   8
 COMPASS_OFS2_X   -0.420265
 COMPASS_OFS2_Y   -0.726942
 COMPASS_OFS2_Z   6.665476
-COMPASS_OFS3_X   0
-COMPASS_OFS3_Y   0
-COMPASS_OFS3_Z   0
+COMPASS_OFS3_X   -0.420265
+COMPASS_OFS3_Y   -0.726942
+COMPASS_OFS3_Z   6.665476
 COMPASS_OFS_X    -0.453570
 COMPASS_OFS_Y    -0.585846
 COMPASS_OFS_Z    6.815743
diff --git a/Tools/autotest/default_params/sub.parm b/Tools/autotest/default_params/sub.parm
index c2c5c1aebd51..83d24e17f88e 100644
--- a/Tools/autotest/default_params/sub.parm
+++ b/Tools/autotest/default_params/sub.parm
@@ -38,6 +38,9 @@ COMPASS_OFS_Z   -18
 COMPASS_OFS2_X   5
 COMPASS_OFS2_Y   13
 COMPASS_OFS2_Z   -18
+COMPASS_OFS3_X   5
+COMPASS_OFS3_Y   13
+COMPASS_OFS3_Z   -18
 INS_ACCOFFS_X   0.001
 INS_ACCOFFS_Y   0.001
 INS_ACCOFFS_Z   0.001
diff --git a/libraries/AP_Compass/AP_Compass_SITL.h b/libraries/AP_Compass/AP_Compass_SITL.h
index 0fa9f0853f74..540d366f458c 100644
--- a/libraries/AP_Compass/AP_Compass_SITL.h
+++ b/libraries/AP_Compass/AP_Compass_SITL.h
@@ -9,7 +9,7 @@
 #include <AP_Math/AP_Math.h>
 #include <AP_Declination/AP_Declination.h>
 
-#define SITL_NUM_COMPASSES 2
+#define SITL_NUM_COMPASSES 3
 
 class AP_Compass_SITL : public AP_Compass_Backend {
 public:
diff --git a/libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp b/libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp
index 207a41b3e44f..4ec5450812ca 100644
--- a/libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp
+++ b/libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp
@@ -247,6 +247,7 @@ void NavEKF2_core::readMagData()
                     magStateResetRequest = true;
                     // declare the field unlearned so that the reset request will be obeyed
                     magFieldLearned = false;
+                    break;
                 }
             }
         }
diff --git a/libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp b/libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp
index e98583efdb2a..75ec47a2f8a2 100644
--- a/libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp
+++ b/libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp
@@ -292,6 +292,7 @@ void NavEKF3_core::readMagData()
                     magStateResetRequest = true;
                     // declare the field unlearned so that the reset request will be obeyed
                     magFieldLearned = false;
+                    break;
                 }
             }
         }
