diff --git a/core/src/main/java/org/elasticsearch/common/settings/Setting.java b/core/src/main/java/org/elasticsearch/common/settings/Setting.java
index 920e1d7f5fc8..1d67008ab0b6 100644
--- a/core/src/main/java/org/elasticsearch/common/settings/Setting.java
+++ b/core/src/main/java/org/elasticsearch/common/settings/Setting.java
@@ -107,9 +107,6 @@
         IndexScope
     }
 
-    private static final ESLogger logger = Loggers.getLogger(Setting.class);
-    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(logger);
-
     private final Key key;
     protected final Function<Settings, String> defaultValue;
     @Nullable
@@ -322,6 +319,7 @@ public String getRaw(Settings settings) {
         // They're using the setting, so we need to tell them to stop
         if (this.isDeprecated() && this.exists(settings)) {
             // It would be convenient to show its replacement key, but replacement is often not so simple
+            final DeprecationLogger deprecationLogger = new DeprecationLogger(Loggers.getLogger(getClass()));
             deprecationLogger.deprecated("[{}] setting was deprecated in Elasticsearch and it will be removed in a future release! " +
                     "See the breaking changes lists in the documentation for details", getKey());
         }
