diff --git a/core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java b/core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java
index c9e38f1d94eb..6fbf814b0066 100644
--- a/core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java
+++ b/core/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java
@@ -329,8 +329,14 @@ public static XContentType xContentType(BytesReference bytes) {
             return XContentType.CBOR;
         }
 
+        int jsonStart = 0;
+        // JSON may be preceded by UTF-8 BOM
+        if (length > 3 && first == (byte) 0xEF && bytes.get(1) == (byte) 0xBB && bytes.get(2) == (byte) 0xBF) {
+            jsonStart = 3;
+        }
+
         // a last chance for JSON
-        for (int i = 0; i < length; i++) {
+        for (int i = jsonStart; i < length; i++) {
             byte b = bytes.get(i);
             if (b == '{') {
                 return XContentType.JSON;
diff --git a/core/src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java b/core/src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java
index fb12c7508c22..7489ea8f9dc6 100644
--- a/core/src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java
+++ b/core/src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java
@@ -97,4 +97,18 @@ public void testEmptyStream() throws Exception {
         is = new ByteArrayInputStream(new byte[] {(byte) 1});
         assertNull(XContentFactory.xContentType(is));
     }
+    
+    public void testJsonFromBytesOptionallyPrecededByUtf8Bom() throws Exception {
+        byte[] bytes = new byte[] {(byte) '{', (byte) '}'};
+        assertThat(XContentFactory.xContentType(bytes), equalTo(XContentType.JSON));
+        
+        bytes = new byte[] {(byte) 0x20, (byte) '{', (byte) '}'};
+        assertThat(XContentFactory.xContentType(bytes), equalTo(XContentType.JSON));
+        
+        bytes = new byte[] {(byte) 0xef, (byte) 0xbb, (byte) 0xbf, (byte) '{', (byte) '}'};
+        assertThat(XContentFactory.xContentType(bytes), equalTo(XContentType.JSON));
+        
+        bytes = new byte[] {(byte) 0xef, (byte) 0xbb, (byte) 0xbf, (byte) 0x20, (byte) '{', (byte) '}'};
+        assertThat(XContentFactory.xContentType(bytes), equalTo(XContentType.JSON));
+    }
 }
