diff --git a/plugins/transport-nio/build.gradle b/plugins/transport-nio/build.gradle
index 197f914f5724..1cc57c3833a6 100644
--- a/plugins/transport-nio/build.gradle
+++ b/plugins/transport-nio/build.gradle
@@ -21,6 +21,7 @@ apply plugin: "nebula.maven-scm"
 esplugin {
     description 'The nio transport.'
     classname 'org.elasticsearch.transport.nio.NioTransportPlugin'
+    hasClientJar = true
 }
 
 compileJava.options.compilerArgs << "-Xlint:-try"
