diff --git a/numpy/core/einsumfunc.py b/numpy/core/einsumfunc.py
index c5b37b7e2a4..f7955516644 100644
--- a/numpy/core/einsumfunc.py
+++ b/numpy/core/einsumfunc.py
@@ -1056,8 +1056,8 @@ def einsum(*operands, **kwargs):
 
     """
 
-    # Grab non-einsum kwargs
-    optimize_arg = kwargs.pop('optimize', True)
+    # Grab non-einsum kwargs; never optimize 2-argument case.
+    optimize_arg = kwargs.pop('optimize', len(operands) > 3)
 
     # If no optimization, run pure einsum
     if optimize_arg is False:
