diff --git a/src/main/java/org/elasticsearch/index/mapper/geo/GeoPointFieldMapper.java b/src/main/java/org/elasticsearch/index/mapper/geo/GeoPointFieldMapper.java
index 765bb8d8f540..a0c4f0c0b225 100644
--- a/src/main/java/org/elasticsearch/index/mapper/geo/GeoPointFieldMapper.java
+++ b/src/main/java/org/elasticsearch/index/mapper/geo/GeoPointFieldMapper.java
@@ -525,7 +525,7 @@ public void parse(ParseContext context) throws IOException {
                 }
             } else if (token == XContentParser.Token.VALUE_STRING) {
                 parsePointFromString(context, sparse, context.parser().text());
-            } else {
+            } else if (token != XContentParser.Token.VALUE_NULL) {
                 parse(context, GeoUtils.parseGeoPoint(context.parser(), sparse), null);
             }
         }
diff --git a/src/test/java/org/elasticsearch/index/mapper/geo/GeohashMappingGeoPointTests.java b/src/test/java/org/elasticsearch/index/mapper/geo/GeohashMappingGeoPointTests.java
index ffa8900dde4e..3aa9823cb046 100644
--- a/src/test/java/org/elasticsearch/index/mapper/geo/GeohashMappingGeoPointTests.java
+++ b/src/test/java/org/elasticsearch/index/mapper/geo/GeohashMappingGeoPointTests.java
@@ -117,4 +117,21 @@ public void testGeoHashPrecisionAsLength() throws Exception {
         GeoPointFieldMapper geoPointFieldMapper = (GeoPointFieldMapper) mapper;
         assertThat(geoPointFieldMapper.geoHashPrecision(), is(10));
     }
+
+    @Test
+    public void testNullValue() throws Exception {
+        String mapping = XContentFactory.jsonBuilder().startObject().startObject("type")
+                .startObject("properties").startObject("point").field("type", "geo_point").endObject().endObject()
+                .endObject().endObject().string();
+
+        DocumentMapper defaultMapper = MapperTestUtils.newParser().parse(mapping);
+
+        ParsedDocument doc = defaultMapper.parse("type", "1", XContentFactory.jsonBuilder()
+                .startObject()
+                .field("point", (Object) null)
+                .endObject()
+                .bytes());
+
+        assertThat(doc.rootDoc().get("point"), nullValue());
+    }
 }
