diff --git a/Marlin/stepper_indirection.cpp b/Marlin/stepper_indirection.cpp
index 778f448937a..8f9f6adf21d 100644
--- a/Marlin/stepper_indirection.cpp
+++ b/Marlin/stepper_indirection.cpp
@@ -85,8 +85,8 @@
   #endif
 
   #define _TMC_INIT(A) do{ \
-    stepper##A.setMicrosteps(A##_MICROSTEPS);
-    stepper##A.start();
+    stepper##A.setMicrosteps(A##_MICROSTEPS); \
+    stepper##A.start(); \
   } while(0)
 
   void tmc_init() {
