diff --git a/Marlin/nozzle.cpp b/Marlin/nozzle.cpp
index 37a4db591ca..6e080307174 100644
--- a/Marlin/nozzle.cpp
+++ b/Marlin/nozzle.cpp
@@ -123,8 +123,8 @@
     for (uint8_t s = 0; s < strokes; s++)
       for (uint8_t i = 0; i < NOZZLE_CLEAN_CIRCLE_FN; i++)
         do_blocking_move_to_xy(
-          middle.x + sin((M_2_PI / NOZZLE_CLEAN_CIRCLE_FN) * i) * radius,
-          middle.y + cos((M_2_PI / NOZZLE_CLEAN_CIRCLE_FN) * i) * radius
+          middle.x + sin((2.0 * M_PI / NOZZLE_CLEAN_CIRCLE_FN) * i) * radius,
+          middle.y + cos((2.0 * M_PI / NOZZLE_CLEAN_CIRCLE_FN) * i) * radius
         );
 
     // Let's be safe
