diff --git a/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java b/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java
index 5f882c6c1e7a..621bfb9868fc 100644
--- a/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java
+++ b/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java
@@ -318,7 +318,7 @@ public void run() {
                 timeInQueue = runnable.timeSinceCreatedInMillis();
             } else {
                 assert false : "expected TimedPrioritizedRunnable got " + task.getClass();
-                source = "unknown";
+                source = "unknown [" + task.getClass() + "]";
                 timeInQueue = 0;
             }
 
