diff --git a/main/fastcgi.c b/main/fastcgi.c
index dd7c7ddeb9bf..db3572816f07 100644
--- a/main/fastcgi.c
+++ b/main/fastcgi.c
@@ -462,6 +462,11 @@ void fcgi_terminate(void)
 	in_shutdown = 1;
 }
 
+void fcgi_request_set_keep(fcgi_request *req, int new_value)
+{
+	req->keep = new_value;
+}
+
 #ifndef HAVE_ATTRIBUTE_WEAK
 void fcgi_set_logger(fcgi_logger lg) {
 	fcgi_log = lg;
@@ -1469,6 +1474,7 @@ int fcgi_accept_request(fcgi_request *req)
 			return -1;
 		}
 		if (fcgi_read_request(req)) {
+			req->hook.on_read();
 #ifdef _WIN32
 			if (is_impersonate && !req->tcp) {
 				pipe = (HANDLE)_get_osfhandle(req->fd);
diff --git a/main/fastcgi.h b/main/fastcgi.h
index bba64016d894..925d60c9bb6c 100644
--- a/main/fastcgi.h
+++ b/main/fastcgi.h
@@ -102,6 +102,7 @@ int fcgi_accept_request(fcgi_request *req);
 int fcgi_finish_request(fcgi_request *req, int force_close);
 const char *fcgi_get_last_client_ip();
 void fcgi_set_in_shutdown(int new_value);
+void fcgi_request_set_keep(fcgi_request *req, int new_value);
 
 #ifndef HAVE_ATTRIBUTE_WEAK
 typedef void (*fcgi_logger)(int type, const char *fmt, ...);
diff --git a/sapi/fpm/fpm/fpm_main.c b/sapi/fpm/fpm/fpm_main.c
index 3ab92b307042..695839cd9a7a 100644
--- a/sapi/fpm/fpm/fpm_main.c
+++ b/sapi/fpm/fpm/fpm_main.c
@@ -1996,7 +1996,8 @@ consult the installation file that came with this distribution, or visit \n\
 
 			requests++;
 			if (UNEXPECTED(max_requests && (requests == max_requests))) {
-				fcgi_finish_request(request, 1);
+				fcgi_request_set_keep(request, 0);
+				fcgi_finish_request(request, 0);
 				break;
 			}
 			/* end of fastcgi loop */
