diff --git a/core/src/main/java/org/elasticsearch/rest/action/cat/RestThreadPoolAction.java b/core/src/main/java/org/elasticsearch/rest/action/cat/RestThreadPoolAction.java
index 111e190638d1..76e5c40cf424 100644
--- a/core/src/main/java/org/elasticsearch/rest/action/cat/RestThreadPoolAction.java
+++ b/core/src/main/java/org/elasticsearch/rest/action/cat/RestThreadPoolAction.java
@@ -60,7 +60,7 @@ public RestThreadPoolAction(Settings settings, RestController controller) {
     @Override
     protected void documentation(StringBuilder sb) {
         sb.append("/_cat/thread_pool\n");
-        sb.append("/_cat/thread_pool/{thread_pools}");
+        sb.append("/_cat/thread_pool/{thread_pools}\n");
     }
 
     @Override
