diff --git a/Marlin/src/Marlin.cpp b/Marlin/src/Marlin.cpp
index 993dd26529f..d6285225c19 100644
--- a/Marlin/src/Marlin.cpp
+++ b/Marlin/src/Marlin.cpp
@@ -661,9 +661,14 @@ void idle(
 ) {
 
   #if ENABLED(SPI_ENDSTOPS)
-    if (endstops.tmc_spi_homing.any && ELAPSED(millis(), sg_guard_period))
+    if (endstops.tmc_spi_homing.any
+      #if ENABLED(IMPROVE_HOMING_RELIABILITY)
+        && ELAPSED(millis(), sg_guard_period)
+      #endif
+    ) {
       for (uint8_t i = 4; i--;) // Read SGT 4 times per idle loop
         if (endstops.tmc_spi_homing_check()) break;
+    }
   #endif
 
   #if ENABLED(MAX7219_DEBUG)
