diff --git a/libraries/SITL/SIM_Gazebo.h b/libraries/SITL/SIM_Gazebo.h
index 0adc21902f06..c542fa842cef 100644
--- a/libraries/SITL/SIM_Gazebo.h
+++ b/libraries/SITL/SIM_Gazebo.h
@@ -54,7 +54,7 @@ class Gazebo : public Aircraft
       double timestamp;
       double imu_angular_velocity_rpy[3];
       double imu_linear_acceleration_xyz[3];
-      double imu_orientation_quat[3];
+      double imu_orientation_quat[4];
       double velocity_xyz[3];
       double position_xyz[3];
     };
