diff --git a/numpy/distutils/system_info.py b/numpy/distutils/system_info.py
index bea120cf96c..93a8e6f7519 100644
--- a/numpy/distutils/system_info.py
+++ b/numpy/distutils/system_info.py
@@ -1876,7 +1876,7 @@ def check_embedded_lapack(self, info):
         c = customized_ccompiler()
 
         tmpdir = tempfile.mkdtemp()
-        s = """void zungqr();
+        s = """void zungqr_();
         int main(int argc, const char *argv[])
         {
             zungqr_();
