diff --git a/Marlin/src/HAL/shared/Delay.h b/Marlin/src/HAL/shared/Delay.h
index 9b5f1aa123d..04771a2e038 100644
--- a/Marlin/src/HAL/shared/Delay.h
+++ b/Marlin/src/HAL/shared/Delay.h
@@ -75,8 +75,8 @@
       }
       #undef MAXNOPS
     }
-    else
-      __delay_4cycles(x / 4);
+    else if ((x >>= 2))
+      __delay_4cycles(x);
   }
   #undef nop
 
@@ -114,8 +114,8 @@
 
       #undef MAXNOPS
     }
-    else
-      __delay_4cycles(x / 4);
+    else if ((x >>= 2))
+      __delay_4cycles(x);
   }
   #undef nop
 
