diff --git a/libraries/AP_HAL_VRBRAIN/Util.cpp b/libraries/AP_HAL_VRBRAIN/Util.cpp
index 309f14004d72..9d96a90c2f98 100644
--- a/libraries/AP_HAL_VRBRAIN/Util.cpp
+++ b/libraries/AP_HAL_VRBRAIN/Util.cpp
@@ -134,7 +134,7 @@ bool VRBRAINUtil::get_system_id(char buf[40])
 */
 uint32_t VRBRAINUtil::available_memory(void)
 {
-    return mallinfo(),fordblks;
+    return mallinfo().fordblks;
 }
 
 #endif // CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
