diff --git a/cmake/checks/cpu_avx2.cpp b/cmake/checks/cpu_avx2.cpp
index 3ab1143b8f2..121fe27ca2a 100644
--- a/cmake/checks/cpu_avx2.cpp
+++ b/cmake/checks/cpu_avx2.cpp
@@ -6,5 +6,6 @@ void test()
 {
     int data[8] = {0,0,0,0, 0,0,0,0};
     __m256i a = _mm256_loadu_si256((const __m256i *)data);
+    __m256i b = _mm256_bslli_epi128(a, 1);  // available in GCC 4.9.3+
 }
 int main() { return 0; }
