diff --git a/ext/phar/phar_object.c b/ext/phar/phar_object.c
index 31d443cd43d0..1799268297ad 100644
--- a/ext/phar/phar_object.c
+++ b/ext/phar/phar_object.c
@@ -5191,7 +5191,7 @@ ZEND_BEGIN_ARG_INFO_EX(arginfo_phar_webPhar, 0, 0, 0)
 ZEND_END_ARG_INFO()
 
 PHAR_ARG_INFO
-ZEND_BEGIN_ARG_INFO_EX(arginfo_phar_running, 0, 0, 1)
+ZEND_BEGIN_ARG_INFO_EX(arginfo_phar_running, 0, 0, 0)
 	ZEND_ARG_INFO(0, retphar)
 ZEND_END_ARG_INFO()
 
diff --git a/ext/phar/tests/bug74383.phpt b/ext/phar/tests/bug74383.phpt
new file mode 100644
index 000000000000..366c4fcb9b48
--- /dev/null
+++ b/ext/phar/tests/bug74383.phpt
@@ -0,0 +1,20 @@
+--TEST--
+Phar: bug #74383: Wrong reflection on Phar::running
+--SKIPIF--
+<?php if (!extension_loaded("phar") || !extension_loaded('reflection')) die("skip"); ?>
+--FILE--
+<?php
+$rc = new ReflectionClass(Phar::class);
+$rm = $rc->getMethod("running");
+echo $rm->getNumberOfParameters();
+echo PHP_EOL;
+echo $rm->getNumberOfRequiredParameters();
+echo PHP_EOL;
+echo (int) $rm->getParameters()[0]->isOptional();
+
+?>
+
+--EXPECT--
+1
+0
+1
\ No newline at end of file
