diff --git a/Marlin/src/pins/sam/pins_RAMPS_DUO.h b/Marlin/src/pins/sam/pins_RAMPS_DUO.h
index 4f9053b8cae..d63d9a32a81 100644
--- a/Marlin/src/pins/sam/pins_RAMPS_DUO.h
+++ b/Marlin/src/pins/sam/pins_RAMPS_DUO.h
@@ -50,7 +50,7 @@
 #define BOARD_INFO_NAME "RAMPS Duo"
 
 #define IS_RAMPS_DUO
-#include "pins_RAMPS.h"
+#include "../ramps/pins_RAMPS.h"
 
 //
 // Temperature Sensors
diff --git a/Marlin/src/pins/sam/pins_RAMPS_SMART.h b/Marlin/src/pins/sam/pins_RAMPS_SMART.h
index f1f593991f7..619b2d594da 100644
--- a/Marlin/src/pins/sam/pins_RAMPS_SMART.h
+++ b/Marlin/src/pins/sam/pins_RAMPS_SMART.h
@@ -67,7 +67,7 @@
 #define BOARD_INFO_NAME "RAMPS-SMART"
 
 #define IS_RAMPS_SMART
-#include "pins_RAMPS.h"
+#include "../ramps/pins_RAMPS.h"
 
 // I2C EEPROM with 4K of space
 #define I2C_EEPROM
