diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 48bf2922483..d572f52829c 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -1013,13 +1013,13 @@ void servo_init() {
       pixels.show(); // initialize to all off
 
       #if ENABLED(NEOPIXEL_STARTUP_TEST)
-        delay(2000);
+        safe_delay(2000);
         set_neopixel_color(pixels.Color(255, 0, 0, 0));  // red
-        delay(2000);
+        safe_delay(2000);
         set_neopixel_color(pixels.Color(0, 255, 0, 0));  // green
-        delay(2000);
+        safe_delay(2000);
         set_neopixel_color(pixels.Color(0, 0, 255, 0));  // blue
-        delay(2000);
+        safe_delay(2000);
       #endif
       set_neopixel_color(pixels.Color(NEO_WHITE));       // white
     }
