diff --git a/modules/imgcodecs/src/loadsave.cpp b/modules/imgcodecs/src/loadsave.cpp
index f5d9c98a3be..d4be6ae4fb1 100644
--- a/modules/imgcodecs/src/loadsave.cpp
+++ b/modules/imgcodecs/src/loadsave.cpp
@@ -396,6 +396,8 @@ static void ApplyExifOrientation(const Mat& buf, Mat& img)
 static void*
 imread_( const String& filename, int flags, int hdrtype, Mat* mat=0 )
 {
+    CV_Assert(mat || hdrtype != LOAD_MAT); // mat is required in LOAD_MAT case
+
     IplImage* image = 0;
     CvMat *matrix = 0;
     Mat temp, *data = &temp;
