diff --git a/numpy/core/tests/test_multiarray.py b/numpy/core/tests/test_multiarray.py
index 183a30b2968..2630f04db2f 100644
--- a/numpy/core/tests/test_multiarray.py
+++ b/numpy/core/tests/test_multiarray.py
@@ -2559,7 +2559,7 @@ def __numpy_ufunc__(self, ufunc, method, i, inputs, **kw):
 
     def test_out_override(self):
         # regression test for github bug 4753
-        class OutClass(ndarray):
+        class OutClass(np.ndarray):
             def __numpy_ufunc__(self, ufunc, method, i, inputs, **kw):
                 if 'out' in kw:
                     tmp_kw = kw.copy()
