diff --git a/numpy/distutils/system_info.py b/numpy/distutils/system_info.py
index 254d6a8d4dc..5cc87ee9884 100644
--- a/numpy/distutils/system_info.py
+++ b/numpy/distutils/system_info.py
@@ -1682,7 +1682,7 @@ def calc_info(self):
 class blas_info(system_info):
     section = 'blas'
     dir_env_var = 'BLAS'
-    _lib_names = ['blas']
+    _lib_names = ['blas', 'cblas']
     notfounderror = BlasNotFoundError
 
     def calc_info(self):
