diff --git a/libraries/SRV_Channel/SRV_Channel_aux.cpp b/libraries/SRV_Channel/SRV_Channel_aux.cpp
index c9b637f8e570..d3ba3e7475f3 100644
--- a/libraries/SRV_Channel/SRV_Channel_aux.cpp
+++ b/libraries/SRV_Channel/SRV_Channel_aux.cpp
@@ -32,6 +32,9 @@ void SRV_Channel::output_ch(void)
     // take care of special function cases
     switch(function)
     {
+    case k_none:
+        // disabled
+        return;
     case k_manual:              // manual
         passthrough_from = ch_num;
         break;
