diff --git a/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java b/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java
index 42a9344eabc1..2a7c185ad8aa 100644
--- a/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java
+++ b/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java
@@ -324,9 +324,11 @@ private void execute() {
             }
         } else {
             boolean success = false;
+            boolean acquired = false;
             try {
                 listener.beforeBulk(executionId, bulkRequest);
                 semaphore.acquire();
+                acquired = true;
                 client.bulk(bulkRequest, new ActionListener<BulkResponse>() {
                     @Override
                     public void onResponse(BulkResponse response) {
@@ -353,7 +355,7 @@ public void onFailure(Throwable e) {
             } catch (Throwable t) {
                 listener.afterBulk(executionId, bulkRequest, t);
             } finally {
-                 if (!success) {  // if we fail on client.bulk() release the semaphore
+                 if (!success && acquired) {  // if we fail on client.bulk() release the semaphore
                      semaphore.release();
                  }
             }
