diff --git a/Marlin/src/inc/Conditionals_LCD.h b/Marlin/src/inc/Conditionals_LCD.h
index fdfecadbab5..049935050e1 100644
--- a/Marlin/src/inc/Conditionals_LCD.h
+++ b/Marlin/src/inc/Conditionals_LCD.h
@@ -158,10 +158,11 @@
   #define ENCODER_STEPS_PER_MENU_ITEM 2
 #endif
 
-// Generic support for SSD1306 / SSD1309 / SH1106 OLED based LCDs.
-#if ENABLED(U8GLIB_SSD1306) || ENABLED(U8GLIB_SSD1309) || ENABLED(U8GLIB_SH1106)
-  #define ULTRA_LCD  //general LCD support, also 16x2
-  #define DOGLCD  // Support for I2C LCD 128x64 (Controller SSD1306 / SSD1309 / SH1106 graphic Display Family)
+// 128x64 I2C OLED LCDs - SSD1306/SSD1309/SH1106
+#define HAS_SSD1306_OLED_I2C (ENABLED(U8GLIB_SSD1306) || ENABLED(U8GLIB_SSD1309) || ENABLED(U8GLIB_SH1106))
+#if HAS_SSD1306_OLED_I2C
+  #define ULTRA_LCD
+  #define DOGLCD
 #endif
 
 #if ENABLED(PANEL_ONE) || ENABLED(U8GLIB_SH1106)
diff --git a/Marlin/src/pins/pins_DUE3DOM.h b/Marlin/src/pins/pins_DUE3DOM.h
index 22b58e351c9..14957a0ef8e 100644
--- a/Marlin/src/pins/pins_DUE3DOM.h
+++ b/Marlin/src/pins/pins_DUE3DOM.h
@@ -146,7 +146,7 @@
     #define SDSS             4
     #define SD_DETECT_PIN   14
 
-  #elif ENABLED(SSD1306_OLED_I2C_CONTROLLER)
+  #elif HAS_SSD1306_OLED_I2C
 
     #define BTN_EN1         50
     #define BTN_EN2         52
diff --git a/Marlin/src/pins/pins_DUE3DOM_MINI.h b/Marlin/src/pins/pins_DUE3DOM_MINI.h
index 703c3d417fa..314d5227473 100644
--- a/Marlin/src/pins/pins_DUE3DOM_MINI.h
+++ b/Marlin/src/pins/pins_DUE3DOM_MINI.h
@@ -138,7 +138,7 @@
     #define SDSS             4
     #define SD_DETECT_PIN   14
 
-  #elif ENABLED(SSD1306_OLED_I2C_CONTROLLER)
+  #elif HAS_SSD1306_OLED_I2C
 
     #define BTN_EN1         50
     #define BTN_EN2         52
diff --git a/Marlin/src/pins/pins_RADDS.h b/Marlin/src/pins/pins_RADDS.h
index 568b7b3f711..473d69dffe5 100644
--- a/Marlin/src/pins/pins_RADDS.h
+++ b/Marlin/src/pins/pins_RADDS.h
@@ -247,7 +247,7 @@
     #define SDSS            10
     #define SD_DETECT_PIN   14
 
-  #elif ENABLED(SSD1306_OLED_I2C_CONTROLLER)
+  #elif HAS_SSD1306_OLED_I2C
 
     #define BTN_EN1         50
     #define BTN_EN2         52
diff --git a/Marlin/src/pins/pins_RURAMPS4D_11.h b/Marlin/src/pins/pins_RURAMPS4D_11.h
index 1e622d507b9..fd50e7ef1d6 100644
--- a/Marlin/src/pins/pins_RURAMPS4D_11.h
+++ b/Marlin/src/pins/pins_RURAMPS4D_11.h
@@ -228,7 +228,7 @@
 
     #define SD_DETECT_PIN   51
 
-  #elif ENABLED(SSD1306_OLED_I2C_CONTROLLER)
+  #elif HAS_SSD1306_OLED_I2C
 
     #define BEEPER_PIN      62
     #define LCD_SDSS        10
diff --git a/Marlin/src/pins/pins_RURAMPS4D_13.h b/Marlin/src/pins/pins_RURAMPS4D_13.h
index 0d71e091162..5f23846db49 100644
--- a/Marlin/src/pins/pins_RURAMPS4D_13.h
+++ b/Marlin/src/pins/pins_RURAMPS4D_13.h
@@ -214,7 +214,7 @@
 
     #define SD_DETECT_PIN   51
 
-  #elif ENABLED(SSD1306_OLED_I2C_CONTROLLER)
+  #elif HAS_SSD1306_OLED_I2C
 
     #define BEEPER_PIN      62
     #define LCD_SDSS        10
