diff --git a/Marlin/temperature.cpp b/Marlin/temperature.cpp
index 9dcab02b0e1..9c6c9890bf1 100644
--- a/Marlin/temperature.cpp
+++ b/Marlin/temperature.cpp
@@ -1269,7 +1269,7 @@ void Temperature::init() {
 
 #if ENABLED(FAST_PWM_FAN)
 
-  void setPwmFrequency(const pin_t pin, int val) {
+  void Temperature::setPwmFrequency(const pin_t pin, int val) {
     val &= 0x07;
     switch (digitalPinToTimer(pin)) {
       #ifdef TCCR0A
diff --git a/Marlin/temperature.h b/Marlin/temperature.h
index b3d9e127e4e..c43776a5de8 100644
--- a/Marlin/temperature.h
+++ b/Marlin/temperature.h
@@ -588,7 +588,7 @@ class Temperature {
   private:
 
     #if ENABLED(FAST_PWM_FAN)
-      void setPwmFrequency(const pin_t pin, int val);
+      static void setPwmFrequency(const pin_t pin, int val);
     #endif
 
     static void set_current_temp_raw();
