diff --git a/core/src/main/java/org/elasticsearch/action/ActionWriteResponse.java b/core/src/main/java/org/elasticsearch/action/ActionWriteResponse.java
index f4152ac85e4a..6fa436fae018 100644
--- a/core/src/main/java/org/elasticsearch/action/ActionWriteResponse.java
+++ b/core/src/main/java/org/elasticsearch/action/ActionWriteResponse.java
@@ -35,6 +35,7 @@
 
 import java.io.IOException;
 import java.util.Collections;
+import java.util.Arrays;
 
 /**
  * Base class for write action responses.
@@ -161,7 +162,11 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
         @Override
         public String toString() {
-            return Strings.toString(this);
+            return "ShardInfo{" +
+                "total=" + total +
+                ", successful=" + successful +
+                ", failures=" + Arrays.toString(failures) +
+                '}';
         }
 
         public static ShardInfo readShardInfo(StreamInput in) throws IOException {
