diff --git a/modules/dnn/src/layers/normalize_bbox_layer.cpp b/modules/dnn/src/layers/normalize_bbox_layer.cpp
index 70d9b7385a2..fbb29292c2a 100644
--- a/modules/dnn/src/layers/normalize_bbox_layer.cpp
+++ b/modules/dnn/src/layers/normalize_bbox_layer.cpp
@@ -276,7 +276,7 @@ class NormalizeBBoxLayerImpl CV_FINAL : public NormalizeBBoxLayer
             {
                 auto weights = InferenceEngine::make_shared_blob<float>(InferenceEngine::Precision::FP32,
                                                                         InferenceEngine::Layout::C,
-                                                                        {numChannels});
+                                                                        {(size_t)numChannels});
                 weights->allocate();
                 std::vector<float> ones(numChannels, 1);
                 weights->set(ones);
@@ -286,7 +286,7 @@ class NormalizeBBoxLayerImpl CV_FINAL : public NormalizeBBoxLayer
             else
             {
                 CV_Assert(numChannels == blobs[0].total());
-                ieLayer->blobs["weights"] = wrapToInfEngineBlob(blobs[0], {numChannels}, InferenceEngine::Layout::C);
+                ieLayer->blobs["weights"] = wrapToInfEngineBlob(blobs[0], {(size_t)numChannels}, InferenceEngine::Layout::C);
                 ieLayer->params["channel_shared"] = blobs[0].total() == 1 ? "1" : "0";
             }
             ieLayer->params["eps"] = format("%f", epsilon);
