diff --git a/src/main/java/org/elasticsearch/action/search/SearchRequest.java b/src/main/java/org/elasticsearch/action/search/SearchRequest.java
index 9acc8a1d62fc..20efe76c79ba 100644
--- a/src/main/java/org/elasticsearch/action/search/SearchRequest.java
+++ b/src/main/java/org/elasticsearch/action/search/SearchRequest.java
@@ -422,9 +422,9 @@ public SearchRequest templateSource(BytesReference template, boolean unsafe) {
     /**
      * The template of the search request.
      */
-    public SearchRequest templateSource(String source) {
-        this.source = new BytesArray(source);
-        this.sourceUnsafe = false;
+    public SearchRequest templateSource(String template) {
+        this.templateSource = new BytesArray(template);
+        this.templateSourceUnsafe = false;
         return this;
     }
 
