diff --git a/tools/dockerfile/interoptest/grpc_interop_node/build_interop.sh b/tools/dockerfile/interoptest/grpc_interop_node/build_interop.sh
index c16efc1d354..aeb82e0c9c3 100755
--- a/tools/dockerfile/interoptest/grpc_interop_node/build_interop.sh
+++ b/tools/dockerfile/interoptest/grpc_interop_node/build_interop.sh
@@ -29,6 +29,4 @@ cp -r /var/local/jenkins/service_account $HOME || true
 cd /var/local/git/grpc-node
 
 # build Node interop client & server
-npm install -g node-gyp gulp
-npm install
-gulp setup
+./setup_interop.sh
diff --git a/tools/dockerfile/interoptest/grpc_interop_nodepurejs/build_interop.sh b/tools/dockerfile/interoptest/grpc_interop_nodepurejs/build_interop.sh
index d41ccacd2f9..db55f5a19f3 100755
--- a/tools/dockerfile/interoptest/grpc_interop_nodepurejs/build_interop.sh
+++ b/tools/dockerfile/interoptest/grpc_interop_nodepurejs/build_interop.sh
@@ -29,8 +29,4 @@ cp -r /var/local/jenkins/service_account $HOME || true
 cd /var/local/git/grpc-node
 
 # build Node interop client & server
-npm install -g gulp
-npm install
-gulp js.core.install
-gulp protobuf.install
-gulp internal.test.install
+./setup_interop_purejs.sh
