diff --git a/ext/zip/php_zip.c b/ext/zip/php_zip.c
index 52f058c70874..541a95d57fda 100644
--- a/ext/zip/php_zip.c
+++ b/ext/zip/php_zip.c
@@ -1095,7 +1095,6 @@ static void php_zip_object_free_storage(void *object TSRMLS_DC) /* {{{ */
 	if (intern->za) {
 		if (zip_close(intern->za) != 0) {
 			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Cannot destroy the zip context");
-			return;
 		}
 		intern->za = NULL;
 	}
diff --git a/ext/zip/tests/bug76524.phpt b/ext/zip/tests/bug76524.phpt
new file mode 100644
index 000000000000..258752067363
--- /dev/null
+++ b/ext/zip/tests/bug76524.phpt
@@ -0,0 +1,25 @@
+--TEST--
+ZipArchive Bug #76524 (memory leak with ZipArchive::OVERWRITE flag and empty archive)
+--SKIPIF--
+<?php
+/* $Id$ */
+if(!extension_loaded('zip')) die('skip');
+?>
+--FILE--
+<?php
+
+$filename = __DIR__ . '/nonexistent.zip';
+
+// The error is not reproduced when file already exist:
+if (!file_exists($filename)) {
+	$zip = new ZipArchive();
+	$zip->open($filename, ZipArchive::CREATE | ZipArchive::OVERWRITE);
+	echo 'ok';
+} else {
+	echo "file $filename exists, something goes wrong";
+}
+
+?>
+--EXPECTF--
+ok
+Warning: Unknown: Cannot destroy the zip context in Unknown on line 0
