diff --git a/Marlin/src/HAL/HAL_STM32/HAL.cpp b/Marlin/src/HAL/HAL_STM32/HAL.cpp
index 2cef2bd0e20..09c598e238f 100644
--- a/Marlin/src/HAL/HAL_STM32/HAL.cpp
+++ b/Marlin/src/HAL/HAL_STM32/HAL.cpp
@@ -29,6 +29,8 @@
 
 #include "HAL.h"
 
+#include "../../inc/MarlinConfig.h"
+
 #if ENABLED(EEPROM_EMULATED_WITH_SRAM)
   #if STM32F7xx
     #include "stm32f7xx_ll_pwr.h"
diff --git a/Marlin/src/HAL/HAL_STM32/HAL.h b/Marlin/src/HAL/HAL_STM32/HAL.h
index f43716871f6..aa2c98f2686 100644
--- a/Marlin/src/HAL/HAL_STM32/HAL.h
+++ b/Marlin/src/HAL/HAL_STM32/HAL.h
@@ -41,6 +41,7 @@
   #include <USBSerial.h>
 #endif
 
+#include "../../inc/MarlinConfigPre.h"
 #include "../shared/math_32bit.h"
 #include "../shared/HAL_SPI.h"
 #include "fastio_STM32.h"
diff --git a/Marlin/src/HAL/HAL_STM32/watchdog_STM32.h b/Marlin/src/HAL/HAL_STM32/watchdog_STM32.h
index ec0c53b4f01..e860d167933 100644
--- a/Marlin/src/HAL/HAL_STM32/watchdog_STM32.h
+++ b/Marlin/src/HAL/HAL_STM32/watchdog_STM32.h
@@ -21,7 +21,5 @@
  */
 #pragma once
 
-#include "../../inc/MarlinConfig.h"
-
 void watchdog_init();
 void watchdog_reset();
