diff --git a/INSTALL.rst.txt b/INSTALL.rst.txt
index 426105929ff..8b135e3090d 100644
--- a/INSTALL.rst.txt
+++ b/INSTALL.rst.txt
@@ -34,7 +34,7 @@ Building NumPy requires the following software installed:
    This is required for testing numpy, but not for using it.
 
 Python__ http://www.python.org
-nose__ http://somethingaboutorange.com/mrl/projects/nose/
+nose__ http://nose.readthedocs.io
 
 
 .. note:: 
diff --git a/doc/TESTS.rst.txt b/doc/TESTS.rst.txt
index 7028ca57a1a..68b0eace4a7 100644
--- a/doc/TESTS.rst.txt
+++ b/doc/TESTS.rst.txt
@@ -9,7 +9,7 @@ Introduction
 ''''''''''''
 
 SciPy uses the `Nose testing system
-<http://www.somethingaboutorange.com/mrl/projects/nose>`__, with some
+<http://nose.readthedocs.io>`__, with some
 minor convenience features added.  Nose is an extension of the unit
 testing framework offered by `unittest.py
 <http://docs.python.org/lib/module-unittest.html>`__. Our goal is that
diff --git a/numpy/lib/financial.py b/numpy/lib/financial.py
index 931b0af5658..95942da164c 100644
--- a/numpy/lib/financial.py
+++ b/numpy/lib/financial.py
@@ -210,9 +210,10 @@ def pmt(rate, nper, pv, fv=0, when='end'):
     (rate, nper, pv, fv, when) = map(np.array, [rate, nper, pv, fv, when])
     temp = (1 + rate)**nper
     mask = (rate == 0.0)
-    np.copyto(rate, 1.0, where=mask)
-    z = np.zeros(np.broadcast(rate, nper, pv, fv, when).shape)
-    fact = np.where(mask != z, nper + z, (1 + rate*when)*(temp - 1)/rate + z)
+    masked_rate = np.where(mask, 1.0, rate)
+    z = np.zeros(np.broadcast(masked_rate, nper, pv, fv, when).shape)
+    fact = np.where(mask != z, nper + z,
+                    (1 + masked_rate*when)*(temp - 1)/masked_rate + z)
     return -(fv + pv*temp) / fact
 
 def nper(rate, pmt, pv, fv=0, when='end'):
diff --git a/numpy/testing/utils.py b/numpy/testing/utils.py
index 683a4f0a61e..cb39a9b4a26 100644
--- a/numpy/testing/utils.py
+++ b/numpy/testing/utils.py
@@ -64,7 +64,7 @@ def import_nose():
 
     if not nose_is_good:
         msg = ('Need nose >= %d.%d.%d for tests - see '
-               'http://somethingaboutorange.com/mrl/projects/nose' %
+               'http://nose.readthedocs.io' %
                minimum_nose_version)
         raise ImportError(msg)
 
