diff --git a/distribution/build.gradle b/distribution/build.gradle
index b4c183d9bd1a..540796fc5a3a 100644
--- a/distribution/build.gradle
+++ b/distribution/build.gradle
@@ -196,8 +196,8 @@ configure(subprojects.findAll { ['zip', 'tar', 'integ-test-zip'].contains(it.nam
       into('bin') {
         with copySpec {
           with binFiles
-          from('../src/main/resources') {
-            include 'bin/*.bat'
+          from('../src/main/resources/bin') {
+            include '*.bat'
             filter(FixCrLfFilter, eol: FixCrLfFilter.CrLf.newInstance('crlf'))
           }
           MavenFilteringHack.filter(it, expansions)
