diff --git a/main/network.c b/main/network.c
index 8a326369a875..23f0f140fa9c 100644
--- a/main/network.c
+++ b/main/network.c
@@ -648,7 +648,7 @@ PHPAPI void php_network_populate_name_from_sockaddr(
 			case AF_INET6:
 				buf = (char*)inet_ntop(sa->sa_family, &((struct sockaddr_in6*)sa)->sin6_addr, (char *)&abuf, sizeof(abuf));
 				if (buf) {
-					*textaddr = strpprintf(0, "%s:%d",
+					*textaddr = strpprintf(0, "[%s]:%d",
 						buf, ntohs(((struct sockaddr_in6*)sa)->sin6_port));
 				}
 
diff --git a/tests/output/bug76136.phpt b/tests/output/bug76136.phpt
new file mode 100644
index 000000000000..49b6932351f6
--- /dev/null
+++ b/tests/output/bug76136.phpt
@@ -0,0 +1,12 @@
+--TEST--
+Bug #76136: stream_socket_get_name should enclose IPv6 in brackets
+--FILE--
+<?php
+$server = stream_socket_server("tcp://[::1]:1337/");
+echo stream_socket_get_name($server, false).PHP_EOL;
+$server = stream_socket_server("tcp://127.0.0.1:1337/");
+echo stream_socket_get_name($server, false);
+?>
+--EXPECT--
+[::1]:1337
+127.0.0.1:1337
