diff --git a/Marlin/Marlin_main.cpp b/Marlin/Marlin_main.cpp
index 2a8d954bc69..7ac74adb0eb 100644
--- a/Marlin/Marlin_main.cpp
+++ b/Marlin/Marlin_main.cpp
@@ -2797,6 +2797,8 @@ inline void gcode_G28() {
       enqueuecommands_P(PSTR(Z_PROBE_END_SCRIPT));
       st_synchronize();
     #endif
+
+    FlushSerialRequestResend();
   }
 
   #ifndef Z_PROBE_SLED
@@ -3360,6 +3362,8 @@ inline void gcode_M42() {
     SERIAL_PROTOCOLPGM("Standard Deviation: ");
     SERIAL_PROTOCOL_F(sigma, 6);
     SERIAL_EOL; SERIAL_EOL;
+
+    FlushSerialRequestResend();
   }
 
 #endif // ENABLE_AUTO_BED_LEVELING && Z_PROBE_REPEATABILITY_TEST
