diff --git a/src/main/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapper.java b/src/main/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapper.java
index 04e715f99428..2026579dc2fd 100644
--- a/src/main/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapper.java
+++ b/src/main/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapper.java
@@ -139,7 +139,7 @@ public FieldNamesFieldMapper(Settings indexSettings) {
     }
 
     public FieldNamesFieldMapper(String name, String indexName, float boost, FieldType fieldType, EnabledAttributeMapper enabledState, @Nullable Settings fieldDataSettings, Settings indexSettings) {
-        super(new Names(name, indexName, indexName, name), boost, fieldType, null, Lucene.KEYWORD_ANALYZER,
+        super(new Names(name, indexName, indexName, name), boost, fieldType, false, Lucene.KEYWORD_ANALYZER,
                 Lucene.KEYWORD_ANALYZER, null, null, fieldDataSettings, indexSettings);
         this.defaultFieldType = Defaults.FIELD_TYPE;
         this.pre13Index = Version.indexCreated(indexSettings).before(Version.V_1_3_0);
@@ -240,9 +240,6 @@ protected void parseCreateField(ParseContext context, List<Field> fields) throws
                     if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
                         document.add(new Field(names().indexName(), fieldName, fieldType));
                     }
-                    if (hasDocValues()) {
-                        document.add(new SortedSetDocValuesField(names().indexName(), new BytesRef(fieldName)));
-                    }
                 }
             }
         }
diff --git a/src/test/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapperTests.java b/src/test/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapperTests.java
index dfd8705612e3..fdf9d00938b3 100644
--- a/src/test/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapperTests.java
+++ b/src/test/java/org/elasticsearch/index/mapper/internal/FieldNamesFieldMapperTests.java
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.index.mapper.internal;
 
+import org.apache.lucene.index.IndexOptions;
 import org.elasticsearch.Version;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.settings.ImmutableSettings;
@@ -62,6 +63,20 @@ public void testExtractFieldNames() {
         assertEquals(set("", ".", ".."), extract(".."));
     }
 
+    public void testFieldType() throws Exception {
+        String mapping = XContentFactory.jsonBuilder().startObject().startObject("type")
+            .startObject("_field_names").field("store", "yes").endObject()
+            .endObject().endObject().string();
+
+        DocumentMapper docMapper = createIndex("test").mapperService().documentMapperParser().parse(mapping);
+        FieldNamesFieldMapper fieldNamesMapper = docMapper.rootMapper(FieldNamesFieldMapper.class);
+        assertFalse(fieldNamesMapper.hasDocValues());
+        assertEquals(IndexOptions.DOCS, fieldNamesMapper.fieldType().indexOptions());
+        assertFalse(fieldNamesMapper.fieldType().tokenized());
+        assertFalse(fieldNamesMapper.fieldType().stored());
+        assertTrue(fieldNamesMapper.fieldType().omitNorms());
+    }
+
     public void testInjectIntoDocDuringParsing() throws Exception {
         String mapping = XContentFactory.jsonBuilder().startObject().startObject("type").endObject().endObject().string();
         DocumentMapper defaultMapper = createIndex("test").mapperService().documentMapperParser().parse(mapping);
