diff --git a/samples/python/letter_recog.py b/samples/python/letter_recog.py
index d4987369bcc..b519379a8f8 100755
--- a/samples/python/letter_recog.py
+++ b/samples/python/letter_recog.py
@@ -41,7 +41,7 @@ class LetterStatModel(object):
     train_ratio = 0.5
 
     def load(self, fn):
-        self.model.load(fn)
+        self.model = self.model.load(fn)
     def save(self, fn):
         self.model.save(fn)
 
