diff --git a/Zend/zend_signal.h b/Zend/zend_signal.h
index c42a3991a53d..449f4f0a8a79 100644
--- a/Zend/zend_signal.h
+++ b/Zend/zend_signal.h
@@ -88,7 +88,7 @@ ZEND_API void zend_signal_handler_unblock(void);
 void zend_signal_activate(void);
 void zend_signal_deactivate(void);
 BEGIN_EXTERN_C()
-void zend_signal_startup(void);
+ZEND_API void zend_signal_startup(void);
 END_EXTERN_C()
 void zend_signal_init(void);
 
