diff --git a/Marlin/src/lcd/ultralcd.h b/Marlin/src/lcd/ultralcd.h
index 565ff2f7927..9c9084dd766 100644
--- a/Marlin/src/lcd/ultralcd.h
+++ b/Marlin/src/lcd/ultralcd.h
@@ -449,10 +449,16 @@ class MarlinUI {
     static void save_previous_screen();
     static void goto_previous_screen(
       #if ENABLED(TURBO_BACK_MENU_ITEM)
-        const bool is_back=false
+        const bool is_back
       #endif
     );
 
+    #if ENABLED(TURBO_BACK_MENU_ITEM)
+      // Various menu items require a "void (*)()" to point to
+      // this function so a default argument *won't* work
+      static inline void goto_previous_screen() { goto_previous_screen(false); }
+    #endif
+
     static void return_to_status();
     static inline bool on_status_screen() { return currentScreen == status_screen; }
     static inline void run_current_screen() { (*currentScreen)(); }
