diff --git a/libraries/AP_ServoRelayEvents/AP_ServoRelayEvents.cpp b/libraries/AP_ServoRelayEvents/AP_ServoRelayEvents.cpp
index 68d73ff8d6dd..56327e20ca06 100644
--- a/libraries/AP_ServoRelayEvents/AP_ServoRelayEvents.cpp
+++ b/libraries/AP_ServoRelayEvents/AP_ServoRelayEvents.cpp
@@ -106,10 +106,6 @@ void AP_ServoRelayEvents::update_events(void)
         return;
     }
 
-    if (channel > NUM_SERVO_CHANNELS || channel == 0) {
-        return;
-    }
-
     start_time_ms = AP_HAL::millis();
 
     switch (type) {
