diff --git a/ext/xml/compat.c b/ext/xml/compat.c
index 3fc203ee23b7..d39afacc0bc4 100644
--- a/ext/xml/compat.c
+++ b/ext/xml/compat.c
@@ -593,10 +593,12 @@ has been defined and none can be detected */
 	}
 #endif
 
+	if (parser->parser->lastError.level >= XML_ERR_WARNING) {
+		return 0;
+	}
+
 	error = xmlParseChunk(parser->parser, (char *) data, data_len, is_final);
-	if (!error) {
-		return 1;
-	} else if (parser->parser->lastError.level > XML_ERR_WARNING ){
+	if (error) {
 		return 0;
 	} else {
 		return 1;
diff --git a/ext/xml/tests/bug73135.phpt b/ext/xml/tests/bug73135.phpt
new file mode 100644
index 000000000000..5fb43e91dbfe
--- /dev/null
+++ b/ext/xml/tests/bug73135.phpt
@@ -0,0 +1,24 @@
+--TEST--
+Bug #73135 (xml_parse() segmentation fault)
+--CREDITS--
+edgarsandi - <edgar.r.sandi@gmail.com>
+--FILE--
+<?php
+    function start_elem($parser, $xml) {
+        xml_parse($parser, $xml);
+    }
+
+    $xml = <<<HERE
+    <a xmlns="ahihi">
+        <bar foo="ahihi"/>
+    </a>
+HERE;
+
+    $parser = xml_parser_create_ns();
+    xml_set_element_handler($parser, 'start_elem', 'ahihi');
+    xml_parse($parser, $xml);
+?>
+--EXPECTF--
+Warning: xml_parse(): Unable to call handler ahihi() in %s%ebug73135.php on line %d
+
+Warning: xml_parse(): Unable to call handler ahihi() in %s%ebug73135.php on line %d
\ No newline at end of file
