diff --git a/Marlin/src/Marlin.cpp b/Marlin/src/Marlin.cpp
index 6f9e527832d..18cbbaa2378 100644
--- a/Marlin/src/Marlin.cpp
+++ b/Marlin/src/Marlin.cpp
@@ -1053,6 +1053,10 @@ void setup() {
     fanmux_init();
   #endif
 
+  #if HAS_TRINAMIC && HAS_LCD_MENU
+    init_tmc_section();
+  #endif
+
   #if ENABLED(MIXING_EXTRUDER)
     mixer.init();
   #endif
diff --git a/Marlin/src/lcd/ultralcd.cpp b/Marlin/src/lcd/ultralcd.cpp
index 8dbec921e19..4035bd34cde 100644
--- a/Marlin/src/lcd/ultralcd.cpp
+++ b/Marlin/src/lcd/ultralcd.cpp
@@ -293,10 +293,6 @@ void MarlinUI::init() {
   #if HAS_ENCODER_ACTION
     encoderDiff = 0;
   #endif
-
-  #if HAS_TRINAMIC && HAS_LCD_MENU
-    init_tmc_section();
-  #endif
 }
 
 bool MarlinUI::get_blink() {
