diff --git a/src/core/lib/iomgr/tcp_posix.cc b/src/core/lib/iomgr/tcp_posix.cc
index 13ceffc6960..68cce8a4655 100644
--- a/src/core/lib/iomgr/tcp_posix.cc
+++ b/src/core/lib/iomgr/tcp_posix.cc
@@ -743,7 +743,7 @@ static void process_errors(grpc_tcp* tcp) {
     if (r == -1) {
       return;
     }
-    if ((msg.msg_flags & MSG_CTRUNC) == 1) {
+    if ((msg.msg_flags & MSG_CTRUNC) != 0) {
       gpr_log(GPR_ERROR, "Error message was truncated.");
     }
 
