diff --git a/test/core/iomgr/socket_utils_test.cc b/test/core/iomgr/socket_utils_test.cc
index a21f3fac62e..8c5b6f3db9f 100644
--- a/test/core/iomgr/socket_utils_test.cc
+++ b/test/core/iomgr/socket_utils_test.cc
@@ -24,6 +24,7 @@
 #include "src/core/lib/iomgr/socket_utils_posix.h"
 
 #include <errno.h>
+#include <netinet/in.h>
 #include <netinet/ip.h>
 #include <string.h>
 
diff --git a/test/cpp/interop/interop_test.cc b/test/cpp/interop/interop_test.cc
index ae155b65f95..8e45b877212 100644
--- a/test/cpp/interop/interop_test.cc
+++ b/test/cpp/interop/interop_test.cc
@@ -16,10 +16,6 @@
  *
  */
 
-#ifndef _POSIX_SOURCE
-#define _POSIX_SOURCE
-#endif
-
 #include <assert.h>
 #include <signal.h>
 #include <stdio.h>
diff --git a/test/cpp/naming/address_sorting_test.cc b/test/cpp/naming/address_sorting_test.cc
index fc6721d0ba8..37fd408cc7a 100644
--- a/test/cpp/naming/address_sorting_test.cc
+++ b/test/cpp/naming/address_sorting_test.cc
@@ -51,6 +51,7 @@
 
 #ifndef GPR_WINDOWS
 #include <arpa/inet.h>
+#include <netinet/in.h>
 #include <sys/socket.h>
 #endif
 
diff --git a/test/cpp/naming/resolver_component_tests_runner_invoker.cc b/test/cpp/naming/resolver_component_tests_runner_invoker.cc
index 68be00a67d6..d8f7100a284 100644
--- a/test/cpp/naming/resolver_component_tests_runner_invoker.cc
+++ b/test/cpp/naming/resolver_component_tests_runner_invoker.cc
@@ -29,6 +29,10 @@
 #include <thread>
 #include <vector>
 
+#ifdef __FreeBSD__
+#include <sys/wait.h>
+#endif
+
 #include "test/cpp/util/subprocess.h"
 #include "test/cpp/util/test_config.h"
 
diff --git a/test/cpp/qps/json_run_localhost.cc b/test/cpp/qps/json_run_localhost.cc
index 948c3088e67..42cc9e0ed76 100644
--- a/test/cpp/qps/json_run_localhost.cc
+++ b/test/cpp/qps/json_run_localhost.cc
@@ -24,6 +24,10 @@
 #include <sstream>
 #include <string>
 
+#ifdef __FreeBSD__
+#include <sys/wait.h>
+#endif
+
 #include <grpc/support/log.h>
 
 #include "src/core/lib/gpr/env.h"
