diff --git a/libraries/RC_Channel/RC_Channel.cpp b/libraries/RC_Channel/RC_Channel.cpp
index 80abef9e7146..2970772cc53e 100644
--- a/libraries/RC_Channel/RC_Channel.cpp
+++ b/libraries/RC_Channel/RC_Channel.cpp
@@ -554,7 +554,7 @@ bool RC_Channels::duplicate_options_exist()
             continue;
         }
         const uint16_t option = c->option.get();
-        if (option > sizeof(auxsw_option_counts)) {
+        if (option >= sizeof(auxsw_option_counts)) {
             continue;
         }
         auxsw_option_counts[option]++;
