diff --git a/Marlin/power.cpp b/Marlin/power.cpp
index df0579153a7..8c9bfcc7eba 100644
--- a/Marlin/power.cpp
+++ b/Marlin/power.cpp
@@ -69,7 +69,9 @@ bool Power::is_power_needed() {
   ) return true;
 
   HOTEND_LOOP() if (thermalManager.degTargetHotend(e) > 0) return true;
-  if (thermalManager.degTargetBed() > 0) return true;
+  #if HAS_HEATED_BED
+    if (thermalManager.degTargetBed() > 0) return true;
+  #endif
 
   return false;
 }
