diff --git a/Marlin/src/pins/pins_MKS_ROBIN_NANO.h b/Marlin/src/pins/pins_MKS_ROBIN_NANO.h
index 1112e17eebe..3acb7f35e5b 100755
--- a/Marlin/src/pins/pins_MKS_ROBIN_NANO.h
+++ b/Marlin/src/pins/pins_MKS_ROBIN_NANO.h
@@ -89,7 +89,7 @@
 #define HEATER_1_PIN       PB0   // HEATER2
 #define HEATER_BED_PIN     PA0   // HOT BED
 
-#define FAN_PIN            PA1   // FAN
+#define FAN_PIN            PB1   // FAN
 
 #define BTN_ENC            PC13  // Pin is not connected. Real pin is needed to enable encoder's push button functionality used by touch screen
 
