diff --git a/Marlin/buzzer.h b/Marlin/buzzer.h
index 2e436259e46..c87fe849d7a 100644
--- a/Marlin/buzzer.h
+++ b/Marlin/buzzer.h
@@ -24,8 +24,8 @@
 #define __BUZZER_H__
 
 #include "fastio.h"
-#include "watchdog.h"
 #include "circularqueue.h"
+#include "temperature.h"
 
 #define TONE_QUEUE_LENGTH 4
 
@@ -106,9 +106,7 @@ class Buzzer {
       while (buffer.isFull()) {
         delay(5);
         this->tick();
-        #if ENABLED(USE_WATCHDOG)
-          watchdog_reset();
-        #endif
+        thermalManager.manage_heater();
       }
       this->buffer.enqueue((tone_t) { duration, frequency });
     }
