diff --git a/numpy/lib/tests/test_type_check.py b/numpy/lib/tests/test_type_check.py
index ce8ef2f1591..9d5908793ae 100644
--- a/numpy/lib/tests/test_type_check.py
+++ b/numpy/lib/tests/test_type_check.py
@@ -442,5 +442,16 @@ def test_asfarray(self):
             asfarray, np.array([1, 2, 3]), dtype=np.array(1.0))
 
 
+class TestAsScalar(object):
+
+    def test_basic(self):
+        a = np.asscalar(np.array([1]))
+        assert_equal(a.__class__, int)
+        assert_raises(ValueError, np.asscalar, [1, 2])
+        b = np.asscalar(3.4)
+        assert_equal(b.__class__, float)
+
+
+
 if __name__ == "__main__":
     run_module_suite()
diff --git a/numpy/lib/type_check.py b/numpy/lib/type_check.py
index 1664e6ebb08..17866553c4e 100644
--- a/numpy/lib/type_check.py
+++ b/numpy/lib/type_check.py
@@ -486,7 +486,10 @@ def asscalar(a):
     24
 
     """
-    return a.item()
+    try:
+        return a.item()
+    except AttributeError:
+        return asarray(a).item()
 
 #-----------------------------------------------------------------------------
 
