diff --git a/core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java b/core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java
index 2e7d5dd3cc6b..59d6cd101c40 100644
--- a/core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java
+++ b/core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java
@@ -38,7 +38,7 @@
  * node. The default is <tt>4</tt></li>
  * <p/>
  * <li><tt>cluster.routing.allocation.node_concurrent_recoveries</tt> -
- * restricts the number of concurrent recovery operations on a single node. The
+ * restricts the number of total concurrent shards initializing on a single node. The
  * default is <tt>2</tt></li>
  * </ul>
  * <p/>
@@ -106,7 +106,7 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
     public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
         int currentRecoveries = 0;
         for (ShardRouting shard : node) {
-            if (shard.initializing() || shard.relocating()) {
+            if (shard.initializing()) {
                 currentRecoveries++;
             }
         }
