diff --git a/buildSrc/src/main/minimumRuntime/org/elasticsearch/gradle/LoggedExec.java b/buildSrc/src/main/minimumRuntime/org/elasticsearch/gradle/LoggedExec.java
index a3f875729320..233431ff3c49 100644
--- a/buildSrc/src/main/minimumRuntime/org/elasticsearch/gradle/LoggedExec.java
+++ b/buildSrc/src/main/minimumRuntime/org/elasticsearch/gradle/LoggedExec.java
@@ -70,7 +70,7 @@ public void setSpoolOutput(boolean spoolOutput) {
             };
         } else {
             out = new ByteArrayOutputStream();
-            outputLogger = logger -> logger.error(((ByteArrayOutputStream) getStandardOutput()).toString(StandardCharsets.UTF_8));
+            outputLogger = logger -> logger.error(((ByteArrayOutputStream) out).toString(StandardCharsets.UTF_8));
         }
         setStandardOutput(out);
         setErrorOutput(out);
