diff --git a/tools/swig/pyfragments.swg b/tools/swig/pyfragments.swg
index 901e6ed9dcb..97ca8cf97c3 100644
--- a/tools/swig/pyfragments.swg
+++ b/tools/swig/pyfragments.swg
@@ -22,7 +22,6 @@
   SWIGINTERN int
   SWIG_AsVal_dec(long)(PyObject * obj, long * val)
   {
-    PyArray_Descr * longDescr = PyArray_DescrNewFromType(NPY_LONG);
     if (PyInt_Check(obj)) {
       if (val) *val = PyInt_AsLong(obj);
       return SWIG_OK;
@@ -56,7 +55,9 @@
     }
 %#endif
     if (!PyArray_IsScalar(obj,Integer)) return SWIG_TypeError;
+    PyArray_Descr * longDescr = PyArray_DescrNewFromType(NPY_LONG);
     PyArray_CastScalarToCtype(obj, (void*)val, longDescr);
+    Py_DECREF(longDescr);
     return SWIG_OK;
   }
 }
@@ -74,7 +75,6 @@
   SWIGINTERN int
   SWIG_AsVal_dec(unsigned long)(PyObject *obj, unsigned long *val)
   {
-    PyArray_Descr * ulongDescr = PyArray_DescrNewFromType(NPY_ULONG);
     %#if PY_VERSION_HEX < 0x03000000
     if (PyInt_Check(obj)) 
     {
@@ -120,7 +120,9 @@
     }
 %#endif
     if (!PyArray_IsScalar(obj,Integer)) return SWIG_TypeError;
+    PyArray_Descr * ulongDescr = PyArray_DescrNewFromType(NPY_ULONG);
     PyArray_CastScalarToCtype(obj, (void*)val, ulongDescr);
+    Py_DECREF(ulongDescr);
     return SWIG_OK;
   }
 }
