diff --git a/numpy/distutils/fcompiler/gnu.py b/numpy/distutils/fcompiler/gnu.py
index 1de658afb2f..1cddf7e8368 100644
--- a/numpy/distutils/fcompiler/gnu.py
+++ b/numpy/distutils/fcompiler/gnu.py
@@ -235,7 +235,9 @@ def get_flags_arch(self):
         return []
 
     def runtime_library_dir_option(self, dir):
-        return '-Wl,-rpath="%s"' % dir
+        sep = ',' if sys.platform == 'darwin' else '='
+        return '-Wl,-rpath%s"%s"' % (sep, dir)
+
 
 class Gnu95FCompiler(GnuFCompiler):
     compiler_type = 'gnu95'
