diff --git a/tools/travis-before-install.sh b/tools/travis-before-install.sh
index 5e5278ce536..3203658d5ff 100755
--- a/tools/travis-before-install.sh
+++ b/tools/travis-before-install.sh
@@ -26,6 +26,6 @@ fi
 source venv/bin/activate
 python -V
 pip install --upgrade pip setuptools
-pip install nose pytz cython
+pip install nose pytz cython==0.26
 if [ -n "$USE_ASV" ]; then pip install asv; fi
 popd
