diff --git a/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_tables.sql b/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_tables.sql
index db40c6b90865..1f11c085ee06 100644
--- a/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_tables.sql
+++ b/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_tables.sql
@@ -9,7 +9,7 @@ CREATE TABLE mock (
   SELF_REFERENCING_COL_NAME VARCHAR,
   REF_GENERATION VARCHAR
 ) AS
-SELECT '', 'test1', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
+SELECT null, 'test1', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
 UNION ALL
-SELECT '', 'test2', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
+SELECT null, 'test2', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
 ;
diff --git a/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_types_of_table.sql b/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_types_of_table.sql
index db40c6b90865..1f11c085ee06 100644
--- a/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_types_of_table.sql
+++ b/x-pack/plugin/sql/qa/src/main/resources/setup_mock_metadata_get_types_of_table.sql
@@ -9,7 +9,7 @@ CREATE TABLE mock (
   SELF_REFERENCING_COL_NAME VARCHAR,
   REF_GENERATION VARCHAR
 ) AS
-SELECT '', 'test1', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
+SELECT null, 'test1', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
 UNION ALL
-SELECT '', 'test2', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
+SELECT null, 'test2', 'BASE TABLE', '', null, null, null, null, null FROM DUAL
 ;
diff --git a/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTables.java b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTables.java
index a3b8f1817415..190224c583c0 100644
--- a/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTables.java
+++ b/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTables.java
@@ -146,7 +146,7 @@ public final void execute(SqlSession session, ActionListener<Page> listener) {
                  .sorted(Comparator.<IndexInfo, String> comparing(i -> legacyName(i.type()))
                            .thenComparing(Comparator.comparing(i -> i.name())))
                  .map(t -> asList(cluster,
-                         EMPTY,
+                         null,
                          t.name(),
                          legacyName(t.type()),
                          EMPTY,
