diff --git a/numpy/core/setup_common.py b/numpy/core/setup_common.py
index e0cb3f6305d..2ea72bcd011 100644
--- a/numpy/core/setup_common.py
+++ b/numpy/core/setup_common.py
@@ -192,7 +192,7 @@ def check_long_double_representation(cmd):
     if sys.platform == "win32" and not mingw32():
         try:
             cmd.compiler.compile_options.remove("/GL")
-        except ValueError:
+        except (ValueError, AttributeError):
             pass
 
     # We need to use _compile because we need the object filename
