diff --git a/Marlin/src/libs/duration_t.h b/Marlin/src/libs/duration_t.h
index 5e9dd1c0059..9db5d2365f9 100644
--- a/Marlin/src/libs/duration_t.h
+++ b/Marlin/src/libs/duration_t.h
@@ -151,7 +151,7 @@ struct duration_t {
     if (with_days) {
       uint16_t d = this->day();
       sprintf_P(buffer, PSTR("%ud %02u:%02u"), d, h % 24, m);
-      return d >= 10 ? 8 : 7;
+      return d >= 10 ? 9 : 8;
     }
     else if (h < 100) {
       sprintf_P(buffer, PSTR("%02u:%02u"), h % 24, m);
