diff --git a/Marlin/src/module/probe.cpp b/Marlin/src/module/probe.cpp
index 4e7436a6341..de7f0948023 100644
--- a/Marlin/src/module/probe.cpp
+++ b/Marlin/src/module/probe.cpp
@@ -107,9 +107,7 @@ float zprobe_zoffset; // Initialized by settings.load()
 
   // Move to the magnet to unlock the probe
   void run_deploy_moves_script() {
-    #ifndef TOUCH_MI_DEPLOY_XPOS
-      #define TOUCH_MI_DEPLOY_XPOS X_MIN_POS
-    #elif TOUCH_MI_DEPLOY_XPOS > X_MAX_BED
+    #if TOUCH_MI_DEPLOY_XPOS > X_MAX_BED
       TemporaryGlobalEndstopsState unlock_x(false);
     #endif
 
@@ -127,7 +125,9 @@ float zprobe_zoffset; // Initialized by settings.load()
       ui.reset_status();
       ui.goto_screen(prev_screen);
     #else
-      do_blocking_move_to_x(TOUCH_MI_DEPLOY_XPOS);
+      #ifdef TOUCH_MI_DEPLOY_XPOS
+        do_blocking_move_to_x(TOUCH_MI_DEPLOY_XPOS);
+      #endif
     #endif
   }
 
