diff --git a/Marlin/src/module/thermistor/thermistor_501.h b/Marlin/src/module/thermistor/thermistor_501.h
index 512ac0d8d80..ac40d119b39 100644
--- a/Marlin/src/module/thermistor/thermistor_501.h
+++ b/Marlin/src/module/thermistor/thermistor_501.h
@@ -28,8 +28,8 @@ const short temptable_501[][2] PROGMEM = {
    {OV(  19), 280},
    {OV(  23), 270},
    {OV(  27), 260},
-   {OV(  32), 250},
-   {OV(  30), 240},
+   {OV(  31), 250},
+   {OV(  37), 240},
    {OV(  47), 230},
    {OV(  57), 220},
    {OV(  68), 210},
