diff --git a/modules/core/include/opencv2/core/cvstd.inl.hpp b/modules/core/include/opencv2/core/cvstd.inl.hpp
index f158c3b46f5..4a85bfd0627 100644
--- a/modules/core/include/opencv2/core/cvstd.inl.hpp
+++ b/modules/core/include/opencv2/core/cvstd.inl.hpp
@@ -73,12 +73,6 @@ template<typename _Tp> class DataType< std::complex<_Tp> >
     typedef Vec<channel_type, channels> vec_type;
 };
 
-static inline
-std::ostream& operator << (std::ostream& os, const String& str)
-{
-    return os << str.c_str();
-}
-
 static inline
 std::ostream& operator << (std::ostream& out, Ptr<Formatted> fmtd)
 {
