diff --git a/numpy/core/arrayprint.py b/numpy/core/arrayprint.py
index 9f5039c57df..f4d6eba89b5 100644
--- a/numpy/core/arrayprint.py
+++ b/numpy/core/arrayprint.py
@@ -1170,7 +1170,7 @@ def dtype_is_implied(dtype):
     array([1, 2, 3], dtype=np.int8)
     """
     dtype = np.dtype(dtype)
-    if _format_options['legacy'] and dtype.type == bool_:
+    if _format_options['legacy'] == '1.13' and dtype.type == bool_:
         return False
     return dtype.type in _typelessdata
 
