diff --git a/Marlin/temperature.cpp b/Marlin/temperature.cpp
index a646136d399..131be977602 100644
--- a/Marlin/temperature.cpp
+++ b/Marlin/temperature.cpp
@@ -655,7 +655,7 @@ void manage_heater() {
 
     #if ENABLED(TEMP_SENSOR_1_AS_REDUNDANT)
       if (fabs(current_temperature[0] - redundant_temperature) > MAX_REDUNDANT_TEMP_SENSOR_DIFF) {
-        _temp_error(0, PSTR(MSG_EXTRUDER_SWITCHED_OFF), PSTR(MSG_ERR_REDUNDANT_TEMP));
+        _temp_error(0, PSTR(MSG_REDUNDANCY), PSTR(MSG_ERR_REDUNDANT_TEMP));
       }
     #endif
 
