diff --git a/numpy/core/src/multiarray/convert_datatype.c b/numpy/core/src/multiarray/convert_datatype.c
index c59979e75b7..4f6b3955b64 100644
--- a/numpy/core/src/multiarray/convert_datatype.c
+++ b/numpy/core/src/multiarray/convert_datatype.c
@@ -422,9 +422,9 @@ PyArray_CanCastSafely(int fromtype, int totype)
         case NPY_BOOL:
         case NPY_DATETIME:
         case NPY_TIMEDELTA:
+        case NPY_VOID:
             return 0;
         case NPY_OBJECT:
-        case NPY_VOID:
             return 1;
     }
 
diff --git a/numpy/core/src/multiarray/scalartypes.c.src b/numpy/core/src/multiarray/scalartypes.c.src
index 715decbde06..ee9a91e6db2 100644
--- a/numpy/core/src/multiarray/scalartypes.c.src
+++ b/numpy/core/src/multiarray/scalartypes.c.src
@@ -4008,8 +4008,10 @@ initialize_casting_tables(void)
     memset(_npy_can_cast_safely_table, 0, sizeof(_npy_can_cast_safely_table));
 
     for (i = 0; i < NPY_NTYPES; ++i) {
-        /* Identity */
-        _npy_can_cast_safely_table[i][i] = 1;
+        /* Identity, OK except for VOID, where it depends on the fields */
+        if (i != NPY_VOID) {
+            _npy_can_cast_safely_table[i][i] = 1;
+        }
         if (i != NPY_DATETIME) {
             /*
              * Bool -> <Anything> except datetime (since
@@ -4017,14 +4019,12 @@ initialize_casting_tables(void)
              */
             _npy_can_cast_safely_table[NPY_BOOL][i] = 1;
         }
-        /* <Anything> -> Object */
+        /* <Anything> -> Object is possible */
         _npy_can_cast_safely_table[i][NPY_OBJECT] = 1;
-        /* <Anything> -> Void */
-        _npy_can_cast_safely_table[i][NPY_VOID] = 1;
+        /* <Anything> -> Void is not possible without further checks */
     }
 
     _npy_can_cast_safely_table[NPY_STRING][NPY_UNICODE] = 1;
-    _npy_can_cast_safely_table[NPY_BOOL][NPY_TIMEDELTA] = 1;
 
 #ifndef NPY_SIZEOF_BYTE
 #define NPY_SIZEOF_BYTE 1
