diff --git a/Marlin/cardreader.cpp b/Marlin/cardreader.cpp
index 8693424d308..b8a940f4ba4 100644
--- a/Marlin/cardreader.cpp
+++ b/Marlin/cardreader.cpp
@@ -56,8 +56,8 @@ void CardReader::lsDive(const char *prepend, SdFile parent, const char * const m
     // If the entry is a directory and the action is LS_SerialPrint
     if (DIR_IS_SUBDIR(&p) && lsAction != LS_Count && lsAction != LS_GetFilename) {
 
-      // Allocate enough stack space for the full path to a folder
-      int len = strlen(prepend) + FILENAME_LENGTH + 1;
+      // Allocate enough stack space for the full path to a folder, trailing slash, and nul
+      int len = strlen(prepend) + FILENAME_LENGTH + 1 + 1;
       char path[len];
 
       // Get the short name for the item, which we know is a folder
