diff --git a/Marlin/ultralcd_impl_DOGM.h b/Marlin/ultralcd_impl_DOGM.h
index ac6fedabba2..8c7c1ff717f 100644
--- a/Marlin/ultralcd_impl_DOGM.h
+++ b/Marlin/ultralcd_impl_DOGM.h
@@ -244,13 +244,13 @@ void lcd_printPGM(const char* str) { for (; char c = pgm_read_byte(str); ++str)
 
 void lcd_print_utf(const char* const str, const uint8_t maxLength=LCD_WIDTH) {
   char c;
-  for (uint8_t i = 0, n = maxLength; n && (c = str[i]); ++i)
+  for (uint8_t i = 0, n = maxLength - 1; n && (c = str[i]); ++i)
     n -= charset_mapper(c);
 }
 
 void lcd_printPGM_utf(const char* str, const uint8_t maxLength=LCD_WIDTH) {
   char c;
-  for (uint8_t i = 0, n = maxLength; n && (c = str[i]); ++i)
+  for (uint8_t i = 0, n = maxLength - 1; n && (c = str[i]); ++i)
     n -= charset_mapper(c);
 }
 
diff --git a/Marlin/ultralcd_impl_HD44780.h b/Marlin/ultralcd_impl_HD44780.h
index d650890d50e..750693a0d5d 100644
--- a/Marlin/ultralcd_impl_HD44780.h
+++ b/Marlin/ultralcd_impl_HD44780.h
@@ -387,13 +387,13 @@ void lcd_printPGM(const char* str) { for (; char c = pgm_read_byte(str); ++str)
 
 void lcd_print_utf(const char * const str, const uint8_t maxLength=LCD_WIDTH) {
   char c;
-  for (uint8_t i = 0, n = maxLength; n && (c = str[i]); ++i)
+  for (uint8_t i = 0, n = maxLength - 1; n && (c = str[i]); ++i)
     n -= charset_mapper(c);
 }
 
 void lcd_printPGM_utf(const char* str, const uint8_t maxLength=LCD_WIDTH) {
   char c;
-  for (uint8_t i = 0, n = maxLength; n && (c = str[i]); ++i)
+  for (uint8_t i = 0, n = maxLength - 1; n && (c = str[i]); ++i)
     n -= charset_mapper(c);
 }
 
