diff --git a/libraries/AP_HAL/board/linux.h b/libraries/AP_HAL/board/linux.h
index 0999dfacb92c..963c495feec1 100644
--- a/libraries/AP_HAL/board/linux.h
+++ b/libraries/AP_HAL/board/linux.h
@@ -146,11 +146,6 @@
 /* HMC5843 can be an external compass */
 #define HAL_COMPASS_HMC5843_I2C_BUS 1
 #define HAL_COMPASS_HMC5843_I2C_ADDR 0x1E
-#define HAL_GPIO_A_LED_PIN        16
-#define HAL_GPIO_B_LED_PIN        16
-#define HAL_GPIO_C_LED_PIN        16
-#define HAL_GPIO_LED_ON           LOW
-#define HAL_GPIO_LED_OFF          HIGH
 #elif CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_ERLEBRAIN2
 #define HAL_INS_DEFAULT HAL_INS_MPU9250_SPI
 #define HAL_INS_DEFAULT_ROTATION ROTATION_YAW_270
@@ -204,7 +199,6 @@
 #elif CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_BLUE
 #define HAL_GPIO_A_LED_PIN 66
 #define HAL_GPIO_B_LED_PIN 67
-#define HAL_GPIO_C_LED_PIN 67
 #define HAL_GPIO_LED_ON    HIGH
 #define HAL_GPIO_LED_OFF   LOW
 #define HAL_INS_DEFAULT HAL_INS_MPU9250_I2C
diff --git a/libraries/AP_Notify/AP_BoardLED.cpp b/libraries/AP_Notify/AP_BoardLED.cpp
index 51f970cbdc4d..8e137c43e077 100644
--- a/libraries/AP_Notify/AP_BoardLED.cpp
+++ b/libraries/AP_Notify/AP_BoardLED.cpp
@@ -16,18 +16,12 @@
 
 #include "AP_Notify.h"
 
-#if (defined(HAL_GPIO_A_LED_PIN) || defined(HAL_GPIO_B_LED_PIN) || \
+#if (defined(HAL_GPIO_A_LED_PIN) && defined(HAL_GPIO_B_LED_PIN) && \
      defined(HAL_GPIO_C_LED_PIN))
 
-#ifndef HAL_GPIO_A_LED_PIN
-#define HAL_GPIO_A_LED_PIN        -1
-#endif
-#ifndef HAL_GPIO_B_LED_PIN
-#define HAL_GPIO_B_LED_PIN        -1
-#endif
-#ifndef HAL_GPIO_C_LED_PIN
-#define HAL_GPIO_C_LED_PIN        -1
-#endif
+static_assert((HAL_GPIO_A_LED_PIN != HAL_GPIO_B_LED_PIN) &&
+              (HAL_GPIO_A_LED_PIN != HAL_GPIO_C_LED_PIN) &&
+              (HAL_GPIO_B_LED_PIN != HAL_GPIO_C_LED_PIN), "Duplicate LED assignments detected");
 
 extern const AP_HAL::HAL& hal;
 
diff --git a/libraries/AP_Notify/AP_BoardLED2.cpp b/libraries/AP_Notify/AP_BoardLED2.cpp
index 333f36469ba1..c375c6d1ec59 100644
--- a/libraries/AP_Notify/AP_BoardLED2.cpp
+++ b/libraries/AP_Notify/AP_BoardLED2.cpp
@@ -19,14 +19,9 @@
 
 // show all status on only 2 leds
 
-#if defined(HAL_GPIO_A_LED_PIN) || defined(HAL_GPIO_B_LED_PIN)
+#if defined(HAL_GPIO_A_LED_PIN) && defined(HAL_GPIO_B_LED_PIN)
 
-#ifndef HAL_GPIO_A_LED_PIN
-#define HAL_GPIO_A_LED_PIN        -1
-#endif
-#ifndef HAL_GPIO_B_LED_PIN
-#define HAL_GPIO_B_LED_PIN        -1
-#endif
+static_assert((HAL_GPIO_A_LED_PIN != HAL_GPIO_B_LED_PIN), "Duplicate LED assignments detected");
 
 extern const AP_HAL::HAL& hal;
 
