diff --git a/src/main/java/io/reactivex/subscribers/SafeSubscriber.java b/src/main/java/io/reactivex/subscribers/SafeSubscriber.java
index bf1ab938c9..540e2e7a0c 100644
--- a/src/main/java/io/reactivex/subscribers/SafeSubscriber.java
+++ b/src/main/java/io/reactivex/subscribers/SafeSubscriber.java
@@ -154,8 +154,11 @@ public void onError(Throwable t) {
             }
         } catch (Throwable e) {
             Exceptions.throwIfFatal(e);
-            t2.suppress(e);
-            
+            if (t2 == null) {
+                t2 = new CompositeException(t, e);
+            } else {
+                t2.suppress(e);
+            }
             RxJavaPlugins.onError(t2);
         }
     }
