diff --git a/src/cpp/common/channel_arguments.cc b/src/cpp/common/channel_arguments.cc
index 50ee9d871f0..ff56538384f 100644
--- a/src/cpp/common/channel_arguments.cc
+++ b/src/cpp/common/channel_arguments.cc
@@ -106,6 +106,8 @@ void ChannelArguments::SetSocketMutator(grpc_socket_mutator* mutator) {
   }
 
   if (!replaced) {
+    strings_.push_back(mutator_arg.key);
+    mutator_arg.key = const_cast<char*>(strings_.back().c_str());
     args_.push_back(mutator_arg);
   }
 }
diff --git a/test/cpp/common/channel_arguments_test.cc b/test/cpp/common/channel_arguments_test.cc
index 183d2afa783..00df5835b26 100644
--- a/test/cpp/common/channel_arguments_test.cc
+++ b/test/cpp/common/channel_arguments_test.cc
@@ -213,6 +213,11 @@ TEST_F(ChannelArgumentsTest, SetSocketMutator) {
   EXPECT_TRUE(HasArg(arg1));
   // arg0 is replaced by arg1
   EXPECT_FALSE(HasArg(arg0));
+
+  // Test channel arg copying.
+  ChannelArguments channel_args_copy(channel_args_);
+  EXPECT_TRUE(HasArg(arg1));
+  EXPECT_FALSE(HasArg(arg0));
 }
 
 TEST_F(ChannelArgumentsTest, SetUserAgentPrefix) {
