diff --git a/ext/date/lib/timezonemap.h b/ext/date/lib/timezonemap.h
index 4601923ac991..a8a2da27896d 100644
--- a/ext/date/lib/timezonemap.h
+++ b/ext/date/lib/timezonemap.h
@@ -412,7 +412,7 @@
 	{ "cest",  1,   7200, "Europe/Copenhagen"             },
 	{ "cest",  1,   7200, "Europe/Gibraltar"              },
 	{ "cest",  1,   7200, "Europe/Kaliningrad"            },
-	{ "cest",  1,   7200, "Europe/Kiev"                   },
+	{ "cest",  1,   7200, "Europe/Kyiv"                   },
 	{ "cest",  1,   7200, "Europe/Lisbon"                 },
 	{ "cest",  1,   7200, "Europe/Ljubljana"              },
 	{ "cest",  1,   7200, "Europe/Luxembourg"             },
@@ -442,7 +442,7 @@
 	{ "cest",  1,   7200, "Europe/Vilnius"                },
 	{ "cest",  1,   7200, "Europe/Warsaw"                 },
 	{ "cest",  1,   7200, "Europe/Zagreb"                 },
-	{ "cest",  1,   7200, "Europe/Zaporozhye"             },
+	{ "cest",  1,   7200, "Europe/Zaporizhia"             },
 	{ "cest",  1,   7200, "Europe/Zurich"                 },
 	{ "cest",  1,   7200, "WET"                           },
 	{ "cet",   0,   3600, "Europe/Berlin"                 },
@@ -465,7 +465,7 @@
 	{ "cet",   0,   3600, "Europe/Copenhagen"             },
 	{ "cet",   0,   3600, "Europe/Gibraltar"              },
 	{ "cet",   0,   3600, "Europe/Kaliningrad"            },
-	{ "cet",   0,   3600, "Europe/Kiev"                   },
+	{ "cet",   0,   3600, "Europe/Kyiv"                   },
 	{ "cet",   0,   3600, "Europe/Lisbon"                 },
 	{ "cet",   0,   3600, "Europe/Ljubljana"              },
 	{ "cet",   0,   3600, "Europe/Luxembourg"             },
@@ -495,7 +495,7 @@
 	{ "cet",   0,   3600, "Europe/Vilnius"                },
 	{ "cet",   0,   3600, "Europe/Warsaw"                 },
 	{ "cet",   0,   3600, "Europe/Zagreb"                 },
-	{ "cet",   0,   3600, "Europe/Zaporozhye"             },
+	{ "cet",   0,   3600, "Europe/Zaporizhia"             },
 	{ "cet",   0,   3600, "Europe/Zurich"                 },
 	{ "cet",   0,   3600, "WET"                           },
 	{ "cet",   0,   7200, "Europe/Kaliningrad"            },
@@ -656,7 +656,7 @@
 	{ "cst",   1,  37800, "Australia/North"               },
 	{ "cst",   1,  37800, "Australia/South"               },
 	{ "cst",   1,  37800, "Australia/Yancowinna"          },
-	{ "cut",   0,   8400, "Europe/Zaporozhye"             },
+	{ "cut",   0,   8400, "Europe/Zaporizhia"             },
 	{ "cvst",  1,  -3600, "Atlantic/Cape_Verde"           },
 	{ "cvt",   0,  -3600, "Atlantic/Cape_Verde"           },
 	{ "cvt",   0,  -7200, "Atlantic/Cape_Verde"           },
@@ -764,7 +764,7 @@
 	{ "eest",  1,  10800, "Europe/Chisinau"               },
 	{ "eest",  1,  10800, "Europe/Istanbul"               },
 	{ "eest",  1,  10800, "Europe/Kaliningrad"            },
-	{ "eest",  1,  10800, "Europe/Kiev"                   },
+	{ "eest",  1,  10800, "Europe/Kyiv"                   },
 	{ "eest",  1,  10800, "Europe/Mariehamn"              },
 	{ "eest",  1,  10800, "Europe/Minsk"                  },
 	{ "eest",  1,  10800, "Europe/Moscow"                 },
@@ -777,7 +777,7 @@
 	{ "eest",  1,  10800, "Europe/Uzhgorod"               },
 	{ "eest",  1,  10800, "Europe/Vilnius"                },
 	{ "eest",  1,  10800, "Europe/Warsaw"                 },
-	{ "eest",  1,  10800, "Europe/Zaporozhye"             },
+	{ "eest",  1,  10800, "Europe/Zaporizhia"             },
 	{ "eet",   0,   7200, "Europe/Helsinki"               },
 	{ "eet",   1,  10800, "Asia/Gaza"                     },
 	{ "eet",   0,   7200, "Africa/Cairo"                  },
@@ -794,7 +794,7 @@
 	{ "eet",   0,   7200, "Europe/Chisinau"               },
 	{ "eet",   0,   7200, "Europe/Istanbul"               },
 	{ "eet",   0,   7200, "Europe/Kaliningrad"            },
-	{ "eet",   0,   7200, "Europe/Kiev"                   },
+	{ "eet",   0,   7200, "Europe/Kyiv"                   },
 	{ "eet",   0,   7200, "Europe/Mariehamn"              },
 	{ "eet",   0,   7200, "Europe/Minsk"                  },
 	{ "eet",   0,   7200, "Europe/Moscow"                 },
@@ -807,7 +807,7 @@
 	{ "eet",   0,   7200, "Europe/Uzhgorod"               },
 	{ "eet",   0,   7200, "Europe/Vilnius"                },
 	{ "eet",   0,   7200, "Europe/Warsaw"                 },
-	{ "eet",   0,   7200, "Europe/Zaporozhye"             },
+	{ "eet",   0,   7200, "Europe/Zaporizhia"             },
 	{ "egst",  1,      0, "America/Scoresbysund"          },
 	{ "egt",   0,  -3600, "America/Scoresbysund"          },
 	{ "ehdt",  1, -16200, "America/Santo_Domingo"         },
@@ -1090,7 +1090,7 @@
 	{ "kmt",   0, -18432, "America/Cayman"                },
 	{ "kmt",   0, -18432, "America/Grand_Turk"            },
 	{ "kmt",   0, -18432, "America/Jamaica"               },
-	{ "kmt",   0,   7324, "Europe/Kiev"                   },
+	{ "kmt",   0,   7324, "Europe/Kyiv"                   },
 	{ "kost",  0,  39600, "Pacific/Kosrae"                },
 	{ "kost",  0,  43200, "Pacific/Kosrae"                },
 	{ "krast", 1,  25200, "Asia/Krasnoyarsk"              },
@@ -1423,7 +1423,7 @@
 	{ "msd",   1,  18000, "Europe/Moscow"                 },
 	{ "msd",   1,  14400, "Europe/Chisinau"               },
 	{ "msd",   1,  14400, "Europe/Kaliningrad"            },
-	{ "msd",   1,  14400, "Europe/Kiev"                   },
+	{ "msd",   1,  14400, "Europe/Kyiv"                   },
 	{ "msd",   1,  14400, "Europe/Minsk"                  },
 	{ "msd",   1,  14400, "Europe/Riga"                   },
 	{ "msd",   1,  14400, "Europe/Simferopol"             },
@@ -1431,11 +1431,11 @@
 	{ "msd",   1,  14400, "Europe/Tiraspol"               },
 	{ "msd",   1,  14400, "Europe/Uzhgorod"               },
 	{ "msd",   1,  14400, "Europe/Vilnius"                },
-	{ "msd",   1,  14400, "Europe/Zaporozhye"             },
+	{ "msd",   1,  14400, "Europe/Zaporizhia"             },
 	{ "msk",   0,  10800, "Europe/Moscow"                 },
 	{ "msk",   0,  10800, "Europe/Chisinau"               },
 	{ "msk",   0,  10800, "Europe/Kaliningrad"            },
-	{ "msk",   0,  10800, "Europe/Kiev"                   },
+	{ "msk",   0,  10800, "Europe/Kyiv"                   },
 	{ "msk",   0,  10800, "Europe/Minsk"                  },
 	{ "msk",   0,  10800, "Europe/Riga"                   },
 	{ "msk",   0,  10800, "Europe/Simferopol"             },
@@ -1443,7 +1443,7 @@
 	{ "msk",   0,  10800, "Europe/Tiraspol"               },
 	{ "msk",   0,  10800, "Europe/Uzhgorod"               },
 	{ "msk",   0,  10800, "Europe/Vilnius"                },
-	{ "msk",   0,  10800, "Europe/Zaporozhye"             },
+	{ "msk",   0,  10800, "Europe/Zaporizhia"             },
 	{ "mst",   0, -25200, "America/Denver"                },
 	{ "mst",   0, -25200, "America/Bahia_Banderas"        },
 	{ "mst",   0, -25200, "America/Boise"                 },
diff --git a/ext/date/tests/bug52577.phpt b/ext/date/tests/bug52577.phpt
index 7a58f2a21acd..c7e9f7f9f949 100644
--- a/ext/date/tests/bug52577.phpt
+++ b/ext/date/tests/bug52577.phpt
@@ -1,8 +1,10 @@
 --TEST--
 Bug #52577 (Incorrect date returning)
+--CREDITS--
+marcosptf - <marcosptf@yahoo.com.br>
 --FILE--
 <?php
-date_default_timezone_set('Europe/Kiev');
+date_default_timezone_set('Europe/Kyiv');
 $date = '7.8.2010';
 echo "String: ".$date."\n";
 $date_format = DATE_RFC2822;
diff --git a/ext/date/tests/bug70840.phpt b/ext/date/tests/bug70840.phpt
new file mode 100644
index 000000000000..65d497a7ff42
--- /dev/null
+++ b/ext/date/tests/bug70840.phpt
@@ -0,0 +1,21 @@
+--TEST--
+Bug #52577 (Incorrect date returning)
+--CREDITS--
+marcosptf - <marcosptf@yahoo.com.br>
+--FILE--
+<?php
+date_default_timezone_set('Europe/Zaporizhia');
+$date = '05.11.2015';
+echo "String: ".$date."\n";
+$date_format = DATE_RFC2822;
+$unixtime = strtotime($date);
+echo "Unixtime: ".$unixtime."\n";
+echo "Date(PHP): ".date($date_format,$unixtime)."\n";
+$date = new DateTime('@'.$unixtime);
+echo "DateTime(PHP Class): ".$date->format($date_format);
+?>
+--EXPECT--
+String: 05.11.2015
+Unixtime: 1446674400
+Date(PHP): Thu, 05 Nov 2015 00:00:00 +0200
+DateTime(PHP Class): Wed, 04 Nov 2015 22:00:00 +0000
