diff --git a/modules/ml/src/ann_mlp.cpp b/modules/ml/src/ann_mlp.cpp
index 7323ab57a70..69e44481692 100644
--- a/modules/ml/src/ann_mlp.cpp
+++ b/modules/ml/src/ann_mlp.cpp
@@ -1535,6 +1535,10 @@ void CvANN_MLP::read( CvFileStorage* fs, CvFileNode* node )
 
     _layer_sizes = (CvMat*)cvReadByName( fs, node, "layer_sizes" );
     CV_CALL( create( _layer_sizes, SIGMOID_SYM, 0, 0 ));
+
+    cvReleaseMat( &_layer_sizes );
+    _layer_sizes = NULL;
+
     l_count = layer_sizes->cols;
 
     CV_CALL( read_params( fs, node ));
