diff --git a/test/cpp/qps/smoke_test.cc b/test/cpp/qps/smoke_test.cc
index e3907308325..2c60a9997c8 100644
--- a/test/cpp/qps/smoke_test.cc
+++ b/test/cpp/qps/smoke_test.cc
@@ -33,6 +33,8 @@
 
 #include <grpc/support/log.h>
 
+#include <signal.h>
+
 #include "test/cpp/qps/driver.h"
 #include "test/cpp/qps/report.h"
 
@@ -136,6 +138,7 @@ static void RunQPS() {
 }  // namespace grpc
 
 int main(int argc, char** argv) {
+  signal(SIGPIPE, SIG_IGN);
   using namespace grpc::testing;
   RunSynchronousStreamingPingPong();
   RunSynchronousUnaryPingPong();
