diff --git a/modules/python/package/cv2/__init__.py b/modules/python/package/cv2/__init__.py
index b176c0d9546..9427365dd24 100644
--- a/modules/python/package/cv2/__init__.py
+++ b/modules/python/package/cv2/__init__.py
@@ -65,7 +65,7 @@ def load_first_config(fnames, required=True):
     if DEBUG: print('OpenCV loader: BINARIES_PATHS={}'.format(str(l_vars['BINARIES_PATHS'])))
 
     for p in reversed(l_vars['PYTHON_EXTENSIONS_PATHS']):
-        sys.path.insert(0, p)
+        sys.path.insert(1, p)
 
     if os.name == 'nt':
         os.environ['PATH'] = ';'.join(l_vars['BINARIES_PATHS']) + ';' + os.environ.get('PATH', '')
