diff --git a/Marlin/src/pins/pins_AZTEEG_X5_MINI.h b/Marlin/src/pins/pins_AZTEEG_X5_MINI.h
index c73f62fa990..91349c0f6ed 100644
--- a/Marlin/src/pins/pins_AZTEEG_X5_MINI.h
+++ b/Marlin/src/pins/pins_AZTEEG_X5_MINI.h
@@ -29,7 +29,7 @@
 #endif
 
 #ifndef BOARD_NAME
-  #define BOARD_NAME        "Azteeg X5 MINI WIFI"
+  #define BOARD_NAME        "Azteeg X5 MINI"
 #endif
 #define BOARD_WEBSITE_URL "http://www.panucatt.com/azteeg_X5_mini_reprap_3d_printer_controller_p/ax5mini.htm"
 
