diff --git a/src/core/lib/surface/server.cc b/src/core/lib/surface/server.cc
index c0fae0f140d..5fa58ffdec2 100644
--- a/src/core/lib/surface/server.cc
+++ b/src/core/lib/surface/server.cc
@@ -1500,5 +1500,8 @@ int grpc_server_has_open_connections(grpc_server* server) {
 
 grpc_core::channelz::ServerNode* grpc_server_get_channelz_node(
     grpc_server* server) {
+  if (server == nullptr) {
+    return nullptr;
+  }
   return server->channelz_server.get();
 }
