diff --git a/Marlin/language_es.h b/Marlin/language_es.h
index 5f0bb03d757..fc54601f139 100644
--- a/Marlin/language_es.h
+++ b/Marlin/language_es.h
@@ -39,7 +39,7 @@
 #define MSG_MAIN                            _UxGT("Menu principal")
 #define MSG_AUTOSTART                       _UxGT("Inicio automatico")
 #define MSG_DISABLE_STEPPERS                _UxGT("Apagar motores")
-#define MSG_DEBUG_MENU                      _UxGT("Menu depurar”)
+#define MSG_DEBUG_MENU                      _UxGT("Menu depurar")
 #define MSG_PROGRESS_BAR_TEST               _UxGT("Prueba barra avance")
 #define MSG_AUTO_HOME                       _UxGT("Llevar al origen")
 #define MSG_AUTO_HOME_X                     _UxGT("Origen X")
@@ -47,7 +47,7 @@
 #define MSG_AUTO_HOME_Z                     _UxGT("Origen Z")
 #define MSG_LEVEL_BED_HOMING                _UxGT("Origen XYZ")
 #define MSG_LEVEL_BED_WAITING               _UxGT("Iniciar (Presione)")
-#define MSG_LEVEL_BED_NEXT_POINT            _UxGT("Siguiente punto”)
+#define MSG_LEVEL_BED_NEXT_POINT            _UxGT("Siguiente punto")
 #define MSG_LEVEL_BED_DONE                  _UxGT("Nivelacion lista!")
 #define MSG_LEVEL_BED_CANCEL                _UxGT("Cancelar")
 #define MSG_SET_HOME_OFFSETS                _UxGT("Ajustar desfases")
@@ -70,7 +70,7 @@
 #define MSG_RETRACT                         _UxGT("Retraer")
 #define MSG_MOVE_AXIS                       _UxGT("Mover ejes")
 #define MSG_LEVEL_BED                       _UxGT("Nivelar plataforma")
-#define MSG_MOVING                          _UxGT("Moviendo...”)
+#define MSG_MOVING                          _UxGT("Moviendo...")
 #define MSG_FREE_XY                         _UxGT("Libre XY")
 #define MSG_MOVE_X                          _UxGT("Mover X")
 #define MSG_MOVE_Y                          _UxGT("Mover Y")
@@ -96,7 +96,7 @@
 #define MSG_PID_I                           _UxGT("PID-I")
 #define MSG_PID_D                           _UxGT("PID-D")
 #define MSG_PID_C                           _UxGT("PID-C")
-#define MSG_SELECT                          _UxGT("Seleccionar”)
+#define MSG_SELECT                          _UxGT("Seleccionar")
 #define MSG_ACC                             _UxGT("Aceleracion")
 #define MSG_VX_JERK                         _UxGT("Vx-jerk")
 #define MSG_VY_JERK                         _UxGT("Vy-jerk")
@@ -153,7 +153,7 @@
 #define MSG_INIT_SDCARD                     _UxGT("Iniciando tarjeta")
 #define MSG_CNG_SDCARD                      _UxGT("Cambiar tarjeta")
 #define MSG_ZPROBE_OUT                      _UxGT("Sonda Z fuera")
-#define MSG_BLTOUCH_SELFTEST                _UxGT("BLTouch Auto-Prueba”)
+#define MSG_BLTOUCH_SELFTEST                _UxGT("BLTouch Auto-Prueba")
 #define MSG_BLTOUCH_RESET                   _UxGT("Reiniciar BLTouch")
 #define MSG_HOME                            _UxGT("Home")  // Used as MSG_HOME " " MSG_X MSG_Y MSG_Z " " MSG_FIRST
 #define MSG_FIRST                           _UxGT("first")
@@ -169,8 +169,8 @@
 #define MSG_ERR_MINTEMP                     _UxGT("Error: Temp Minima")
 #define MSG_ERR_MAXTEMP_BED                 _UxGT("Error: Temp Max Plat")
 #define MSG_ERR_MINTEMP_BED                 _UxGT("Error: Temp Min Plat")
-#define MSG_ERR_Z_HOMING                    _UxGT("G28 Z Prohibido”)
-#define MSG_HALTED                          _UxGT("IMPRESORA PARADA”)
+#define MSG_ERR_Z_HOMING                    _UxGT("G28 Z Prohibido")
+#define MSG_HALTED                          _UxGT("IMPRESORA PARADA")
 #define MSG_PLEASE_RESET                    _UxGT("Por favor, reinicie")
 #define MSG_SHORT_DAY                       _UxGT("d") // One character only
 #define MSG_SHORT_HOUR                      _UxGT("h") // One character only
@@ -193,7 +193,7 @@
 #define MSG_INFO_BAUDRATE                   _UxGT("Baudios")
 #define MSG_INFO_PROTOCOL                   _UxGT("Protocolo")
 #define MSG_LIGHTS_ON                       _UxGT("Luz cabina ON")
-#define MSG_LIGHTS_OFF                      _UxGT("Luz cabina OFF”)
+#define MSG_LIGHTS_OFF                      _UxGT("Luz cabina OFF")
 #if LCD_WIDTH > 19
   #define MSG_INFO_PRINT_COUNT              _UxGT("Conteo de impresion")
   #define MSG_INFO_COMPLETED_PRINTS         _UxGT("Completadas")
