diff --git a/core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java b/core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java
index bd721adaf127..89be2ecabeb2 100644
--- a/core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java
+++ b/core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java
@@ -316,8 +316,8 @@ public final ShardSearchTransportRequest buildShardSearchRequest(SearchShardIter
 
     @Override
     protected void skipShard(SearchShardIterator iterator) {
-        super.skipShard(iterator);
         successfulOps.incrementAndGet();
         skippedOps.incrementAndGet();
+        super.skipShard(iterator);
     }
 }
