diff --git a/ArduPlane/system.cpp b/ArduPlane/system.cpp
index 8313962f8bae..b49b3776648c 100644
--- a/ArduPlane/system.cpp
+++ b/ArduPlane/system.cpp
@@ -171,7 +171,7 @@ void Plane::init_ardupilot()
     airspeed.init();
 
     if (g.compass_enabled==true) {
-        if (!compass.init() || !compass.read()) {
+        if (!compass.init() || (!g.hil_mode && !compass.read())) {
             cliSerial->println_P(PSTR("Compass initialisation failed!"));
             g.compass_enabled = false;
         } else {
diff --git a/libraries/AP_InertialSensor/AP_InertialSensor.cpp b/libraries/AP_InertialSensor/AP_InertialSensor.cpp
index 3e3420f30dcc..0d45f79c7547 100644
--- a/libraries/AP_InertialSensor/AP_InertialSensor.cpp
+++ b/libraries/AP_InertialSensor/AP_InertialSensor.cpp
@@ -734,6 +734,10 @@ bool AP_InertialSensor::calibrate_trim(float &trim_roll, float &trim_pitch)
  */
 bool AP_InertialSensor::accel_calibrated_ok_all() const
 {
+    // calibration is not applicable for HIL mode
+    if (_hil_mode)
+        return true;
+
     // check each accelerometer has offsets saved
     for (uint8_t i=0; i<get_accel_count(); i++) {
         // exactly 0.0 offset is extremely unlikely
