diff --git a/tools/run_tests/run_tests.py b/tools/run_tests/run_tests.py
index bd5959cd000..b1cde270448 100755
--- a/tools/run_tests/run_tests.py
+++ b/tools/run_tests/run_tests.py
@@ -346,6 +346,7 @@ def runs_per_test_type(arg_str):
     try:
         n = int(arg_str)
         if n <= 0: raise ValueError
+        return n
     except:
         msg = "'{}' isn't a positive integer or 'inf'".format(arg_str)
         raise argparse.ArgumentTypeError(msg)
