diff --git a/bin/elasticsearch.bat b/bin/elasticsearch.bat
index b6da883d6fab..f6e7ba8c0aa4 100644
--- a/bin/elasticsearch.bat
+++ b/bin/elasticsearch.bat
@@ -3,7 +3,7 @@
 SETLOCAL
 TITLE Elasticsearch ${project.version}
 
-CALL %~dp0elasticsearch.in.bat
+CALL "%~dp0elasticsearch.in.bat"
 
 "%JAVA_HOME%\bin\java" %JAVA_OPTS% %ES_JAVA_OPTS% %ES_PARAMS% %* -cp "%ES_CLASSPATH%" "org.elasticsearch.bootstrap.Elasticsearch"
 
diff --git a/bin/service.bat b/bin/service.bat
index d4040be9b478..553a87931c15 100644
--- a/bin/service.bat
+++ b/bin/service.bat
@@ -130,7 +130,7 @@ call:convertxm %ES_MAX_MEM% JVM_XMX
 REM java_opts might be empty - init to avoid tripping commons daemon (if the command starts with ;)
 if "%JAVA_OPTS%" == "" set JAVA_OPTS=-XX:+UseParNewGC
 
-CALL %ES_HOME%\bin\elasticsearch.in.bat
+CALL "%ES_HOME%\bin\elasticsearch.in.bat"
 
 rem thread stack size
 set JVM_SS=256
@@ -201,4 +201,4 @@ set /a conv=%conv% * 1024
 set "%~2=%conv%"
 goto:eof
 
-ENDLOCAL
\ No newline at end of file
+ENDLOCAL
