diff --git a/Marlin/src/pins/pins_RADDS.h b/Marlin/src/pins/pins_RADDS.h
index f1434434082..8f140debc38 100644
--- a/Marlin/src/pins/pins_RADDS.h
+++ b/Marlin/src/pins/pins_RADDS.h
@@ -193,7 +193,6 @@
 //
 // Misc. Functions
 //
-#define SDSS                4
 #define SD_DETECT_PIN      14
 #define PS_ON_PIN          40   // SERVO3_PIN
 
@@ -227,7 +226,6 @@
 
     #define BTN_BACK        71
 
-    #undef SDSS
     #define SDSS            10
     #define SD_DETECT_PIN   14
 
@@ -246,6 +244,9 @@
     #define BTN_EN2         52
     #define BTN_ENC         48
 
+    #define SDSS            10
+    #define SD_DETECT_PIN   14
+
   #elif ENABLED(SSD1306_OLED_I2C_CONTROLLER)
 
     #define BTN_EN1         50
@@ -268,3 +269,7 @@
   #endif // SPARK_FULL_GRAPHICS
 
 #endif // ULTRA_LCD
+
+#ifndef SDSS
+  #define SDSS              4
+#endif
