diff --git a/modules/videoio/src/cap_images.cpp b/modules/videoio/src/cap_images.cpp
index 8e23793e8c9..92de9247a8f 100644
--- a/modules/videoio/src/cap_images.cpp
+++ b/modules/videoio/src/cap_images.cpp
@@ -437,6 +437,7 @@ Ptr<IVideoWriter> create_Images_writer(const std::string &filename, int, double,
     {
         if( writer->open( filename.c_str() ))
             return makePtr<LegacyWriter>(writer);
+        delete writer;
     }
     catch (...)
     {
diff --git a/platforms/scripts/valgrind.supp b/platforms/scripts/valgrind.supp
index 7664c632c1b..321ec9e60bd 100644
--- a/platforms/scripts/valgrind.supp
+++ b/platforms/scripts/valgrind.supp
@@ -226,3 +226,12 @@
    fun:_ZN20CvVideoWriter_FFMPEG10writeFrameEPKhiiiii
    fun:cvWriteFrame_FFMPEG
 }
+
+{
+   OpenCV-test-gapi-thread-tls
+   Memcheck:Leak
+   match-leak-kinds: possible
+   fun:calloc
+   fun:allocate_dtv
+   fun:_dl_allocate_tls
+}
