diff --git a/ext/mbstring/php_unicode.c b/ext/mbstring/php_unicode.c
index 09ba877bb9d7..3ea50d1beac9 100644
--- a/ext/mbstring/php_unicode.c
+++ b/ext/mbstring/php_unicode.c
@@ -186,7 +186,7 @@ MBSTRING_API unsigned long php_unicode_toupper(unsigned long code, enum mbfl_no_
 		 */
 		field = 1;
 		l = _uccase_len[0] + _uccase_len[1];
-		r = _uccase_size - 3;
+		r = (_uccase_size * 3) - 3;
 	}
 	return case_lookup(code, l, r, field);
 }
@@ -217,7 +217,7 @@ MBSTRING_API unsigned long php_unicode_tolower(unsigned long code, enum mbfl_no_
 		 */
 		field = 2;
 		l = _uccase_len[0] + _uccase_len[1];
-		r = _uccase_size - 3;
+		r = (_uccase_size * 3) - 3;
 	}
 	return case_lookup(code, l, r, field);
 }
diff --git a/ext/mbstring/tests/bug69267.phpt b/ext/mbstring/tests/bug69267.phpt
new file mode 100644
index 000000000000..b6827a6731b3
--- /dev/null
+++ b/ext/mbstring/tests/bug69267.phpt
@@ -0,0 +1,13 @@
+--TEST--
+Bug #69267 (mb_strtolower fails on titlecase characters)
+--SKIPIF--
+<?php extension_loaded('mbstring') or die('skip mbstring not available'); ?>
+--FILE--
+<?php
+$str = "\u{01c5}";
+var_dump(
+    "\u{01c6}" === mb_strtolower($str)
+);
+?>
+--EXPECT--
+bool(true)
\ No newline at end of file
