diff --git a/numpy/conftest.py b/numpy/conftest.py
index 4d4d055ec20..7834dd39dff 100644
--- a/numpy/conftest.py
+++ b/numpy/conftest.py
@@ -13,6 +13,13 @@
 _collect_results = {}
 
 
+def pytest_configure(config):
+    config.addinivalue_line("markers",
+        "valgrind_error: Tests that are known to error under valgrind.")
+    config.addinivalue_line("markers",
+        "slow: Tests that are very slow.")
+
+
 #FIXME when yield tests are gone.
 @pytest.hookimpl()
 def pytest_itemcollected(item):
diff --git a/pytest.ini b/pytest.ini
index 1ddfb60095c..4748e3575a2 100644
--- a/pytest.ini
+++ b/pytest.ini
@@ -24,7 +24,3 @@ filterwarnings =
 
 env =
     PYTHONHASHSEED=0
-
-markers =
-    valgrind_error: Known to cause errors under valgrind
-    slow: Tests which are slow
