diff --git a/src/main/java/org/elasticsearch/rest/action/admin/indices/settings/RestUpdateSettingsAction.java b/src/main/java/org/elasticsearch/rest/action/admin/indices/settings/RestUpdateSettingsAction.java
index e880f2af0cf7..d19b5c9a1f42 100644
--- a/src/main/java/org/elasticsearch/rest/action/admin/indices/settings/RestUpdateSettingsAction.java
+++ b/src/main/java/org/elasticsearch/rest/action/admin/indices/settings/RestUpdateSettingsAction.java
@@ -33,12 +33,18 @@
 import java.util.Map;
 
 import static org.elasticsearch.client.Requests.updateSettingsRequest;
+import com.google.common.collect.ImmutableSet;
 
 /**
  *
  */
 public class RestUpdateSettingsAction extends BaseRestHandler {
 
+    private static final ImmutableSet<String> VALUES_TO_EXCLUDE = ImmutableSet.<String>builder()
+            .add("pretty").add("timeout").add("master_timeout").add("index")
+            .add("expand_wildcards").add("ignore_unavailable").add("allow_no_indices")
+            .build();
+
     @Inject
     public RestUpdateSettingsAction(Settings settings, RestController controller, Client client) {
         super(settings, controller, client);
@@ -69,7 +75,7 @@ public void handleRequest(final RestRequest request, final RestChannel channel,
             }
         }
         for (Map.Entry<String, String> entry : request.params().entrySet()) {
-            if (entry.getKey().equals("pretty") || entry.getKey().equals("timeout") || entry.getKey().equals("master_timeout") || entry.getKey().equals("index")) {
+            if (VALUES_TO_EXCLUDE.contains(entry.getKey())) {
                 continue;
             }
             updateSettings.put(entry.getKey(), entry.getValue());
