diff --git a/ext/date/lib/interval.c b/ext/date/lib/interval.c
index dee9f92eedb1..8763ef0fcc1a 100644
--- a/ext/date/lib/interval.c
+++ b/ext/date/lib/interval.c
@@ -72,7 +72,6 @@ timelib_rel_time *timelib_diff(timelib_time *one, timelib_time *two)
 	}
 
 	rt->days = fabs(floor((one->sse - two->sse - (dst_h_corr * 3600) - (dst_m_corr * 60)) / 86400));
-
 	timelib_do_rel_normalize(rt->invert ? one : two, rt);
 
 	/* We need to do this after normalisation otherwise we can't get "24H" */
@@ -84,6 +83,10 @@ timelib_rel_time *timelib_diff(timelib_time *one, timelib_time *two)
 			rt->h += dst_h_corr;
 			rt->i += dst_m_corr;
 		}
+		// because of adding DST correlation there could be negative values which need to be normalized.
+		if (rt->d < 0 || rt->h < 0 || rt->i < 0) {
+			timelib_do_rel_normalize(rt->invert ? one : two, rt);
+		}
 	}
 
 	/* Restore old TZ info */
diff --git a/ext/date/tests/bug74524.phpt b/ext/date/tests/bug74524.phpt
new file mode 100644
index 000000000000..a72c523d98c8
--- /dev/null
+++ b/ext/date/tests/bug74524.phpt
@@ -0,0 +1,13 @@
+--TEST--
+Bug #74524 Date diff is bad calculated, in same time zone
+--FILE--
+<?php
+date_default_timezone_set('Europe/Amsterdam');
+$a = new DateTime("2017-11-17 22:05:26.000000");
+$b = new DateTime("2017-04-03 22:29:15.079459");
+
+$diff = $a->diff($b);
+echo sprintf("%s %s %s %s", $diff->y, $diff->m, $diff->d, $diff->h);
+?>
+--EXPECT--
+0 7 13 23
