diff --git a/Zend/zend_ast.h b/Zend/zend_ast.h
index 2231b0a32ca0..818d77779a7c 100644
--- a/Zend/zend_ast.h
+++ b/Zend/zend_ast.h
@@ -144,7 +144,7 @@ enum _zend_ast_kind {
 
 	/* 4 child nodes */
 	ZEND_AST_FOR = 4 << ZEND_AST_NUM_CHILDREN_SHIFT,
-	ZEND_AST_FOREACH,
+	ZEND_AST_FOREACH
 };
 
 typedef uint16_t zend_ast_kind;
diff --git a/Zend/zend_types.h b/Zend/zend_types.h
index f4c22ccdd5c9..7bfbe440b2dc 100644
--- a/Zend/zend_types.h
+++ b/Zend/zend_types.h
@@ -48,7 +48,7 @@ typedef unsigned char zend_uchar;
 
 typedef enum {
   SUCCESS =  0,
-  FAILURE = -1,		/* this MUST stay a negative number, or it may affect functions! */
+  FAILURE = -1		/* this MUST stay a negative number, or it may affect functions! */
 } ZEND_RESULT_CODE;
 
 #ifdef ZEND_ENABLE_ZVAL_LONG64
diff --git a/ext/intl/breakiterator/breakiterator_iterators.h b/ext/intl/breakiterator/breakiterator_iterators.h
index 71a1e35025ae..98d81d8a2449 100644
--- a/ext/intl/breakiterator/breakiterator_iterators.h
+++ b/ext/intl/breakiterator/breakiterator_iterators.h
@@ -26,7 +26,7 @@ U_CDECL_END
 typedef enum {
 	PARTS_ITERATOR_KEY_SEQUENTIAL,
 	PARTS_ITERATOR_KEY_LEFT,
-	PARTS_ITERATOR_KEY_RIGHT,
+	PARTS_ITERATOR_KEY_RIGHT
 } parts_iter_key_type;
 
 #ifdef __cplusplus
diff --git a/ext/mysqlnd/mysqlnd_enum_n_def.h b/ext/mysqlnd/mysqlnd_enum_n_def.h
index 0a3022058d1d..52ea9b1efcba 100644
--- a/ext/mysqlnd/mysqlnd_enum_n_def.h
+++ b/ext/mysqlnd/mysqlnd_enum_n_def.h
@@ -207,7 +207,7 @@ typedef enum mysqlnd_parse_exec_response_type
 	MYSQLND_PARSE_EXEC_RESPONSE_IMPLICIT = 0,
 	MYSQLND_PARSE_EXEC_RESPONSE_IMPLICIT_NEXT_RESULT,
 	MYSQLND_PARSE_EXEC_RESPONSE_IMPLICIT_OUT_VARIABLES,
-	MYSQLND_PARSE_EXEC_RESPONSE_EXPLICIT,
+	MYSQLND_PARSE_EXEC_RESPONSE_EXPLICIT
 } enum_mysqlnd_parse_exec_response_type;
 
 typedef enum mysqlnd_client_option
@@ -666,7 +666,7 @@ enum php_mysqlnd_server_command
 	/* Here follow own, non-protocol, commands */
 	COM_REAP_RESULT=240,	/* own command */
 	COM_ENABLE_SSL,			/* own command */
-	COM_HANDSHAKE,			/* own command */
+	COM_HANDSHAKE			/* own command */
 };
 
 
diff --git a/ext/readline/readline_cli.c b/ext/readline/readline_cli.c
index 66cb34c900e4..8ef4c3d7e75a 100644
--- a/ext/readline/readline_cli.c
+++ b/ext/readline/readline_cli.c
@@ -135,7 +135,7 @@ typedef enum {
 	comment_block,
 	heredoc_start,
 	heredoc,
-	outside,
+	outside
 } php_code_type;
 
 static zend_string *cli_get_prompt(char *block, char prompt) /* {{{ */
diff --git a/sapi/phpdbg/phpdbg_watch.h b/sapi/phpdbg/phpdbg_watch.h
index 4d5e2153ae91..f087853c8d85 100644
--- a/sapi/phpdbg/phpdbg_watch.h
+++ b/sapi/phpdbg/phpdbg_watch.h
@@ -47,7 +47,7 @@ typedef enum {
 	WATCH_ON_REFCOUNTED,
 	WATCH_ON_STR,
 	WATCH_ON_HASHDATA,
-	WATCH_ON_BUCKET,
+	WATCH_ON_BUCKET
 } phpdbg_watchtype;
 
 
diff --git a/win32/ioutil.h b/win32/ioutil.h
index f96719ec1961..1d6e151a19b8 100644
--- a/win32/ioutil.h
+++ b/win32/ioutil.h
@@ -90,7 +90,7 @@ typedef enum {
 typedef enum {
 	PHP_WIN32_IOUTIL_NORM_OK,
 	PHP_WIN32_IOUTIL_NORM_PARTIAL,
-	PHP_WIN32_IOUTIL_NORM_FAIL,
+	PHP_WIN32_IOUTIL_NORM_FAIL
 } php_win32_ioutil_normalization_result;
 
 #define PHP_WIN32_IOUTIL_FW_SLASHW L'/'
