diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java
index 617fc6998a62..044caee41c55 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java
@@ -97,13 +97,6 @@ public void delete(BlobContainer blobContainer, String name) throws IOException
         blobContainer.deleteBlob(blobName(name));
     }
 
-    /**
-     * Checks obj in the blob container
-     */
-    public boolean exists(BlobContainer blobContainer, String name) {
-        return blobContainer.blobExists(blobName(name));
-    }
-
     public String blobName(String name) {
         return String.format(Locale.ROOT, blobNameFormat, name);
     }
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
index 39a4011783a5..0ef1d3ab149f 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -395,9 +395,6 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
             if (repositoryData.getAllSnapshotIds().stream().anyMatch(s -> s.getName().equals(snapshotName))) {
                 throw new InvalidSnapshotNameException(metadata.name(), snapshotId.getName(), "snapshot with the same name already exists");
             }
-            if (snapshotFormat.exists(blobContainer(), snapshotId.getUUID())) {
-                throw new InvalidSnapshotNameException(metadata.name(), snapshotId.getName(), "snapshot with the same name already exists");
-            }
 
             // Write Global MetaData
             globalMetaDataFormat.write(clusterMetaData, blobContainer(), snapshotId.getUUID());
