diff --git a/modules/videoio/src/cap_gstreamer.cpp b/modules/videoio/src/cap_gstreamer.cpp
index 7b3236f1e96..79515d5df63 100644
--- a/modules/videoio/src/cap_gstreamer.cpp
+++ b/modules/videoio/src/cap_gstreamer.cpp
@@ -1769,7 +1769,7 @@ CvResult CV_API_CALL cv_capture_open(const char* filename, int camera_index, CV_
         cap = new GStreamerCapture();
         bool res;
         if (filename)
-            res = cap->open(string(filename));
+            res = cap->open(std::string(filename));
         else
             res = cap->open(camera_index);
         if (res)
