diff --git a/Marlin/src/feature/pause.cpp b/Marlin/src/feature/pause.cpp
index 28f2a9fd003..131acdd7c94 100644
--- a/Marlin/src/feature/pause.cpp
+++ b/Marlin/src/feature/pause.cpp
@@ -683,7 +683,9 @@ void resume_print(const float &slow_load_length/*=0*/, const float &fast_load_le
 
   #if HAS_DISPLAY
     ui.reset_status();
-    ui.return_to_status();
+    #if HAS_LCD_MENU
+      ui.return_to_status();
+    #endif
   #endif
 }
 
diff --git a/Marlin/src/lcd/ultralcd.cpp b/Marlin/src/lcd/ultralcd.cpp
index 6dba900cb7e..3d2c25f267b 100644
--- a/Marlin/src/lcd/ultralcd.cpp
+++ b/Marlin/src/lcd/ultralcd.cpp
@@ -1424,6 +1424,10 @@ void MarlinUI::update() {
     set_status_P(msg, -1);
   }
 
+  #if ENABLED(SDSUPPORT)
+    extern bool wait_for_user, wait_for_heatup;
+  #endif
+
   void MarlinUI::abort_print() {
     #if ENABLED(SDSUPPORT)
       wait_for_heatup = wait_for_user = false;
@@ -1442,6 +1446,10 @@ void MarlinUI::update() {
     #endif
   }
 
+  #if ANY(PARK_HEAD_ON_PAUSE, SDSUPPORT)
+    #include "../gcode/queue.h"
+  #endif
+
   void MarlinUI::pause_print() {
     #if HAS_LCD_MENU
       synchronize(PSTR(MSG_PAUSE_PRINT));
