diff --git a/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParser.java b/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParser.java
index 797b3d9ab5b6..6c8f31cff6ee 100644
--- a/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParser.java
+++ b/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParser.java
@@ -152,7 +152,7 @@ public boolean token(String currentFieldName, XContentParser.Token token, XConte
             return config;
         }
 
-        FieldMapper<?> mapper = context.smartNameFieldMapper(input.field);
+        FieldMapper<?> mapper = context.mapperService().smartNameFieldMapper(input.field);
         if (mapper == null) {
             Class<VS> valuesSourceType = valueType != null ? (Class<VS>) valueType.getValuesSourceType() : this.valuesSourceType;
             ValuesSourceConfig<VS> config = new ValuesSourceConfig<>(valuesSourceType);
diff --git a/src/test/java/org/elasticsearch/search/aggregations/bucket/ChildrenTests.java b/src/test/java/org/elasticsearch/search/aggregations/bucket/ChildrenTests.java
index 967708286ee6..4c7b1d29ca3e 100644
--- a/src/test/java/org/elasticsearch/search/aggregations/bucket/ChildrenTests.java
+++ b/src/test/java/org/elasticsearch/search/aggregations/bucket/ChildrenTests.java
@@ -310,8 +310,8 @@ public void testPostCollection() throws Exception {
                 .setQuery(hasChildQuery(childType, termQuery("color", "orange")))
                 .addAggregation(children("my-refinements")
                                 .childType(childType)
-                                .subAggregation(terms("my-colors").field(childType + ".color"))
-                                .subAggregation(terms("my-sizes").field(childType + ".size"))
+                                .subAggregation(terms("my-colors").field("color"))
+                                .subAggregation(terms("my-sizes").field("size"))
                 ).get();
         assertNoFailures(response);
         assertHitCount(response, 1);
