diff --git a/numpy/distutils/misc_util.py b/numpy/distutils/misc_util.py
index 38edf4691b2..1d08942f6c1 100644
--- a/numpy/distutils/misc_util.py
+++ b/numpy/distutils/misc_util.py
@@ -2304,7 +2304,8 @@ def generate_config_py(target):
         windll.kernel32.SetDefaultDllDirectories(0x1000)
     except AttributeError:
         def _AddDllDirectory(dll_directory):
-            os.environ["PATH"] += os.pathsep + dll_directory
+            os.environ.setdefault('PATH', '')
+            os.environ['PATH'] += os.pathsep + dll_directory
 
     _AddDllDirectory(extra_dll_dir)
 
