diff --git a/docs/plugins/repository-azure.asciidoc b/docs/plugins/repository-azure.asciidoc
index df09b28093c8..9249efd5d174 100644
--- a/docs/plugins/repository-azure.asciidoc
+++ b/docs/plugins/repository-azure.asciidoc
@@ -124,7 +124,7 @@ The Azure repository supports following settings:
 
     When set to `true` metadata files are stored in compressed format. This
     setting doesn't affect index files that are already compressed by default.
-    Defaults to `false`.
+    Defaults to `true`.
 
 `readonly`::
 
diff --git a/docs/plugins/repository-gcs.asciidoc b/docs/plugins/repository-gcs.asciidoc
index e3978e65f447..f655d2930707 100644
--- a/docs/plugins/repository-gcs.asciidoc
+++ b/docs/plugins/repository-gcs.asciidoc
@@ -238,7 +238,7 @@ The following settings are supported:
 
     When set to `true` metadata files are stored in compressed format. This
     setting doesn't affect index files that are already compressed by default.
-    Defaults to `false`.
+    Defaults to `true`.
 
 `application_name`::
 
diff --git a/docs/plugins/repository-hdfs.asciidoc b/docs/plugins/repository-hdfs.asciidoc
index 005cc3089555..e798682a3869 100644
--- a/docs/plugins/repository-hdfs.asciidoc
+++ b/docs/plugins/repository-hdfs.asciidoc
@@ -62,7 +62,7 @@ The following settings are supported:
 
 `compress`::
 
-    Whether to compress the metadata or not. (Disabled by default)
+    Whether to compress the metadata or not. (Enabled by default)
 
 `chunk_size`::
 
diff --git a/docs/plugins/repository-s3.asciidoc b/docs/plugins/repository-s3.asciidoc
index 2ce6ae1121b5..7c4e763a3b04 100644
--- a/docs/plugins/repository-s3.asciidoc
+++ b/docs/plugins/repository-s3.asciidoc
@@ -211,7 +211,7 @@ The following settings are supported:
 
     When set to `true` metadata files are stored in compressed format. This
     setting doesn't affect index files that are already compressed by default.
-    Defaults to `false`.
+    Defaults to `true`.
 
 `server_side_encryption`::
 
diff --git a/docs/reference/migration/migrate_8_0/snapshots.asciidoc b/docs/reference/migration/migrate_8_0/snapshots.asciidoc
index 52db460fa7a1..4a00d91e1acc 100644
--- a/docs/reference/migration/migrate_8_0/snapshots.asciidoc
+++ b/docs/reference/migration/migrate_8_0/snapshots.asciidoc
@@ -11,3 +11,13 @@ The `repositories.fs.compress` setting has been removed.
 
 Instead use the repository specific `compress` setting to enable compression. See <<modules-snapshots>> for information
 on the `compress` setting.
+
+[float]
+==== Compression of meta data files is now on by default
+
+Previously, the default value for `compress` was `false`. The default has been changed to `true`.
+
+This change will affect both newly created repositories and existing repositories where `compress=false` has not been
+explicitly specified.
+
+For more information on the compress option, see <<modules-snapshots>>
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
index 4fee2fad4160..e71114533f3d 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -199,7 +199,7 @@
      * files that are already compressed by default. Changing the setting does not invalidate existing files since reads
      * do not observe the setting, instead they examine the file to see if it is compressed or not.
      */
-    public static final Setting<Boolean> COMPRESS_SETTING = Setting.boolSetting("compress", false, Setting.Property.NodeScope);
+    public static final Setting<Boolean> COMPRESS_SETTING = Setting.boolSetting("compress", true, Setting.Property.NodeScope);
 
     private final Settings settings;
 
diff --git a/server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java b/server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java
index 7246b708dd16..2cad99c48bc4 100644
--- a/server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java
+++ b/server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java
@@ -254,8 +254,8 @@ private BlobStoreRepository setupRepo() {
 
         Settings.Builder repoSettings = Settings.builder().put(node().settings()).put("location", location);
         boolean compress = randomBoolean();
-        if (compress) {
-            repoSettings.put(BlobStoreRepository.COMPRESS_SETTING.getKey(), true);
+        if (compress == false) {
+            repoSettings.put(BlobStoreRepository.COMPRESS_SETTING.getKey(), false);
         }
         AcknowledgedResponse putRepositoryResponse =
             client.admin().cluster().preparePutRepository(repositoryName)
