diff --git a/modules/imgproc/test/test_imgwarp.cpp b/modules/imgproc/test/test_imgwarp.cpp
index 99cffbc37dd..a14ea532a92 100644
--- a/modules/imgproc/test/test_imgwarp.cpp
+++ b/modules/imgproc/test/test_imgwarp.cpp
@@ -1420,6 +1420,8 @@ TEST(Imgproc_cvWarpAffine, regression)
     IplImage* src = cvCreateImage(cvSize(100, 100), IPL_DEPTH_8U, 1);
     IplImage* dst = cvCreateImage(cvSize(100, 100), IPL_DEPTH_8U, 1);
 
+    cvZero(src);
+
     float m[6];
     CvMat M = cvMat( 2, 3, CV_32F, m );
     int w = src->width;
