diff --git a/APMrover2/AP_MotorsUGV.cpp b/APMrover2/AP_MotorsUGV.cpp
index 3a57260cdb13..4ff5f35680d6 100644
--- a/APMrover2/AP_MotorsUGV.cpp
+++ b/APMrover2/AP_MotorsUGV.cpp
@@ -502,8 +502,8 @@ void AP_MotorsUGV::setup_omni()
 
     case FRAME_TYPE_OMNI3:
         _motors_num = 3;
-        add_omni_motor(0, 1.0f, 1.0f, -1.0f);
-        add_omni_motor(1, 0.0f, 1.0f, 1.0f);
+        add_omni_motor(0, 1.0f, -1.0f, -1.0f);
+        add_omni_motor(1, 0.0f, -1.0f, 1.0f);
         add_omni_motor(2, 1.0f, 1.0f, 1.0f);
         break;
 
