diff --git a/ext/standard/info.c b/ext/standard/info.c
index d6911a42ddd9..9c895255751c 100644
--- a/ext/standard/info.c
+++ b/ext/standard/info.c
@@ -37,6 +37,7 @@
 #ifdef HAVE_SYS_UTSNAME_H
 #include <sys/utsname.h>
 #endif
+#include "url.h"
 
 
 #ifdef PHP_WIN32
@@ -146,8 +147,7 @@ PHPAPI void php_info_print_module(zend_module_entry *zend_module TSRMLS_DC) /* {
 {
 	if (zend_module->info_func || zend_module->version) {
 		if (!sapi_module.phpinfo_as_text) {
-			int len = 0;
-			zend_string *url_name = php_url_encode(zend_module->name, strlen(zend_module->name), &len);
+			zend_string *url_name = php_url_encode(zend_module->name, strlen(zend_module->name));
 
 			php_strtolower(url_name->val, url_name->len);
 			php_info_printf("<h2><a name=\"module_%s\">%s</a></h2>\n", url_name->val, zend_module->name);
diff --git a/ext/standard/tests/general_functions/phpinfo2.phpt b/ext/standard/tests/general_functions/phpinfo2.phpt
index 891867f2eca2..9f3bf9a17b63 100644
--- a/ext/standard/tests/general_functions/phpinfo2.phpt
+++ b/ext/standard/tests/general_functions/phpinfo2.phpt
@@ -17,7 +17,7 @@ var_dump(phpinfo(INFO_LICENSE));
 
 ?>
 --EXPECTF--
-<!DOCTYPE %s>
+%a<!DOCTYPE %s>
 %a</html>bool(true)
 --
 
