diff --git a/tools/interop_matrix/client_matrix.py b/tools/interop_matrix/client_matrix.py
index 22e522bf6e2..2d43e3ea12e 100644
--- a/tools/interop_matrix/client_matrix.py
+++ b/tools/interop_matrix/client_matrix.py
@@ -388,5 +388,5 @@ def should_build_docker_interop_image_from_release_tag(lang):
     'python_v1.7.2': 'python__v1.0.x',
     'python_v1.8.1': 'python__v1.0.x',
     'python_v1.9.1': 'python__v1.0.x',
-    'python_v1.10.0': 'python__v1.0.x',
+    'python_v1.10.1': 'python__v1.0.x',
 }
