diff --git a/numpy/random/mtrand/mtrand.pyx b/numpy/random/mtrand/mtrand.pyx
index 9861204d9d5..bfbeb3c7bcf 100644
--- a/numpy/random/mtrand/mtrand.pyx
+++ b/numpy/random/mtrand/mtrand.pyx
@@ -127,6 +127,7 @@ cdef extern from "initarray.h":
 # Initialize numpy
 import_array()
 
+cimport cython
 import numpy as np
 import operator
 import warnings
@@ -4475,7 +4476,7 @@ cdef class RandomState:
         mnarr = <ndarray>multin
         mnix = <long*>PyArray_DATA(mnarr)
         sz = PyArray_SIZE(mnarr)
-        with self.lock, nogil:
+        with self.lock, nogil, cython.cdivision(True):
             i = 0
             while i < sz:
                 Sum = 1.0
