diff --git a/numpy/random/__init__.py b/numpy/random/__init__.py
index 2e83e419d67..21c1adddf40 100644
--- a/numpy/random/__init__.py
+++ b/numpy/random/__init__.py
@@ -141,9 +141,12 @@
     'rand',
     'randint',
     'randn',
+    'random',
     'random_integers',
     'random_sample',
+    'ranf',
     'rayleigh',
+    'sample',
     'seed',
     'set_state',
     'shuffle',
@@ -176,10 +179,6 @@
 __all__ += ['Generator', 'DSFMT', 'MT19937', 'Philox', 'PCG64', 'PCG32',
             'ThreeFry', 'Xoshiro256', 'Xoshiro512', 'RandomState']
 
-# Some aliases:
-ranf = random = sample = random_sample
-__all__.extend(['ranf', 'random', 'sample'])
-
 
 def __RandomState_ctor():
     """Return a RandomState instance.
diff --git a/numpy/random/mtrand.pyx b/numpy/random/mtrand.pyx
index 48ff6b0a673..1d3313446fd 100644
--- a/numpy/random/mtrand.pyx
+++ b/numpy/random/mtrand.pyx
@@ -4196,7 +4196,7 @@ power = _rand.power
 rand = _rand.rand
 randint = _rand.randint
 randn = _rand.randn
-random = _rand.random_sample
+random = _rand.random
 random_integers = _rand.random_integers
 random_sample = _rand.random_sample
 rayleigh = _rand.rayleigh
@@ -4261,6 +4261,7 @@ __all__ = [
     'rand',
     'randint',
     'randn',
+    'random',
     'random_integers',
     'random_sample',
     'ranf',
