diff --git a/APMrover2/system.cpp b/APMrover2/system.cpp
index ec3c9790da3c..aee4e31dc609 100644
--- a/APMrover2/system.cpp
+++ b/APMrover2/system.cpp
@@ -27,6 +27,8 @@ void Rover::init_ardupilot()
                         AP::fwversion().fw_string,
                         (unsigned)hal.util->available_memory());
 
+    init_capabilities();
+
     //
     // Check the EEPROM format version before loading any parameters from EEPROM.
     //
@@ -142,8 +144,6 @@ void Rover::init_ardupilot()
     // initialize SmartRTL
     g2.smart_rtl.init();
 
-    init_capabilities();
-
     startup_ground();
 
     Mode *initial_mode = mode_from_mode_num((enum Mode::Number)g.initial_mode.get());
diff --git a/AntennaTracker/system.cpp b/AntennaTracker/system.cpp
index c38f111dbb4f..3cf134601844 100644
--- a/AntennaTracker/system.cpp
+++ b/AntennaTracker/system.cpp
@@ -17,6 +17,8 @@ void Tracker::init_tracker()
                         AP::fwversion().fw_string,
                         (unsigned)hal.util->available_memory());
 
+    init_capabilities();
+
     // Check the EEPROM format version before loading any parameters from EEPROM
     load_parameters();
 
@@ -100,8 +102,6 @@ void Tracker::init_tracker()
         get_home_eeprom(current_loc);
     }
 
-    init_capabilities();
-
     gcs().send_text(MAV_SEVERITY_INFO,"Ready to track");
     hal.scheduler->delay(1000); // Why????
 
diff --git a/ArduCopter/system.cpp b/ArduCopter/system.cpp
index 2939bb791ea0..ddcf95ef5834 100644
--- a/ArduCopter/system.cpp
+++ b/ArduCopter/system.cpp
@@ -23,14 +23,13 @@ void Copter::init_ardupilot()
     // initialise serial port
     serial_manager.init_console();
 
-    // init vehicle capabilties
-    init_capabilities();
-
     hal.console->printf("\n\nInit %s"
                         "\n\nFree RAM: %u\n",
                         AP::fwversion().fw_string,
                         (unsigned)hal.util->available_memory());
 
+    init_capabilities();
+
     //
     // Report firmware version code expect on console (check of actual EEPROM format version is done in load_parameters function)
     //
diff --git a/ArduPlane/system.cpp b/ArduPlane/system.cpp
index 3c449d234a09..f9ccc0453342 100644
--- a/ArduPlane/system.cpp
+++ b/ArduPlane/system.cpp
@@ -27,6 +27,7 @@ void Plane::init_ardupilot()
                         AP::fwversion().fw_string,
                         (unsigned)hal.util->available_memory());
 
+    init_capabilities();
 
     //
     // Check the EEPROM format version before loading any parameters from EEPROM
@@ -180,8 +181,6 @@ void Plane::init_ardupilot()
      */
     hal.scheduler->register_timer_failsafe(failsafe_check_static, 1000);
 
-    init_capabilities();
-
     quadplane.setup();
 
     AP_Param::reload_defaults_file(true);
diff --git a/ArduSub/system.cpp b/ArduSub/system.cpp
index 60ac2a928727..658f4e1ddec8 100644
--- a/ArduSub/system.cpp
+++ b/ArduSub/system.cpp
@@ -27,6 +27,8 @@ void Sub::init_ardupilot()
                         AP::fwversion().fw_string,
                         (unsigned)hal.util->available_memory());
 
+    init_capabilities();
+
     // load parameters from EEPROM
     load_parameters();
 
@@ -203,9 +205,6 @@ void Sub::init_ardupilot()
 
     ins.set_log_raw_bit(MASK_LOG_IMU_RAW);
 
-    // init vehicle capabilties
-    init_capabilities();
-
     // disable safety if requested
     BoardConfig.init_safety();    
     
