diff --git a/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java b/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java
index b871b4eb0b11..a96de5ef696e 100644
--- a/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java
+++ b/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java
@@ -221,8 +221,17 @@ private void stop() {
             keepAliveLatch.countDown();
         }
     }
+    
+    /** Calls doMain(), but with special formatting of errors */
+    public static void main(String[] args) throws StartupError {
+        try {
+            doMain(args);
+        } catch (Throwable t) {
+            throw new StartupError(t);
+        }
+    }
 
-    public static void main(String[] args) throws Throwable {
+    public static void doMain(String[] args) throws Throwable {
         BootstrapCLIParser bootstrapCLIParser = new BootstrapCLIParser();
         CliTool.ExitStatus status = bootstrapCLIParser.execute(args);
 
@@ -291,7 +300,7 @@ public static void main(String[] args) throws Throwable {
                 Loggers.enableConsoleLogging();
             }
             
-            throw new StartupError(e);
+            throw e;
         }
     }
 
