diff --git a/libraries/WiFi/src/WiFi.cpp b/libraries/WiFi/src/WiFi.cpp
index 4cb750e636..88b18b0d25 100644
--- a/libraries/WiFi/src/WiFi.cpp
+++ b/libraries/WiFi/src/WiFi.cpp
@@ -32,8 +32,6 @@ uint16_t 	WiFiClass::_server_port[MAX_SOCK_NUM] = { 0, 0, 0, 0 };
 
 WiFiClass::WiFiClass()
 {
-	// Driver initialization
-	init();
 }
 
 void WiFiClass::init()
diff --git a/libraries/WiFi/src/utility/spi_drv.cpp b/libraries/WiFi/src/utility/spi_drv.cpp
index 638947a66b..2bf1bb7151 100644
--- a/libraries/WiFi/src/utility/spi_drv.cpp
+++ b/libraries/WiFi/src/utility/spi_drv.cpp
@@ -43,8 +43,8 @@ void SpiDrv::begin()
 	  pinMode(SLAVEREADY, INPUT);
 	  pinMode(WIFILED, OUTPUT);
 
-	  digitalWrite(SCK, LOW);
-	  digitalWrite(MOSI, LOW);
+	  // digitalWrite(SCK, LOW);
+	  // digitalWrite(MOSI, LOW);
 	  digitalWrite(SS, HIGH);
 	  digitalWrite(SLAVESELECT, HIGH);
 	  digitalWrite(WIFILED, LOW);
diff --git a/libraries/WiFi/src/utility/spi_drv.h b/libraries/WiFi/src/utility/spi_drv.h
index da71a0bc76..ab8d97d66b 100644
--- a/libraries/WiFi/src/utility/spi_drv.h
+++ b/libraries/WiFi/src/utility/spi_drv.h
@@ -30,11 +30,15 @@
 
 #define DUMMY_DATA  0xFF
 
-#define WAIT_FOR_SLAVE_SELECT()	 \
-	SpiDrv::waitForSlaveReady(); \
+#define WAIT_FOR_SLAVE_SELECT()	      \
+	if (!initialized) {           \
+		SpiDrv::begin();      \
+		initialized = true;   \
+	}                             \
+	SpiDrv::waitForSlaveReady();  \
 	SpiDrv::spiSlaveSelect();
 
-
+static bool initialized = false;
 
 class SpiDrv
 {
