diff --git a/ext/gd/gd.c b/ext/gd/gd.c
index b0fba30bfdcc..3e7d778ba343 100644
--- a/ext/gd/gd.c
+++ b/ext/gd/gd.c
@@ -372,6 +372,8 @@ ZEND_END_ARG_INFO()
 ZEND_BEGIN_ARG_INFO_EX(arginfo_imagepng, 0, 0, 1)
 	ZEND_ARG_INFO(0, im)
 	ZEND_ARG_INFO(0, to)
+	ZEND_ARG_INFO(0, quality)
+	ZEND_ARG_INFO(0, filters)
 ZEND_END_ARG_INFO()
 #endif
 
diff --git a/ext/gd/tests/bug74031.phpt b/ext/gd/tests/bug74031.phpt
new file mode 100644
index 000000000000..e5d6d7091ec2
--- /dev/null
+++ b/ext/gd/tests/bug74031.phpt
@@ -0,0 +1,10 @@
+--TEST--
+(Bug #74031) ReflectionFunction for imagepng returns wrong number of parameters
+--FILE--
+<?php
+
+$ref = new ReflectionFunction('imagepng');
+var_dump(count($ref->getParameters()));
+?>
+--EXPECT--
+int(4)
