diff --git a/ext/exif/exif.c b/ext/exif/exif.c
index 7b51e2f9a77e..48e937216194 100644
--- a/ext/exif/exif.c
+++ b/ext/exif/exif.c
@@ -2415,7 +2415,7 @@ static void add_assoc_image_info(zval *value, int sub_array, image_info_type *im
 									break;
 
 								case TAG_FMT_URATIONAL:
-									snprintf(buffer, sizeof(buffer), "%i/%i", info_value->ur.num, info_value->ur.den);
+									snprintf(buffer, sizeof(buffer), "%u/%u", info_value->ur.num, info_value->ur.den);
 									if (l==1) {
 										add_assoc_string(&tmpi, name, buffer);
 									} else {
diff --git a/ext/exif/tests/bug77184/DJI_0245_tiny.JPG b/ext/exif/tests/bug77184/DJI_0245_tiny.JPG
new file mode 100644
index 000000000000..dd5adbd7ce1f
Binary files /dev/null and b/ext/exif/tests/bug77184/DJI_0245_tiny.JPG differ
diff --git a/ext/exif/tests/bug77184/bug77184.phpt b/ext/exif/tests/bug77184/bug77184.phpt
new file mode 100644
index 000000000000..bbe7c93b614b
--- /dev/null
+++ b/ext/exif/tests/bug77184/bug77184.phpt
@@ -0,0 +1,19 @@
+--TEST--
+Bug #74428 (Unsigned rational numbers are written out as signed rationals)
+--SKIPIF--
+<?php if (!extension_loaded('exif')) print 'skip exif extension not available';?>
+--INI--
+output_handler=
+zlib.output_compression=0
+--FILE--
+<?php
+$infile = dirname(__FILE__).'/DJI_0245_tiny.JPG';
+$tags = exif_read_data($infile);
+echo $tags['GPSLatitude'][2], PHP_EOL;
+echo $tags['GPSLongitude'][2], PHP_EOL;
+?>
+===DONE===
+--EXPECTF--
+156691934/70000000
+2672193571/70000000
+===DONE===
