diff --git a/distribution/bwc/build.gradle b/distribution/bwc/build.gradle
index f878878e6b60..a44e670542bc 100644
--- a/distribution/bwc/build.gradle
+++ b/distribution/bwc/build.gradle
@@ -179,6 +179,9 @@ subprojects {
     } else {
       executable new File(checkoutDir, 'gradlew').toString()
     }
+    if (gradle.startParameter.isOffline()) {
+      args "--offline"
+    }
     for (String dir : projectDirs) {
       args ":${dir.replace('/', ':')}:assemble"
     }
