diff --git a/core/src/main/java/org/elasticsearch/index/query/SimpleQueryParser.java b/core/src/main/java/org/elasticsearch/index/query/SimpleQueryParser.java
index 9ae0703282f2..ff10b3f05ce5 100644
--- a/core/src/main/java/org/elasticsearch/index/query/SimpleQueryParser.java
+++ b/core/src/main/java/org/elasticsearch/index/query/SimpleQueryParser.java
@@ -128,8 +128,10 @@ public Query newPrefixQuery(String text) {
             try {
                 if (settings.analyzeWildcard()) {
                     Query analyzedQuery = newPossiblyAnalyzedQuery(entry.getKey(), text);
-                    analyzedQuery.setBoost(entry.getValue());
-                    bq.add(analyzedQuery, BooleanClause.Occur.SHOULD);
+                    if (analyzedQuery != null) {
+                        analyzedQuery.setBoost(entry.getValue());
+                        bq.add(analyzedQuery, BooleanClause.Occur.SHOULD);
+                    }
                 } else {
                     PrefixQuery prefix = new PrefixQuery(new Term(entry.getKey(), text));
                     prefix.setBoost(entry.getValue());
diff --git a/core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java b/core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java
index e41c451b981a..902490035095 100644
--- a/core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java
+++ b/core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java
@@ -346,4 +346,31 @@ public void testSimpleQueryStringAnalyzeWildcard() throws ExecutionException, In
         assertSearchHits(searchResponse, "1");
     }
 
+    @Test
+    public void testEmptySimpleQueryStringWithAnalysis() throws Exception {
+        // https://github.com/elastic/elasticsearch/issues/18202
+        String mapping = XContentFactory.jsonBuilder()
+                .startObject()
+                .startObject("type1")
+                .startObject("properties")
+                .startObject("body")
+                .field("type", "string")
+                .field("analyzer", "stop")
+                .endObject()
+                .endObject()
+                .endObject()
+                .endObject().string();
+
+        CreateIndexRequestBuilder mappingRequest = client().admin().indices()
+                .prepareCreate("test1")
+                .addMapping("type1", mapping);
+        mappingRequest.execute().actionGet();
+        indexRandom(true, client().prepareIndex("test1", "type1", "1").setSource("body", "Some Text"));
+        refresh();
+
+        SearchResponse searchResponse = client().prepareSearch()
+                .setQuery(simpleQueryStringQuery("the*").analyzeWildcard(true).field("body")).get();
+        assertNoFailures(searchResponse);
+        assertHitCount(searchResponse, 0l);
+    }
 }
