diff --git a/ext/spl/spl_directory.c b/ext/spl/spl_directory.c
index 41654625a479..bc5ac46d88ef 100644
--- a/ext/spl/spl_directory.c
+++ b/ext/spl/spl_directory.c
@@ -307,7 +307,7 @@ static int spl_filesystem_file_open(spl_filesystem_object *intern, int use_inclu
 
 	intern->u.file.delimiter = ',';
 	intern->u.file.enclosure = '"';
-	intern->u.file.escape = '\\';
+	intern->u.file.escape = '\0';
 
 	intern->u.file.func_getCurr = zend_hash_str_find_ptr(&intern->std.ce->function_table, "getcurrentline", sizeof("getcurrentline") - 1);
 
diff --git a/ext/spl/tests/SplFileObject_fgetcsv_escape_default.phpt b/ext/spl/tests/SplFileObject_fgetcsv_escape_default.phpt
index 69089636c116..4dd6e2472644 100644
--- a/ext/spl/tests/SplFileObject_fgetcsv_escape_default.phpt
+++ b/ext/spl/tests/SplFileObject_fgetcsv_escape_default.phpt
@@ -3,7 +3,7 @@ SplFileObject::fgetcsv with default escape character
 --FILE--
 <?php
 $fp = fopen('SplFileObject__fgetcsv7.csv', 'w+');
-fwrite($fp, '"aa\"","bb","\"c"');
+fwrite($fp, '"aa\""","bb","\""c"');
 fclose($fp);
 
 $fo = new SplFileObject('SplFileObject__fgetcsv7.csv');
diff --git a/ext/spl/tests/SplFileObject_fputcsv.phpt b/ext/spl/tests/SplFileObject_fputcsv.phpt
index 66fdbfd1b9be..20c02cc588ed 100644
--- a/ext/spl/tests/SplFileObject_fputcsv.phpt
+++ b/ext/spl/tests/SplFileObject_fputcsv.phpt
@@ -76,10 +76,10 @@ $list = array (
   13 => 'aaa,"""bbb   """',
   14 => '"aaa""aaa""","""bbb""bbb"',
   15 => '"aaa""aaa""""""",bbb',
-  16 => 'aaa,"""\\"bbb",ccc',
-  17 => '"aaa""\\"a""","""bbb"""',
-  18 => '"""\\"""","""aaa"""',
-  19 => '"""\\"""""",aaa',
+  16 => 'aaa,"""\\""bbb",ccc',
+  17 => '"aaa""\\""a""","""bbb"""',
+  18 => '"""\\""""","""aaa"""',
+  19 => '"""\\""""""",aaa',
 );
 $list = array (
   0 => 'aaa,bbb',
diff --git a/ext/spl/tests/SplFileObject_fputcsv_variation10.phpt b/ext/spl/tests/SplFileObject_fputcsv_variation10.phpt
index 08a2461b4dd8..138c862dfcfa 100644
--- a/ext/spl/tests/SplFileObject_fputcsv_variation10.phpt
+++ b/ext/spl/tests/SplFileObject_fputcsv_variation10.phpt
@@ -61,87 +61,87 @@ echo "Done\n";
 *** Testing fputcsv() : with no CSV format in the field ***
 
 -- file opened in r+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in r+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in r+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in a+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in a+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in a+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in w+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in w+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in w+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in x+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in x+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in x+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in r+ --
diff --git a/ext/spl/tests/SplFileObject_getCsvControl_basic_001.phpt b/ext/spl/tests/SplFileObject_getCsvControl_basic_001.phpt
index 4e3de4bab2eb..2259f15a0799 100644
--- a/ext/spl/tests/SplFileObject_getCsvControl_basic_001.phpt
+++ b/ext/spl/tests/SplFileObject_getCsvControl_basic_001.phpt
@@ -13,5 +13,5 @@ array(3) {
   [1]=>
   string(1) """
   [2]=>
-  string(1) "\"
+  string(0) ""
 }
diff --git a/ext/standard/file.c b/ext/standard/file.c
index c2711225bd75..f081da377ffc 100644
--- a/ext/standard/file.c
+++ b/ext/standard/file.c
@@ -1864,7 +1864,7 @@ PHP_FUNCTION(fputcsv)
 {
 	char delimiter = ',';	 /* allow this to be set as parameter */
 	char enclosure = '"';	 /* allow this to be set as parameter */
-	char escape_char = '\\'; /* allow this to be set as parameter */
+	char escape_char = '\0'; /* allow this to be set as parameter */
 	php_stream *stream;
 	zval *fp = NULL, *fields = NULL;
 	size_t ret;
@@ -1987,7 +1987,7 @@ PHP_FUNCTION(fgetcsv)
 {
 	char delimiter = ',';	/* allow this to be set as parameter */
 	char enclosure = '"';	/* allow this to be set as parameter */
-	char escape = '\\';
+	char escape = '\0';
 
 	/* first section exactly as php_fgetss */
 
@@ -2231,7 +2231,7 @@ PHPAPI void php_fgetcsv(php_stream *stream, char delimiter, char enclosure, char
 							default:
 								if (*bptr == enclosure) {
 									state = 2;
-								} else if (*bptr == escape_char) {
+								} else if (*bptr != '\0' && *bptr == escape_char) {
 									state = 1;
 								}
 								bptr++;
diff --git a/ext/standard/tests/file/bug43225.phpt b/ext/standard/tests/file/bug43225.phpt
new file mode 100644
index 000000000000..b4685b091062
--- /dev/null
+++ b/ext/standard/tests/file/bug43225.phpt
@@ -0,0 +1,7 @@
+--TEST--
+Bug #43225: fputcsv incorrectly handles cells ending in \ followed by "
+--FILE--
+<?php
+fputcsv(STDOUT, ['a\\"', 'bbb']);
+--EXPECT--
+"a\""",bbb
diff --git a/ext/standard/tests/file/fgetcsv.phpt b/ext/standard/tests/file/fgetcsv.phpt
index 868212c033b6..2cc8b185474f 100644
--- a/ext/standard/tests/file/fgetcsv.phpt
+++ b/ext/standard/tests/file/fgetcsv.phpt
@@ -13,11 +13,11 @@ various fgetcsv() functionality tests
 		'aaa,',
 		',"aaa"',
 		'"",""',
-		'"\\"","aaa"',
+		'"\\""","aaa"',
 		'"""""",',
 		'""""",aaa',
 		'"\\""",aaa',
-		'aaa,"\\"bbb,ccc',
+		'aaa,"\\""bbb",ccc',
 		'aaa,bbb   ',
 		'aaa,"bbb   "',
 		'aaa"aaa","bbb"bbb',
@@ -114,17 +114,19 @@ array(1) {
   string(7) """,aaa
 "
 }
-array(1) {
+array(2) {
   [0]=>
-  string(8) "\"",aaa
-"
+  string(2) "\""
+  [1]=>
+  string(3) "aaa"
 }
-array(2) {
+array(3) {
   [0]=>
   string(3) "aaa"
   [1]=>
-  string(10) "\"bbb,ccc
-"
+  string(5) "\"bbb"
+  [2]=>
+  string(3) "ccc"
 }
 array(2) {
   [0]=>
diff --git a/ext/standard/tests/file/fputcsv.phpt b/ext/standard/tests/file/fputcsv.phpt
index 63c41509bdd9..2b0550dc5e9a 100644
--- a/ext/standard/tests/file/fputcsv.phpt
+++ b/ext/standard/tests/file/fputcsv.phpt
@@ -75,10 +75,10 @@ $list = array (
   13 => 'aaa,"""bbb   """',
   14 => '"aaa""aaa""","""bbb""bbb"',
   15 => '"aaa""aaa""""""",bbb',
-  16 => 'aaa,"""\\"bbb",ccc',
-  17 => '"aaa""\\"a""","""bbb"""',
-  18 => '"""\\"""","""aaa"""',
-  19 => '"""\\"""""",aaa',
+  16 => 'aaa,"""\\""bbb",ccc',
+  17 => '"aaa""\\""a""","""bbb"""',
+  18 => '"""\\""""","""aaa"""',
+  19 => '"""\\""""""",aaa',
 );
 $list = array (
   0 => 'aaa,bbb',
diff --git a/ext/standard/tests/file/fputcsv_variation10.phpt b/ext/standard/tests/file/fputcsv_variation10.phpt
index 21d845ffa769..3350481a2146 100644
--- a/ext/standard/tests/file/fputcsv_variation10.phpt
+++ b/ext/standard/tests/file/fputcsv_variation10.phpt
@@ -70,87 +70,87 @@ echo "Done\n";
 *** Testing fputcsv() : with no CSV format in the field ***
 
 -- file opened in r+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in r+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in r+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in a+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in a+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in a+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in w+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in w+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in w+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in x+ --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in x+b --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(16) ""water_fruit\n"
+string(14) "water_fruit\n
 "
 
 -- file opened in x+t --
-int(16)
-int(16)
+int(14)
+int(14)
 bool(false)
-string(%d) ""water_fruit\n"
+string(%d) "water_fruit\n
 "
 
 -- file opened in r+ --
diff --git a/ext/standard/tests/file/test2.csv b/ext/standard/tests/file/test2.csv
index d8164641703c..5d992df91808 100644
--- a/ext/standard/tests/file/test2.csv
+++ b/ext/standard/tests/file/test2.csv
@@ -1 +1 @@
-"One","\"Two\"","Three\"","Four","\\","\\\\\\\\\\\\\\\\\\\\\\\"\\\\"
+"One","\""Two\""","Three\""","Four","\\","\\\\\\\\\\\\\\\\\\\\\\\""\\\\"
