diff --git a/src/main/java/org/elasticsearch/gateway/GatewayService.java b/src/main/java/org/elasticsearch/gateway/GatewayService.java
index 827a6559bf9c..b5468aaf29e9 100644
--- a/src/main/java/org/elasticsearch/gateway/GatewayService.java
+++ b/src/main/java/org/elasticsearch/gateway/GatewayService.java
@@ -104,6 +104,7 @@ public GatewayService(Settings settings, Gateway gateway, AllocationService allo
     @Override
     protected void doStart() throws ElasticsearchException {
         gateway.start();
+        clusterService.addLast(this);
         // if we received initial state, see if we can recover within the start phase, so we hold the
         // node from starting until we recovered properly
         if (discoveryService.initialStateReceived()) {
@@ -114,7 +115,6 @@ protected void doStart() throws ElasticsearchException {
         } else {
             logger.debug("can't wait on start for (possibly) reading state from gateway, will do it asynchronously");
         }
-        clusterService.addLast(this);
     }
 
     @Override
