diff --git a/.travis.yml b/.travis.yml
index 6fb5d200c17..7d73787e3ea 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -62,9 +62,10 @@ script:
   - opt_set TEMP_SENSOR_0 -2
   - opt_set TEMP_SENSOR_1 1
   - opt_set TEMP_SENSOR_BED 1
+  - opt_set POWER_SUPPLY 1
   - opt_enable PIDTEMPBED FIX_MOUNTED_PROBE Z_SAFE_HOMING ARC_P_CIRCLES CNC_WORKSPACE_PLANES CNC_COORDINATE_SYSTEMS
   - opt_enable REPRAP_DISCOUNT_SMART_CONTROLLER SDSUPPORT EEPROM_SETTINGS
-  - opt_enable BLINKM PCA9632 RGB_LED NEOPIXEL_LED
+  - opt_enable BLINKM PCA9632 RGB_LED NEOPIXEL_LED AUTO_POWER_CONTROL
   - opt_enable AUTO_BED_LEVELING_LINEAR Z_MIN_PROBE_REPEATABILITY_TEST DEBUG_LEVELING_FEATURE SKEW_CORRECTION SKEW_CORRECTION_FOR_Z SKEW_CORRECTION_GCODE
   - opt_enable_adv FWRETRACT MAX7219_DEBUG LED_CONTROL_MENU CASE_LIGHT_ENABLE CASE_LIGHT_USE_NEOPIXEL
   - opt_set ABL_GRID_POINTS_X 16
diff --git a/Marlin/src/feature/power.cpp b/Marlin/src/feature/power.cpp
index d713cb3ddfe..c72d7100e4a 100644
--- a/Marlin/src/feature/power.cpp
+++ b/Marlin/src/feature/power.cpp
@@ -29,8 +29,9 @@
 #if ENABLED(AUTO_POWER_CONTROL)
 
 #include "power.h"
-#include "temperature.h"
-#include "stepper_indirection.h"
+#include "../module/temperature.h"
+#include "../module/stepper_indirection.h"
+#include "../Marlin.h"
 
 Power powerManager;
 
@@ -50,7 +51,7 @@ bool Power::is_power_needed() {
   #endif
 
   if (X_ENABLE_READ == X_ENABLE_ON || Y_ENABLE_READ == Y_ENABLE_ON || Z_ENABLE_READ == Z_ENABLE_ON ||
-      thermalManager.soft_pwm_bed > 0
+      thermalManager.soft_pwm_amount_bed > 0
       || E0_ENABLE_READ == E_ENABLE_ON // If any of the drivers are enabled...
       #if E_STEPPERS > 1
         || E1_ENABLE_READ == E_ENABLE_ON
diff --git a/Marlin/src/feature/power.h b/Marlin/src/feature/power.h
index c61c64e6f72..225e4dc6f3d 100644
--- a/Marlin/src/feature/power.h
+++ b/Marlin/src/feature/power.h
@@ -27,6 +27,8 @@
 #ifndef POWER_H
 #define POWER_H
 
+#include "../core/types.h"
+
 class Power {
   public:
     static void check();
diff --git a/Marlin/src/module/temperature.h b/Marlin/src/module/temperature.h
index 49019ebfe02..4210803011a 100644
--- a/Marlin/src/module/temperature.h
+++ b/Marlin/src/module/temperature.h
@@ -35,7 +35,7 @@
 #endif
 
 #if ENABLED(AUTO_POWER_CONTROL)
-  #include "power.h"
+  #include "../feature/power.h"
 #endif
 
 #if ENABLED(PID_EXTRUSION_SCALING)
