diff --git a/numpy/distutils/system_info.py b/numpy/distutils/system_info.py
index 1f4552aea99..21d134c6ba7 100644
--- a/numpy/distutils/system_info.py
+++ b/numpy/distutils/system_info.py
@@ -126,6 +126,8 @@
 import re
 import copy
 import warnings
+import subprocess
+
 from glob import glob
 from functools import reduce
 if sys.version_info[0] < 3:
@@ -298,13 +300,12 @@ def add_system_root(library_root):
             default_x11_include_dirs.extend(['/usr/lib/X11/include',
                                              '/usr/include/X11'])
 
-    import subprocess as sp
     tmp = None
     try:
         # Explicitly open/close file to avoid ResourceWarning when
         # tests are run in debug mode Python 3.
         tmp = open(os.devnull, 'w')
-        p = sp.Popen(["gcc", "-print-multiarch"], stdout=sp.PIPE,
+        p = subprocess.Popen(["gcc", "-print-multiarch"], stdout=subprocess.PIPE,
                      stderr=tmp)
     except (OSError, DistutilsError):
         # OSError if gcc is not installed, or SandboxViolation (DistutilsError
