diff --git a/src/main/java/org/elasticsearch/transport/PlainTransportFuture.java b/src/main/java/org/elasticsearch/transport/PlainTransportFuture.java
index 812eec9f64b0..739ad3040cfa 100644
--- a/src/main/java/org/elasticsearch/transport/PlainTransportFuture.java
+++ b/src/main/java/org/elasticsearch/transport/PlainTransportFuture.java
@@ -62,7 +62,6 @@ public V txGet(long timeout, TimeUnit unit) throws ElasticsearchException {
         } catch (TimeoutException e) {
             throw new ElasticsearchTimeoutException(e.getMessage());
         } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
             throw new ElasticsearchIllegalStateException("Future got interrupted", e);
         } catch (ExecutionException e) {
             if (e.getCause() instanceof ElasticsearchException) {
