diff --git a/core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java b/core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java
index 5f6893ed2aa3..a32bd23cccb9 100644
--- a/core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java
+++ b/core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java
@@ -24,6 +24,7 @@
 import com.google.common.base.Function;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterators;
+
 import org.apache.lucene.document.Field;
 import org.apache.lucene.document.FieldType;
 import org.apache.lucene.index.IndexOptions;
@@ -467,9 +468,17 @@ protected void doXContentAnalyzers(XContentBuilder builder, boolean includeDefau
             if (includeDefaults) {
                 builder.field("analyzer", "default");
             }
-        } else if (includeDefaults || fieldType().indexAnalyzer().name().startsWith("_") == false && fieldType().indexAnalyzer().name().equals("default") == false) {
-            builder.field("analyzer", fieldType().indexAnalyzer().name());
-            if (fieldType().searchAnalyzer().name().equals(fieldType().indexAnalyzer().name()) == false) {
+        } else {
+            boolean writeSearchAnalyzer = includeDefaults || fieldType().searchAnalyzer().name().startsWith("_") == false
+                    && fieldType().searchAnalyzer().name().equals("default") == false
+                    && fieldType().searchAnalyzer().name().equals(fieldType().indexAnalyzer().name()) == false;
+            // If we are going to write the search_analyzer then we need to write the analyzer as well since the search_analyzer 
+            // should not be specified without the analyzer
+            if (writeSearchAnalyzer || includeDefaults || fieldType().indexAnalyzer().name().startsWith("_") == false
+                    && fieldType().indexAnalyzer().name().equals("default") == false) {
+                builder.field("analyzer", fieldType().indexAnalyzer().name());
+            }
+            if (writeSearchAnalyzer) {
                 builder.field("search_analyzer", fieldType().searchAnalyzer().name());
             }
         }
diff --git a/core/src/main/java/org/elasticsearch/index/mapper/core/TypeParsers.java b/core/src/main/java/org/elasticsearch/index/mapper/core/TypeParsers.java
index 0588bd1e0449..433cb585362f 100644
--- a/core/src/main/java/org/elasticsearch/index/mapper/core/TypeParsers.java
+++ b/core/src/main/java/org/elasticsearch/index/mapper/core/TypeParsers.java
@@ -291,7 +291,13 @@ public static void parseField(FieldMapper.Builder builder, String name, Map<Stri
 
         if (indexAnalyzer == null) {
             if (searchAnalyzer != null) {
-                throw new MapperParsingException("analyzer on field [" + name + "] must be set when search_analyzer is set");
+                // If the index was created before 2.0 then we are trying to upgrade the mappings so use the default indexAnalyzer 
+                // instead of throwing an exception so the user is able to upgrade
+                if (parserContext.indexVersionCreated().before(Version.V_2_0_0_beta1)) {
+                    indexAnalyzer = parserContext.analysisService().defaultIndexAnalyzer();
+                } else {
+                    throw new MapperParsingException("analyzer on field [" + name + "] must be set when search_analyzer is set");
+                }
             }
         } else if (searchAnalyzer == null) {
             searchAnalyzer = indexAnalyzer;
diff --git a/core/src/test/java/org/elasticsearch/index/mapper/string/SimpleStringMappingTests.java b/core/src/test/java/org/elasticsearch/index/mapper/string/SimpleStringMappingTests.java
index 8315ce55b2ff..e8ed6a98e50c 100644
--- a/core/src/test/java/org/elasticsearch/index/mapper/string/SimpleStringMappingTests.java
+++ b/core/src/test/java/org/elasticsearch/index/mapper/string/SimpleStringMappingTests.java
@@ -26,26 +26,29 @@
 import org.apache.lucene.index.IndexOptions;
 import org.apache.lucene.index.IndexableField;
 import org.apache.lucene.index.IndexableFieldType;
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.xcontent.ToXContent;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
 import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.common.xcontent.json.JsonXContent;
-import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.fielddata.FieldDataType;
 import org.elasticsearch.index.mapper.ContentPath;
 import org.elasticsearch.index.mapper.DocumentMapper;
 import org.elasticsearch.index.mapper.DocumentMapperParser;
 import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.Mapper;
 import org.elasticsearch.index.mapper.Mapper.BuilderContext;
+import org.elasticsearch.index.mapper.MapperParsingException;
 import org.elasticsearch.index.mapper.MergeResult;
 import org.elasticsearch.index.mapper.ParseContext.Document;
 import org.elasticsearch.index.mapper.ParsedDocument;
 import org.elasticsearch.index.mapper.core.StringFieldMapper;
-import org.elasticsearch.index.mapper.MapperParsingException;
-import org.elasticsearch.Version;
+import org.elasticsearch.index.mapper.core.StringFieldMapper.Builder;
 import org.elasticsearch.test.ESSingleNodeTestCase;
 import org.elasticsearch.test.VersionUtils;
 import org.junit.Before;
@@ -54,11 +57,11 @@
 import java.util.Arrays;
 import java.util.Map;
 
-import static org.elasticsearch.index.mapper.core.StringFieldMapper.Builder;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
-import static org.hamcrest.Matchers.containsString;
 
 /**
  */
@@ -567,4 +570,59 @@ public void testBackwardCompatible() throws Exception {
 
         assertThat(parser.parse(mapping).mapping().toString(), containsString("\"position_increment_gap\":10"));
     }
+
+    /**
+     * Test backward compatibility when a search_analyzer is specified without
+     * an index_analyzer
+     */
+    public void testBackwardCompatibleSearchAnalyzerMigration() throws Exception {
+
+        Settings settings = Settings.settingsBuilder()
+                .put(IndexMetaData.SETTING_VERSION_CREATED, VersionUtils.randomVersionBetween(random(), Version.V_1_0_0, Version.V_1_7_1))
+                .build();
+
+        DocumentMapperParser parser = createIndex("backward_compatible_index", settings).mapperService().documentMapperParser();
+
+        String mapping = XContentFactory.jsonBuilder().startObject().startObject("type").startObject("properties").startObject("field1")
+                .field("type", "string").field("search_analyzer", "keyword").endObject().endObject().endObject().endObject().string();
+        parser.parse(mapping);
+
+        assertThat(parser.parse(mapping).mapping().toString(), containsString("\"search_analyzer\":\"keyword\""));
+        assertThat(parser.parse(mapping).mapping().toString(), containsString("\"analyzer\":\"default\""));
+        assertThat(parser.parse(mapping).mapping(), notNullValue());
+        assertThat(parser.parse(mapping).mapping().root(), notNullValue());
+        Mapper mapper = parser.parse(mapping).mapping().root().getMapper("field1");
+        assertThat(mapper, notNullValue());
+        assertThat(mapper, instanceOf(StringFieldMapper.class));
+        StringFieldMapper stringMapper = (StringFieldMapper) mapper;
+        MappedFieldType fieldType = stringMapper.fieldType();
+        assertThat(fieldType, notNullValue());
+        assertThat(fieldType.indexAnalyzer(), notNullValue());
+        assertThat(fieldType.indexAnalyzer().name(), equalTo("default"));
+        assertThat(fieldType.searchAnalyzer().name(), equalTo("keyword"));
+    }
+
+    /**
+     * Test for indexes created on or after 2.0 an index analyzer must be
+     * specified when declaring a search analyzer
+     */
+    public void testSearchAnalyzer() throws Exception {
+
+        try {
+            Settings settings = Settings
+                    .settingsBuilder()
+                    .put(IndexMetaData.SETTING_VERSION_CREATED,
+                            VersionUtils.randomVersionBetween(random(), Version.V_2_0_0, Version.CURRENT)).build();
+
+            DocumentMapperParser parser = createIndex("index", settings).mapperService().documentMapperParser();
+
+            String mapping = XContentFactory.jsonBuilder().startObject().startObject("type").startObject("properties")
+                    .startObject("field1").field("type", "string").field("search_analyzer", "keyword").endObject().endObject().endObject()
+                    .endObject().string();
+            parser.parse(mapping);
+            fail("Expected a MapperParsingException");
+        } catch (MapperParsingException e) {
+            assertThat(e.getMessage(), equalTo("analyzer on field [field1] must be set when search_analyzer is set"));
+        }
+    }
 }
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.Beta1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.Beta1.zip
index 5bbdea4a9632..c3b4b8b99214 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.Beta1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.Beta1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC1.zip
index d9072ce465c8..04afef478042 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC2.zip
index dce299b7d608..1d307ed11c5a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.RC2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.zip
index 3ec908ddc27e..02d956cbe497 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.1.zip
index 67db98fc3ce9..edebf819c5f4 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.10.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.10.zip
index 6bdb9f27c8d2..35c84cda25df 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.10.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.10.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.11.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.11.zip
index b5253f9a673b..cccd306fa495 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.11.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.11.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.12.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.12.zip
index 0392049bb9dd..4107da372455 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.12.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.12.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.13.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.13.zip
index 025b4c38b13f..7437aeafa565 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.13.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.13.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.2.zip
index 413e08e65816..b3f68352965a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.3.zip
index c31d4de7c5bb..3597c0e56d56 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.4.zip
index 8b07a9249353..01a85295a99b 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.5.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.5.zip
index dfd0fd09e7bc..dc3a0468df4a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.5.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.5.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.6.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.6.zip
index 1f3cff28e7a2..2bb8abdba071 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.6.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.6.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.7.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.7.zip
index 6d0e65c28c42..3c9441a85cda 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.7.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.7.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.8.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.8.zip
index 8ff8ac3ddf2f..22e732239be5 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.8.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.8.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.9.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.9.zip
index 4445b3905acc..1b70aa49f251 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.9.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-0.90.9.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta1.zip
index 167dde888dfe..ceded1555748 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta2.zip
index 95fbfefb8702..572d4608f7b8 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.Beta2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC1.zip
index 3ced97aa2a99..5433cc91e020 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC2.zip
index 1298cfbfd1df..3a245d7d8507 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.RC2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.zip
index 2cb9abc43dcc..3915db767ee4 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.1.zip
index 844271b58c45..fa122a192ca5 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.2.zip
index dd8e39388dff..c862c40450d8 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.3.zip
index e4437effa243..2f0cfd72e274 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.0.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.0.zip
index 4f05370b1298..95a4f4f56344 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.1.zip
index effeb94287b2..51323d9b4287 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.2.zip
index bedffa44f195..ea098e5d9019 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.1.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.0.zip
index 4644a3893337..cf1501027dee 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.1.zip
index 553b46b06ad3..808780bd7146 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.2.zip
index 3f51a47534e0..ee2f3138a18d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.3.zip
index 8c8bfbd908b2..32a04ada9228 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.4.zip
index e3a151930e7b..81cdb56a76bf 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.2.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.0.zip
index d98958dea334..4ec5219f4b1a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.1.zip
index 167d0f4c9463..9f12362ad872 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.2.zip
index 756eaf68ac2c..591e195b03f8 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.3.zip
index 8470deefd8f3..df364b82df23 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.4.zip
index 2175012fc6dc..d8dbc0f4c825 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.5.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.5.zip
index 19d1e568d849..7d386d7bdcc8 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.5.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.5.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.6.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.6.zip
index ad8e8bd003c6..ef53542683cf 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.6.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.6.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.7.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.7.zip
index 3a645a9dd55c..c1d2bd6f639b 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.7.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.7.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.8.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.8.zip
index f8ab0a26dc62..73fc42175f3d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.8.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.8.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.9.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.9.zip
index 5ef35b21ced7..af3457c0d7c7 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.9.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.3.9.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.Beta1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.Beta1.zip
index 4546f5d4aa32..c09e555dd2f6 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.Beta1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.Beta1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.zip
index 467d19aa5edc..6fb7e4969197 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.1.zip
index 2adbb28caedb..7bc5e9e4382e 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.2.zip
index 4fac2086e5d5..01837f7de94f 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.3.zip
index 1a0d66745a99..81ae3637502c 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.4.zip
index 0328a9e2d570..3b393bbbcf63 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.5.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.5.zip
index eeb25ab6ed95..b13d9e969316 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.5.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.4.5.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.0.zip
index f1dab0883149..40e67cf1177b 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.1.zip
index 342e311e56db..547b41f51cdc 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.2.zip
index fb36b197c4a8..a338556e32e7 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.5.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.0.zip
index 02a5806638b6..6a4c61ce6d0d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.1.zip
index 04820f92b070..072fa01754b3 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.2.zip
index af6ce561fa6a..1e248741b49e 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.6.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.0.zip
index 941be645c65b..1b612babeec1 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.1.zip
index debd797162b7..6db7803519c2 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.2.zip
index 18bb6c7dfc9b..2fd0adc4b54d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.3.zip
index 9fcc1788ea95..33c46d3af630 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-1.7.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta1.zip
index c0054fd37408..29446794e0a4 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta2.zip
index 169a0b71d750..6ea7cb79b999 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-beta2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-rc1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-rc1.zip
index 44e78b4a9054..a6e70370aff3 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-rc1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0-rc1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0.zip
index 7110fb424a8b..c3748a99e6db 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/index-2.0.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.Beta2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.Beta2.zip
index 020f6f4e59fd..7e3a160a63cb 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.Beta2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.Beta2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC1.zip
index a84c50759a7d..2839677b45c2 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC2.zip
index e5d65ebaf499..2657dd2de6a1 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.RC2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.zip
index 13f778d965b7..1f1b679cdd46 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.1.zip
index 76ed278fade3..0afbef3cb513 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.2.zip
index 762eabf82864..bf102b257eb7 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.3.zip
index ba79dea3459b..8ccced841e9a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.0.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.0.zip
index cbf84c717ac2..3d8602a0a448 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.1.zip
index 00ea0447b261..a130ef00b2c6 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.2.zip
index 18abd996259b..f6b6cb4529ed 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.1.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.0.zip
index f5e62a1a94b5..186d3c25aad2 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.1.zip
index 935e71b0c126..80e8ba2ef571 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.2.zip
index d69b22a0887f..c8cc0d55200a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.3.zip
index 295f9f758cc3..f8290d686cc0 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.4.zip
index e9efc00c97f3..7d9bb1b3a8c8 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.2.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.0.zip
index 5a59e21930d9..71ebe1122b51 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.1.zip
index 2ae1d7c0ad3e..4de20b038b94 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.2.zip
index c67b99717a5f..4dd61640304e 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.3.zip
index 64e3235142dc..c4ab16b90b41 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.4.zip
index 55e67445549c..b214ccf437c6 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.5.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.5.zip
index 35a5fd4836c3..c9e420ed4596 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.5.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.5.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.6.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.6.zip
index f1eb21ca4b0d..fb20176cd29d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.6.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.6.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.7.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.7.zip
index 543c13c2ba22..01446efc017b 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.7.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.7.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.8.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.8.zip
index 93abac386f9a..df6f1c9df60f 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.8.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.8.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.9.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.9.zip
index 7dc194df58e2..df143eb9fda8 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.9.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.3.9.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.Beta1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.Beta1.zip
index 5adf788623f8..8d8c1049f382 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.Beta1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.Beta1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.zip
index 21f867c85ead..7e3915233b8a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.1.zip
index 18166eabbd97..e280c40780b3 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.2.zip
index f03625a1ce16..07f7ac9c4219 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.3.zip
index d78fbb1975bb..5fc837679ae1 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.4.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.4.zip
index 1e8944618ea4..002835458f95 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.4.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.4.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.5.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.5.zip
index fcae43982ae1..0ad763531319 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.5.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.4.5.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.0.zip
index a55e6c4c3537..214ddccb1c7a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.1.zip
index 7cccbbe25a9a..f736bc322714 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.2.zip
index cee5783e10aa..02b9223baadb 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.5.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.0.zip
index 1c31a0240be1..0dab2a222f4f 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.1.zip
index 746f3ce613c8..0d631e598c82 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.2.zip
index de4c5be79289..20f6233955fc 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.6.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.0.zip
index 893689b0f4ce..09c16d2996ef 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.0.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.1.zip
index cd5c6c07c982..bf986683ef7a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.2.zip
index a0daaac39d44..b3e3691b561d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.3.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.3.zip
index 12e13dafc597..648c8351d3f6 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.3.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-1.7.3.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta1.zip
index f57314c601cd..b3724e751fcd 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta2.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta2.zip
index 1eb19f1d61eb..f66282a19b43 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta2.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-beta2.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-rc1.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-rc1.zip
index b15cedf4d664..c0b133f8173a 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-rc1.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0-rc1.zip differ
diff --git a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0.zip b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0.zip
index 9605830a12c0..54424217102d 100644
Binary files a/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0.zip and b/core/src/test/resources/org/elasticsearch/bwcompat/repo-2.0.0.zip differ
diff --git a/dev-tools/create_bwc_index.py b/dev-tools/create_bwc_index.py
index 623b368e28c6..ede39bc0aea5 100644
--- a/dev-tools/create_bwc_index.py
+++ b/dev-tools/create_bwc_index.py
@@ -179,6 +179,10 @@ def generate_index(client, version, index_name):
           'type': 'string',
           'index_analyzer': 'standard'
         },
+        'string_with_search_analyzer': {
+          'type': 'string',
+          'search_analyzer': 'keyword'
+        }
       }
     }
     # completion type was added in 0.90.3
