diff --git a/include/grpcpp/impl/codegen/async_generic_service.h b/include/grpcpp/impl/codegen/async_generic_service.h
index 957bb776f1d..2a0e1b40881 100644
--- a/include/grpcpp/impl/codegen/async_generic_service.h
+++ b/include/grpcpp/impl/codegen/async_generic_service.h
@@ -52,12 +52,12 @@ class GenericServerContext final : public ServerContext {
 //   ServerBuilder builder;
 //   auto cq = builder.AddCompletionQueue();
 //   AsyncGenericService generic_service;
-//   builder.RegisterAsyncGeneicService(&generic_service);
+//   builder.RegisterAsyncGenericService(&generic_service);
 //   auto server = builder.BuildAndStart();
 //
 //   // request a new call
 //   GenericServerContext context;
-//   GenericAsyncReaderWriter stream;
+//   GenericServerAsyncReaderWriter stream;
 //   generic_service.RequestCall(&context, &stream, cq.get(), cq.get(), tag);
 //
 // When tag is retrieved from cq->Next(), context.method() can be used to look
