diff --git a/Zend/tests/generators/bug75396.phpt b/Zend/tests/generators/bug75396.phpt
new file mode 100644
index 000000000000..4e802eb61bdd
--- /dev/null
+++ b/Zend/tests/generators/bug75396.phpt
@@ -0,0 +1,19 @@
+--TEST--
+Bug #75396 (Exit inside generator finally results in fatal error)
+--FILE--
+<?php
+
+$gen = (function () {
+
+    yield 42;
+
+    try {
+    } finally {
+        exit;
+    }
+})();
+
+$gen->send("x");
+
+?>
+--EXPECT--
\ No newline at end of file
diff --git a/Zend/zend_generators.c b/Zend/zend_generators.c
index ce7f4374cf76..6a39a0dfe384 100644
--- a/Zend/zend_generators.c
+++ b/Zend/zend_generators.c
@@ -223,6 +223,8 @@ static void zend_generator_dtor_storage(zend_object *object) /* {{{ */
 
 		ex->opline = &ex->func->op_array.opcodes[finally_op_num];
 		generator->flags |= ZEND_GENERATOR_FORCED_CLOSE;
+		generator->flags &= ~ZEND_GENERATOR_CURRENTLY_RUNNING;
+		
 		zend_generator_resume(generator);
 	}
 }
