diff --git a/Tools/ardupilotwaf/git_submodule.py b/Tools/ardupilotwaf/git_submodule.py
index af452326f833..218656a2783e 100644
--- a/Tools/ardupilotwaf/git_submodule.py
+++ b/Tools/ardupilotwaf/git_submodule.py
@@ -80,7 +80,7 @@ def git_submodule_update(self, name):
         module_node = self.bld.srcnode.make_node(os.path.join('modules', name))
 
         tsk = self.create_task('update_submodule', submodule=name)
-        tsk.cwd = self.bld.srcnode.abspath()
+        tsk.cwd = self.bld.srcnode
         tsk.env.SUBMODULE_PATH = module_node.abspath()
 
         _submodules_tasks[name] = tsk
