diff --git a/Marlin/cardreader.cpp b/Marlin/cardreader.cpp
index 0f402029124..80e330a7281 100644
--- a/Marlin/cardreader.cpp
+++ b/Marlin/cardreader.cpp
@@ -532,7 +532,7 @@ void CardReader::write_command(char *buf) {
 }
 
 void CardReader::checkautostart(bool force) {
-  if (!force && (!autostart_stilltocheck || ELAPSED(millis(), next_autostart_ms)))
+  if (!force && (!autostart_stilltocheck || ! ELAPSED(millis(), next_autostart_ms)))
     return;
 
   autostart_stilltocheck = false;
