diff --git a/src/main/java/rx/internal/util/SynchronizedQueue.java b/src/main/java/rx/internal/util/SynchronizedQueue.java
index 9fe867d93a..8f0c4a7372 100644
--- a/src/main/java/rx/internal/util/SynchronizedQueue.java
+++ b/src/main/java/rx/internal/util/SynchronizedQueue.java
@@ -99,13 +99,25 @@ public synchronized String toString() {
     }
 
     @Override
-    public synchronized boolean equals(Object o) {
-        return list.equals(o);
+    public int hashCode() {
+        return list.hashCode();
     }
 
     @Override
-    public synchronized int hashCode() {
-        return list.hashCode();
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        SynchronizedQueue<?> other = (SynchronizedQueue<?>) obj;
+        if (list == null) {
+            if (other.list != null)
+                return false;
+        } else if (!list.equals(other.list))
+            return false;
+        return true;
     }
 
     @Override
diff --git a/src/test/java/rx/internal/util/SynchronizedQueueTest.java b/src/test/java/rx/internal/util/SynchronizedQueueTest.java
new file mode 100644
index 0000000000..98779ea1b9
--- /dev/null
+++ b/src/test/java/rx/internal/util/SynchronizedQueueTest.java
@@ -0,0 +1,15 @@
+package rx.internal.util;
+
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+public class SynchronizedQueueTest {
+    
+    @Test
+    public void testEquals() {
+         SynchronizedQueue<Object> q = new SynchronizedQueue<Object>();
+         assertTrue(q.equals(q));
+    }
+
+}
